---
title: "KRAS AALE Manuscript code 2020"
author: 'Roman E. Reggiardo'
date: '10/02/2020'
output: html_document
editor_options: 
  chunk_output_type: inline
---

# Setup:

## nb settings -- here
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
library(here)
```
## loading
```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(ggsci)
  library(ggpubr)
  library(ggsignif)
  library(ggthemes) 
  library(ggrepel)
  library(ggforce)
  library(ggpmisc)
  library(extrafont)
  library(matrixStats)
  library(ggrepel)
  library(pheatmap)
  library(viridis)
  library(VennDiagram)
  library(stringr)
  library(patchwork)
  library(fgsea)
  library(msigdbr)
  library(gtsummary)
  library(SingleCellExperiment)
  library(tximport)
  library(Seurat)
  library(janitor)
})
suppressMessages(loadfonts())

```
## setting ggplot theme
```{r}
theme_set(theme_foundation(base_size = 7, base_family = 'Helvetica') + 
            theme(plot.title = element_blank(),
                panel.background = element_rect(colour = NA),
                plot.background = element_rect(colour = NA),
                panel.border = element_rect(colour = NA),
                axis.title = element_text(face = "bold",size = rel(1)),
                axis.title.y = element_text(angle=90,vjust =2),
                axis.title.x = element_text(vjust = -0.2),
                axis.text = element_text(size = rel(1)), 
                axis.line = element_blank(),
                strip.text = element_text(),
                strip.background = element_blank(),
                legend.key.size= unit(0.2, "cm"),
                legend.spacing = unit(0, "cm"),
                legend.key = element_rect(colour = NA),
                legend.title = element_text(face="italic"),
                legend.text = element_text(face = 'bold'),
                plot.margin=margin(2,2,2,2,unit = "mm"),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(),
          ))

set_aale_theme  <- function(){theme_set(theme_foundation(base_size = 7, base_family = 'Helvetica') + 
            theme(plot.title = element_blank(),
                panel.background = element_rect(colour = NA),
                plot.background = element_rect(colour = NA),
                panel.border = element_rect(colour = NA),
                axis.title = element_text(face = "bold",size = rel(1)),
                axis.title.y = element_text(angle=90,vjust =2),
                axis.title.x = element_text(vjust = -0.2),
                axis.text = element_text(), 
                axis.line = element_blank(),
                strip.text = element_text(),
                strip.background = element_blank(),
                legend.key.size= unit(0.2, "cm"),
                legend.spacing = unit(0, "cm"),
                legend.key = element_rect(colour = NA),
                legend.title = element_text(face="italic"),
                legend.text = element_text(face = 'bold'),
                plot.margin=margin(2,2,2,2,unit = "mm"),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank()
            ))}
```
## figure dirs, and other globally relevant constants
```{r}
figure.dir <-  function(figure, name){
  return(paste0(here('figures'), '/', 'fig.', figure, '/', name))
}

homer.dir <- here('data', 'atac.seq', 'output/atac-seq-pipeline/analysis/homer.annotations/')
atac.dir <- here('data', 'atac.seq', 'output/atac-seq-pipeline/analysis/')

chr.order <- c('chr1', 'chr2', 'chr3', 
               'chr4', 'chr5', 'chr6', 
               'chr7', 'chr8', 'chr9',
               'chr10', 'chr11', 'chr12', 
               'chr13', 'chr14', 'chr15', 
               'chr16', 'chr17',
               'chr18', 'chr19', 'chr20', 
               'chr21', 'chr22', 'chrX', 'chrY')

chr.19.bins <- data.frame(chr = 'chr19',
                          bin = c('p13.3', 'p13.2', 'p13.13', 'p13.12', 'p13.11',
                                  'p12', 'q12', 'q13.11', 'q13.12', 'q13.13',
                                  'q13.2', 'q13.31', 'q13.32', 'q13.33', 'q13.41', 'q13.42',
                                  'q13.43'),
                          start = c(1, 6900001, 12600001, 13800001, 16100001, 19900001,
                                    28100001, 31900001, 35100001, 37800001, 38200001, 
                                    42900001, 44700001, 47500001, 50900001, 53100001,
                                    55800001),
                          end = c(6900000, 12600000, 13800000, 16100000, 19900000,
                                    28100000, 31900000, 35100000, 37800000, 38200000, 
                                    42900000, 44700000, 47500000, 50900000, 53100000,
                                    55800000, 58617616))

trono.znf.targets <- read_csv(here('data', 'other', 'wetlab/all.znf.te.score.csv')) %>% 
  gather(target, score, -X1)

te.families.clades <- 
  read_csv(here('data', 'other', 'reference/te.family.clade.csv'),
           col_names = c('gene', 'family', 'clade')) %>% 
  distinct()
```
## gene sets of interest
```{r echo=FALSE}
kras.isg.signature <- c("DDX58", "IFIH1", "IFNL1", "IRF9", "IFI27", "OAS2", "IFIT3", "IFIT1", "OASL", "ISG15", "OAS3", "RSAD2", "MX1", "OAS1", "IFI44", "CH25H", "IFIT2", "ISG20", "RNASEL", "IFI6", "HERC5", "HERC6", "GBP3", "GBP1", "IRF1", "ISG15")

liu.isg.signature <- c('ADAR', 'DDX60', 'HERC6', 'IRF7', 'OASL', 'PSM2', 'STAT2', 'TRIM25', 'BST2', 'DHX58', 'IFI35',
                       'ISG15', 'OGFR', 'RSAD2', 'TDRD7', 'UBE2L6', 'CASP1', 'EIF2AK2', 'IFIH1', 'ISG20', 'PARP12',
                       'RTP4', 'TRAFD1', 'USP18', 'CMPK2', 'EPSTI1', 'IFIT2', 'MX1', 'PARP14', 'SAMD9L', 'TRIM14',
                       'CXCL10', 'GBP4', 'IFIT3', 'NMI', 'PNPT1', 'SP110', 'TRIM21')

gen.32.lnc <- read_table(here('data', 'other', 'wetlab/gen.32.pure.lnc.names'), col_names = 'gene')

gs.names = c("KRAS ISG SIGNATURE" , 
             "LIU ISG SIGNATURE" ,
             'HALLMARK IFNA RESPONSE',
             'HALLMARK IFNG RESPONSE',
             'HALLMARK KRAS UP')

rig.i.mda5.induction <- 
  msigdbr(species = 'Homo sapiens') %>% 
  subset(grepl('REACTOME_DDX58', gs_name)) %>% 
  mutate(gs_name = 'RIG_I_MDA5_INDUCTION') %>% 
  select(gene_symbol) %>% 
  unlist() %>% 
  unname()

rig.i.mda5.neg.regulation <- 
  msigdbr(species = 'Homo sapiens') %>% 
  subset(grepl('OF_DDX58', gs_name)) %>% 
  mutate(gs_name = 'RIG_I_MDA5_NEG_REGULATION') %>% 
  select(gene_symbol) %>% 
  unlist() %>% 
  unname()

msig.df.h.ifng <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  subset(grepl('_GAMMA_RESPONSE', gs_name)) %>% 
  mutate(gs_name = 'HALLMARK_IFNG_RESPONSE') %>% 
  select(gene_symbol) %>% 
  unlist() %>% 
  unname()

msig.df.h.ifna <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  subset(grepl('_ALPHA_RESPONSE', gs_name)) %>% 
  mutate(gs_name = 'HALLMARK_IFNA_RESPONSE') %>% 
  select(gene_symbol) %>% 
  unlist() %>% 
  unname()

kras.signal.up <- 
  msigdbr(species = 'Homo sapiens') %>% 
  filter(gs_name == 'HALLMARK_KRAS_SIGNALING_UP')  %>% 
  select(gene_symbol) %>% 
  unlist() %>% 
  unname()



```
## global functions:
```{r}
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=width, height=height)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}

```

# Data set post-processing:

## Bulk:

### DESeqII
```{r}

ha1em.de.seq <- 
  read_csv(here('data','bulk.rna.seq','ha1em','output', 'ha1em.kras.v.ctrl.de-seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj <= 0.01)

aale.de.seq <- 
  read_csv(here('data','bulk.rna.seq','aale','output', 'aale.kras.v.ctrl.de-seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj <= 0.01)

gencode.te.locus.aale.norm.counts <- 
  read.csv(here('data','bulk.rna.seq','aale','output', 'gencode.tx.aale.kras.v.ctrl.de-seq.counts.csv')) %>% 
  rename(X = 'tx') %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

gencode.te.locus.aale.norm.counts <- 
  read.csv(here('data','bulk.rna.seq','aale','output', 'te.locus.aale.kras.v.ctrl.de-seq.counts.csv')) %>% 
  rename(X = 'tx') %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

aale.te.de.seq <- 
  read_csv(here('data','bulk.rna.seq','aale','output', 'te.locus.aale.kras.v.ctrl.de-seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj < 0.01, ! gene %in% aale.de.seq$Gene) 

zscan18.row <- 
  tibble('gene' = 'ZSCAN18', 'biotype' = 'protein_coding', 'ensg' = 'placehold',
                       'chr' = 19, 'start' = 58083839, 
         'end' = 58098210,  'strand' = '-', 'anno' ='SCAN-ZNF')

trono.de.krab.znfs <- read_csv(here('data', 'other/wetlab/all.human.znf.trono.db.csv'),
                            col_names = c('gene', 'biotype', 'ensg', 'chr',
                                          'start', 'end', 'strand', 'anno')) %>%
  mutate(anno = gsub('s','', anno),
         anno = gsub('[(]', '', anno),
         anno = gsub('[)]', '', anno)) %>%
  rbind(zscan18.row) %>%
  merge(aale.de.seq, by = 'gene')

ha1em.trono.de.krab.znfs <- read_csv(here('data', 'other/wetlab/all.human.znf.trono.db.csv'),
                            col_names = c('gene', 'biotype', 'ensg', 'chr',
                                          'start', 'end', 'strand', 'anno')) %>%
  mutate(anno = gsub('s','', anno),
         anno = gsub('[(]', '', anno),
         anno = gsub('[)]', '', anno)) %>%
  rbind(zscan18.row) %>%
  merge(ha1em.de.seq, by = 'gene')

aale.exo.locus.count <- 
  read_csv(here('data','bulk.rna.seq','aale','output', 'tx.exo.gencode.kras.v.ctrl.de-seq.counts.csv')) %>% 
  # filter(grepl('Alu', X1)) %>% 
  separate(X1, sep = '_', into = c('hg', 'rmsk', 'te', 'range')) %>% 
  merge(te.families.clades, by.x = 'te', by.y = 'gene')
  
exo.aale.de.seq <- 
  read_csv(here('data','bulk.rna.seq','exo','output', 'exo.gencode.kras.v.ctrl.de-seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj <= 0.01)

exo.aale.norm.counts <- 
  read_csv(here('data','bulk.rna.seq','aale','output', 'exo.gencode.normalized.counts.csv'))

exo.vs.intra.kras.de.seq <- 
  read_csv(here('data', 'bulk.rna.seq', 'exo.ctrl.compare/output/exo.gencode.intra.v.exo.de-seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj < 0.01) 

exo.gencode.te.locus.aale.norm.counts <- 
  read_csv(here('data', 'bulk.rna.seq', 'exo/output/exo.gencode.tx.aale.kras.v.ctrl.de-seq.counts.csv')) %>% 
  # read.csv('../data/bulk.rna.seq/exo/output/exo.gencode.tx.aale.kras.v.ctrl.de-seq.counts.csv') %>% 
  rename(X1 = 'tx') %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

exo.aale.te.de.seq <- 
  read_csv(here('data', 'bulk.rna.seq', 'aale/output/exo.te.de.seq.csv'),
           col_names = c('gene', 'baseMean', 'log2fc', 'lfcse', 'pval', 'padj'),
           skip=1) %>% 
  filter(padj < 0.01, ! gene %in% aale.de.seq$Gene) 
```
### Sleuth -- TEs
```{r}
te.clades.merge <- read_csv(here('data', 'other', 'wetlab', 'te.family.clade.csv'),
                            col_names = c('te','family','clade')) %>% 
  unique()

aale.sleuth.out <- read_csv(here('data', 'bulk.rna.seq', 'aale/output/quant.data.log2fc.csv')) %>% 
  filter(qval <= 0.05) %>% 
  separate(Name, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'te-only')

aale.exo.sleuth.out <- read_csv(here('data', 'bulk.rna.seq', 'exo/output/quant.data.log2fc.csv')) %>% 
  #filter(qval <= 0.05) %>% 
  separate(Name, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'te-only')

aale.exo.vs.intra.sleuth.out.actual <- read_csv(here('data', 'bulk.rna.seq', 'exo.kras.te.compare/output/quant.data.log2fc.csv')) %>% 
  #filter(qval <= 0.05) %>% 
  separate(Name, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'te-only')

# write_csv(aale.exo.vs.intra.sleuth.out.actual, '../data/bulk.rna.seq/exo.kras.te.compare/output/aale.exo.vs.intra.te.de.csv')

aale.exo.vs.intra.sleuth.out <- read_csv(here('data', 'bulk.rna.seq', 
                                              'exo.kras.te.compare/output/exo.te.aale.kras.v.ctrl.de-seq.counts.csv')) %>% 
  separate(X1, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'te-only')

aale.combined.sleuth.out <- 
  read_csv(here('data', 'bulk.rna.seq', 'aale/output/combined.quant.data.log2fc.csv')) %>% 
  filter(qval <= 0.05,
         ! grepl('^ENST', Name)) %>% 
  separate(Name, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'combined')

ha1em.sleuth.out <- read_csv(here('data', 'bulk.rna.seq', 'ha1em/output/quant.data.log2fc.csv')) %>% 
  filter(qval <= 0.05) %>% 
  separate(Name, sep = '=', into = c('te', 'position', NA, NA, 'strand', NA)) %>% 
  separate(position, sep = '_', into = c('pos', NA)) %>% 
  separate(pos, sep = ':', into = c('chr', 'startend')) %>% 
  separate(startend, sep = '-', into = c('start', 'stop')) %>% 
  separate(strand, sep = '_', into = c('strand', NA)) %>% 
  separate(te, sep = '_', into = c(NA, NA, 'te', NA)) %>% 
  merge(te.clades.merge, by = 'te') %>% 
  mutate(approach = 'te-only')

total.sleuth.out <- rbind(aale.sleuth.out, aale.combined.sleuth.out)

```

## Single cell:

### import and QA
```{r}
# load in alevin quant matrix as a seurat object, assess standard measurements
set.seed(431)

files <- 
  file.path(here('data', 'single.cell.rna.seq', 'output/gencode.te.locus.v.1.index.alevin.out/alevin/quants_mat.gz'))
file.exists(files)

txi <- tximport(files, type="alevin")

kras.aale.sc <- CreateSeuratObject(counts = txi$counts, min.cells = 3, 
                                   min.features = 200, project = "KRAS.AALE")

kras.aale.sc[["percent.mt"]] <- PercentageFeatureSet(kras.aale.sc, pattern = "^MT-")
kras.aale.sc[["percent.rp"]] <- PercentageFeatureSet(kras.aale.sc, pattern = '^RP')

VlnPlot(kras.aale.sc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
plot1 <- FeatureScatter(kras.aale.sc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(kras.aale.sc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3 <- FeatureScatter(kras.aale.sc, feature1 = "nCount_RNA", feature2 = "percent.rp")
# scatter plot of MT, RNA, Ribo
CombinePlots(plots = list(plot1, plot2, plot3))
```
### filter and PCA
```{r}
## Filter according to seurat guidelines
## I've reworked this extensively and, due to the relatively small number of truly variable genes and cells, I've filtered heavily
## these settings seem to provide the most robust characterization of the immune phenotype we're seeing
set.seed(431)

kras.aale.sc <- subset(kras.aale.sc, 
                       subset = nFeature_RNA >= 2000  & 
                         percent.mt < 5 & 
                         nFeature_RNA <= 7000)

kras.aale.sc <- NormalizeData(kras.aale.sc, 
                              normalization.method = "LogNormalize", scale.factor = 10000)

kras.aale.sc <- FindVariableFeatures(kras.aale.sc, selection.method = "vst", nfeatures = 500)

# vector of all remaining genes in the seurat object
all.genes <- rownames(kras.aale.sc)

kras.aale.sc <- ScaleData(kras.aale.sc, features = all.genes)
kras.aale.sc <- RunPCA(kras.aale.sc, features = VariableFeatures(object = kras.aale.sc))

# jackstraw is just for EDA, arguments here won't affect the clustering, but results should be used to aid
# in determining appropriate UMAP args
kras.aale.sc <- JackStraw(kras.aale.sc, num.replicate = 100)
kras.aale.sc <- ScoreJackStraw(kras.aale.sc, dims = 1:20)

JackStrawPlot(kras.aale.sc, dims = 1:20)

ElbowPlot(kras.aale.sc)
```
### clustering
```{r}
## run clustering via suerat with manually determined args
set.seed(431)
# knn, umap
kras.aale.sc <- FindNeighbors(kras.aale.sc, dims = 1:10)
kras.aale.sc <- FindClusters(kras.aale.sc, resolution = 0.6)
kras.aale.sc <- RunUMAP(kras.aale.sc, dims = 1:10)
# assess QC features on a cluster-to-cluster basis to detect bias
VlnPlot(kras.aale.sc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))

```
### tidying normalized counts
```{r}
# extract scaled, normalized counts
kras.aale.sc.norm.counts <- as.data.frame(as.matrix(kras.aale.sc[["RNA"]]@data))
# extract barcodes
kras.barcode.counts <- 
  as.data.frame(as.matrix(kras.aale.sc@active.ident)) %>% 
  rownames_to_column('barcode')
# rename column to reflect data
names(kras.barcode.counts)[2] <- 'cluster'
# tidy 
kras.aale.sc.norm.counts <- 
  kras.aale.sc.norm.counts %>% 
  rownames_to_column('gene') %>% 
  gather(barcode, norm.counts, -gene) 
# pair barcodes and counts
kras.aale.sc.norm.counts <- 
  merge(kras.aale.sc.norm.counts, kras.barcode.counts, by = 'barcode')
```
### Building gene sets from the count matrix
```{r}
# make TE class gene sets
alu.gene.set <- subset(kras.aale.sc.norm.counts, grepl('^Alu', gene))$gene
mer.gene.set <- subset(kras.aale.sc.norm.counts, grepl('^MER', gene))$gene
herv.gene.set <- subset(kras.aale.sc.norm.counts, grepl('^HERV', gene))$gene
ltr.gene.set <- subset(kras.aale.sc.norm.counts, grepl('^LTR', gene))$gene
line.gene.set <- subset(kras.aale.sc.norm.counts, grepl('^L1', gene))$gene
znf.gene.set <- trono.de.krab.znfs$gene
gene.set.list <- lst(alu.gene.set,
                   mer.gene.set,
                   herv.gene.set,
                   ltr.gene.set,
                   line.gene.set,
                   znf.gene.set,
                   liu.isg.signature,
                   kras.isg.signature,
                   rig.i.mda5.induction,
                   rig.i.mda5.neg.regulation,
                   msig.df.h.ifna,
                   msig.df.h.ifng,
                   kras.signal.up)
```
### Calculate gene sets within the suerat counts matrix:
```{r}
## using seurat wrappers for GGPLOT functions to visualize the means of gene sets on a per cell/per cluster basis
# make a copy so we don't overwrite everything above accidentallu
gene.set.kras.aale.sc <- kras.aale.sc  
# add the gene set as an attribute of the single cell object (requires the SCexperiment library)
add.gene.set.to.sc <- function(gene.set, name){
  gene.set <- 
    rownames(gene.set.kras.aale.sc[['RNA']]@data)[rownames(gene.set.kras.aale.sc[['RNA']]@data) %in% gene.set]
  
  # Get mean expression of genes of interest per cell
  mean.exp <- 
    colMeans(x = gene.set.kras.aale.sc[['RNA']]@data[rownames(gene.set.kras.aale.sc[['RNA']]@data) %in% gene.set, ], 
             na.rm = TRUE)
  # Add mean expression values in 'object@meta.data$gene.set.score'
  name <- deparse(substitute(name))
  gene.set.kras.aale.sc@meta.data[, `name`] <- mean.exp 
  return(gene.set.kras.aale.sc)
}

# need to pass the name like this in order to use it as a string...unless I overcomplicated things
gene.set.kras.aale.sc <- add.gene.set.to.sc(liu.isg.signature, `liu.isg.signature`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(kras.isg.signature, `kras.isg.signature`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(msig.df.h.ifna, `msig.df.h.ifna`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(msig.df.h.ifng, `msig.df.h.ifng`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(kras.signal.up, `kras.signal.up`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(rig.i.mda5.induction, `rig.i.mda5.induction`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(alu.gene.set, `alu`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(mer.gene.set, `mer`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(line.gene.set, `line`)
gene.set.kras.aale.sc <- add.gene.set.to.sc(ltr.gene.set, `ltr`)

# spread data into gene columns
norm.counts.plot <- 
  kras.aale.sc.norm.counts %>% 
  group_by(barcode) %>% 
  subset(gene %in% unlist(eval(gene.set.list))) %>%
  spread(gene, norm.counts)

norm.counts.plot %>% 
  gather(gene, count, -cluster, -barcode) %>% 
  group_by(gene) %>% 
  summarise(count = mean(count)) %>% 
  filter(! gene %in% aale.de.seq$gene) %>% 
  top_n(20, count)

# iterate through named list and create a column for each, populating with the avg counts of that set
for(name in names(gene.set.list)){
  norm.counts.plot[, name] <-
  rowMeans(norm.counts.plot[, colnames(norm.counts.plot) %>% 
                            subset(. %in% unlist(eval(gene.set.list[name])))])
}

# make a slimmer df to work with going forward
norm.counts.plot.final <-
  norm.counts.plot %>%
  select(barcode, cluster, unlist(names(gene.set.list)))
```

## ATAC Seq:

### functions:
```{r}

gene.set.hist <- function(file.name){
  input.data <- read_tsv(paste0(atac.dir, 'peak.files/', file.name),
                            col_names = c('pos', 'ctrl', 'ctrl_pos',
                                          'ctrl_neg', 'kras', 'kras_pos', 'kras_neg'),
                            skip = 1)

  input.data %>% 
    select(pos, ctrl, kras) %>% 
    mutate(`relative ATAC enrichment` = log2((kras + 1) / (ctrl + 1))) %>% 
    ggplot(aes(pos/1000, `relative ATAC enrichment`)) +
    geom_line(alpha = 0.2, size = rel(0.75)) + 
    stat_smooth(geom='line', se = F, color = 'black', alpha = 1) + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
    xlab('distance from peaks (kb)') +
    #stat_smooth(geom='line', se = F, color = 'black', size = rel(0.75), alpha = 1) +
    geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) +
    #theme(plot.margin =  margin(5,1,-1.5,0,unit = "mm")) +
    scale_y_continuous(limits = c(-3,3), breaks = seq(-2,2, 1)) +
    geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) %>%
    return()
}

homer.motif.hist <- function(file.name){
  input.data <- read_tsv(paste0(atac.dir, 'peak.files/', file.name),
                            col_names = c('pos', 'motif_sites', 'motif_pos', 'motif_neg', 
                                          'ctrl', 'ctrl_pos', 'ctrl_neg', 'kras', 
                                          'kras_pos', 'kras_neg', "a",'c','g','t'),
                            skip = 1)

  input.data %>% 
    select(pos, motif_sites, ctrl, kras) %>% 
    mutate(`relative ATAC enrichment` = log2((kras + 1) / (ctrl + 1))) %>% 
    ggplot(aes(pos, `relative ATAC enrichment`)) + 
    #ggplot(aes(pos, motif_sites)) + 
    geom_line(alpha = 1, size = rel(1)) +
    xlab('') + 
    stat_smooth(geom='line', se = F, color = 'black', size = rel(1), alpha = 0.4) + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) +
    geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) -> plt.1
  
  input.data %>% 
    select(pos, motif_sites, ctrl, kras) %>% 
    mutate(`relative ATAC enrichment` = log2((kras + 1) / (ctrl + 1))) %>% 
    #ggplot(aes(pos, `relative ATAC enrichment`)) + 
    ggplot(aes(pos, motif_sites)) + 
    geom_line(alpha = 1, size = rel(1)) +
    xlab('distance from kras.isg signature peaks') + 
    ylab('motifs per bp per peak') + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) +
    geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) -> plt.2
  
  input.data %>% 
    select(pos, motif_sites, ctrl, kras) %>% 
    mutate(`relative ATAC enrichment` = log2((kras + 1) / (ctrl + 1))) %>% 
    ggplot(aes(`relative ATAC enrichment`, motif_sites)) + 
    geom_point() + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) +
    geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) -> plt.3
      
    return( (plt.1 / plt.2))
}

```
### annotated peak output (peak files ~ bed 12):
```{r}
read.anno.peaks <- function(path, exp){
  anno.file <- read_tsv(path, skip = 1, 
                              col_names = c('peak','chr', 'start', 'end',
                                            'strand', 'score', 'focus', 
                                            'annotation', 'details', 'tss.dist', 
                                            'prom.id', 'entrez.id', 'unigene', 
                                            'refeseq', 'ensembl', 'gene', 
                                            'alias', 'description','type')) %>% 
  select(-focus) %>% 
  mutate(condition = exp)
  return(anno.file)
  remove(anno.file)
}


kras.anno.peaks <- read.anno.peaks(paste0(homer.dir, 'kras.peaks.anno.bed'), 'KRAS')
ctrl.anno.peaks <- read.anno.peaks(paste0(homer.dir, 'ctrl.peaks.anno.bed'), 'CTRL')

# combine rowwise into a comprehensive data set
aale.anno.peaks <- 
  rbind(ctrl.anno.peaks,
        kras.anno.peaks)
```
### collapsed and merged peaks for comparison:
```{r}
collapse.peaks <- read_tsv(paste0(atac.dir, 'peak.files/aale.collapse.merge.sort.peak.bed'), 
                           col_names = c('chr', 'start', 'end', 
                                         'name', 'fc', 'pval', 
                                         'qval', 'rel.summit')) %>%
  separate_rows(c(name, fc, pval, qval, rel.summit), sep = ',') %>% 
  separate(name, sep = '_', into = c('condition', 'peak', 'number')) %>% 
  unite(col = 'peak', c(peak, number), sep = '_') %>% 
  mutate(start = start + 1) %>% 
  mutate(fc = as.numeric(fc),
         pval = as.numeric(pval),
         qval = as.numeric(qval),
         rel.summit = as.numeric(rel.summit)) %>% 
  group_by(chr, start, end, condition) %>%
  dplyr::summarize(fc = mean(fc),
            pval = mean(pval),
            qval = mean(qval),
            summit = mean(rel.summit))

## need to generate the comparison sets separately in order to include peak annotations
collapse.peaks %>% 
  select(-pval, -qval, -summit) %>% 
  spread(condition, fc) %>% 
  mutate(KRAS = replace_na(KRAS, 0),
         CTRL = replace_na(CTRL, 0),
         KRAS = KRAS + 0.001,
         CTRL = CTRL + 0.001) %>% 
  ungroup() %>% 
  mutate(log2FC = log2(KRAS/CTRL),
         start = start + 1) %>% # gotta adjust the starts to match the bedtools stuff to homer
  merge(kras.anno.peaks, by = c('chr', 'start', 'end')) -> kras.anno.collapsed.peaks

collapse.peaks %>% 
  select(-pval, -qval, -summit) %>% 
  spread(condition, fc) %>% 
  mutate(KRAS = replace_na(KRAS, 0),
         CTRL = replace_na(CTRL, 0),
         KRAS = KRAS + 0.001,
         CTRL = CTRL + 0.001) %>% 
  ungroup() %>% 
  mutate(log2FC = log2(KRAS/CTRL),
         start = start + 1) %>% # gotta adjust the starts to match the bedtools stuff to homer
  merge(ctrl.anno.peaks, by = c('chr', 'start', 'end')) -> ctrl.anno.collapsed.peaks

## and then bind into the reference data set
aale.anno.collapsed.peaks <- 
  rbind(kras.anno.collapsed.peaks,
        ctrl.anno.collapsed.peaks) %>% 
  unique() %>% 
  separate(annotation, sep = '[(]', into = c('anno', 'explained')) %>% 
  separate(explained, sep = '[)]', into = c('motif.type', NA))

# aale.anno.collapsed.peaks %>% 
#   filter(gene %in% filter(trono.de.krab.znfs, log2fc < 2)$gene) %>%
#   select(peak, chr, start, end, strand, score, gene, condition, anno) %>% 
#   write_tsv(paste0(atac.dir, 'peak.files/total.znf.dn.peaks.file'))
# 
# ctrl.anno.collapsed.peaks %>% 
#   unique() %>% 
#   separate(annotation, sep = '[(]', into = c('anno', 'explained')) %>% 
#   separate(explained, sep = '[)]', into = c('motif.type', NA)) %>% 
#   filter(gene %in% filter(trono.de.krab.znfs, log2fc < -2)$gene) %>%
#   select(peak, chr, start, end, strand, score, gene, condition, anno) #%>% 
#   # write_tsv(paste0(atac.dir, 'peak.files/total.znf.dn.peaks.file'))


```
### Homer merged/unique peaks mergePeaks:
```{r}
kras.unique.peaks <- read_tsv(paste0(atac.dir, 'peak.files/kras.unique.peaks'),
                              col_names = c('name', 'chr', 'start', 'end', 
                                            'strand', 'stat', 'file', 
                                            'total_subpeaks', 'other', 'peak'),
                              skip = 1) %>% 
  separate_rows(c(name, chr, start, end, strand, stat, file, other, peak), sep = ',') %>%
  merge(kras.anno.peaks, by = c('peak','chr', 'start', 'end')) %>% 
  separate(annotation, sep = '[(]', into = c('anno', 'explained')) %>% 
  select(-explained, -other)

ctrl.unique.peaks <- read_tsv(paste0(atac.dir, 'peak.files/ctrl.unique.peaks'),
                              col_names = c('name', 'chr', 'start', 'end', 
                                            'strand', 'stat', 'file', 
                                            'total_subpeaks', 'peak', 'other'),
                              skip = 1) %>% 
  separate_rows(c(name, chr, start, end, strand, stat, file, other, peak), sep = ',') %>% 
  merge(ctrl.anno.peaks, by = c('peak','chr', 'start', 'end')) %>% 
  separate(annotation, sep = '[(]', into = c('anno', 'explained')) %>% 
  select(-explained, -other)
  
ctrl.unique.peaks %>% 
  filter(gene %in% filter(trono.de.krab.znfs, log2fc < -1)$gene) %>%
  select(name, chr, start, end, strand=strand.y, stat, file, total_subpeaks, peak) %>% 
  write_tsv(paste0(atac.dir, 'peak.files/ctrl.unique.dn.znf.peaks'))

combined.unique.peaks <- rbind(kras.unique.peaks, ctrl.unique.peaks)
```
### Merged peaks from annoPeaks
```{r}
kras.isg.motifs <- 
  read_tsv(paste0(atac.dir, 'peak.files/kras.isg.motifs.txt')) %>% 
  select_if(function(x) sum(!is.na(x)) >= 4) %>% 
  select(-contains('FOXA1'),
         -contains('FoxL2'),
         -contains('Nearest'),
         -contains('bHLH'),
         -contains('STAT6'),
         -contains('Nr5a2'),
         -contains('Pit1'),
         -contains('AP-2'),
         -`Not Used`, -`Gene Description`, -`Gene Alias`) %>%
  rename('Gene Type' = 'biotype') %>%
  dplyr::rename_at(.vars = vars(contains('(')),
            .funs = funs(vapply(str_split(., pattern = '[(]', n = 3), '[', '', 1))) %>%
  janitor::clean_names() %>% 
  select(-starts_with('n_'),
         -starts_with('b_'),
         -starts_with('ro_'),
         -starts_with('c_'),
         -starts_with('ar_')) %>% 
  rename('kras_label_narrow_peak_tag_tag_count_in_1250_bp' = 'kras',
         'ctrl_label_narrow_peak_tag_tag_count_in_1250_bp' = 'ctrl') %>% 
  gather(motif, detail, -peak_id:-kras)%>% 
  separate_rows(peak_id:detail, sep = '[)],') %>% 
  filter(!is.na(detail)) %>% 
  separate(detail, sep = '[(]', into = c('tss.dist', 'sequence')) %>% 
  separate(sequence, sep = ',', into = c('sequence', 'motif.strand', 'conservation')) %>% 
  filter(strand == motif.strand) %>% 
  mutate(motif = toupper(motif),
         motif = ifelse(motif == 'ISRE', 'IRF9', motif),
         motif = ifelse(motif == 'AP_1', 'FOS', motif),
         kras = as.numeric(kras), ctrl = as.numeric(ctrl), tss.dist = as.numeric(tss.dist)) %>% 
  select(-conservation) %>% 
  filter(motif %in% aale.de.seq$gene) %>% 
  merge(aale.de.seq, by.x = 'motif', by.y = 'gene')
```



# Figures

## Figure 1:

### B: GSEA of both HA1E and AALE cell types:
```{r}
msig.gs.name.list <- 
            c('DER_IFN_ALPHA_RESPONSE_UP',
             'DER_IFN_GAMMA_RESPONSE_UP',
             'REACTOME_RIG_I_MDA5_MEDIATED_INDUCTION_OF_IFN_ALPHA_BETA_PATHWAYS',
             'DER_IFN_BETA_RESPONSE_UP',
             'HALLMARK_INTERFERON_GAMMA_RESPONSE',
             'HALLMARK_INTERFERON_ALPHA_RESPONSE',
             'KRAS_ISG_SIGNATURE',
             'LIU_ISG_SIGNATURE',
             'HALLMARK_KRAS_SIGNALING_UP')

fgsea.kras.isg.signature <- 
  tibble(gene_symbol = kras.isg.signature,
         gs_name = 'kras_isg_signature')

fgsea.liu.isg.signature <- 
  tibble(gene_symbol = liu.isg.signature,
         gs_name = 'liu_isg_signature')

fgsea.rig.i.mda5.induction <- 
  msigdbr(species = 'Homo sapiens') %>% 
  subset(grepl('REACTOME_DDX58', gs_name)) %>% 
  mutate(gs_name = 'rig_i_mda5_induction') %>% 
  select(gs_name, gene_symbol)

fgsea.rig.i.mda5.neg.regulation <- 
  msigdbr(species = 'Homo sapiens') %>% 
  subset(grepl('OF_DDX58', gs_name)) %>% 
  mutate(gs_name = 'rig_i_mda5_neg_regulation') %>% 
  select(gs_name, gene_symbol)

fgsea.msig.df.h.ifng <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  subset(grepl('_GAMMA_RESPONSE', gs_name)) %>% 
  mutate(gs_name = 'hallmark_ifng_response') %>% 
  select(gs_name, gene_symbol)

fgsea.msig.df.h.ifna <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  subset(grepl('_ALPHA_RESPONSE', gs_name)) %>% 
  mutate(gs_name = 'hallmark_ifna_response') %>% 
  select(gs_name, gene_symbol)

fgsea.kras.signal.up <- 
  msigdbr(species = 'Homo sapiens') %>% 
  filter(gs_name == 'HALLMARK_KRAS_SIGNALING_UP') %>% 
  mutate(gs_name = 'hallmark_kras_signaling_up') %>% 
  select(gs_name, gene_symbol)

msig.df <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  select(gene_symbol, gs_name) %>% 
  filter(! gs_name %in% msig.gs.name.list) %>% 
  rbind(fgsea.msig.df.h.ifng) %>% 
  rbind(fgsea.msig.df.h.ifna) %>% 
  rbind(fgsea.kras.isg.signature) %>% 
  rbind(fgsea.liu.isg.signature) %>% 
  rbind(fgsea.kras.signal.up) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

make.fgsea.ranks <- function(de.seq){
  
  # build a ranked gene list in the appropriate format (named list)
  # using shrunken log2 Fold Change values from DESeqII
  de.seq <- 
    de.seq %>%
    mutate(rank = as.double(log2fc),
           Gene = as.character(gene)) %>% 
    select(Gene, rank) %>%
    mutate(rank = scale(rank)) 
  
  de.seq.rnk <- 
    de.seq$rank
  de.seq.rnk <- 
    setNames(de.seq$rank, de.seq$Gene)
  # run the GSEA implementation on the hallmark sets supplemented with custom sets
  print('fgsea')
  de.seq.fgsea.res <- 
    fgsea(pathways = msig.df, 
          stats = de.seq.rnk, 
          nperm = 100000)
  
  print('padj filter and clean names')
  # convert to a tidy format, clean the names for display
  de.seq.fgsea.df <- 
    as_tibble(de.seq.fgsea.res) %>% 
    filter(padj < 0.01) %>% 
    mutate(pathway = gsub('_', ' ', pathway))
  
  return(de.seq.fgsea.df)
  # clean tmp objects
  rm(de.seq, de.seq.rnk, de.seq.fgsea.res, de.seq.fgsea.df)
}

aale.rnk.df <- make.fgsea.ranks(aale.de.seq %>% filter(padj > 0))
ha1e.rnk.df <- make.fgsea.ranks(ha1em.de.seq %>%  filter(padj > 0))
  
aale.rnk.df %>% 
  mutate(condition = 'aale') %>% 
  filter(pathway %in% ha1e.rnk.df$pathway) %>% 
  rbind(ha1e.rnk.df %>% 
          filter(pathway %in% aale.rnk.df$pathway) %>% 
          mutate(condition = 'ha1e')) %>% 
  ggplot(aes(as.factor(reorder(pathway, NES)), NES, fill = condition)) + 
  geom_bar(stat = 'identity', position = 'dodge', width = rel(0.5)) +
  xlab('') +
  scale_fill_manual('cell line', values = c(viridis(3)[1], viridis(3)[2])) +
  geom_hline(yintercept = seq(-3,3), alpha = 1, size = 0.5, color = 'white') +
  geom_vline(xintercept = 4.5, alpha = 0.3, size = 0.25, color = c('black')) +
  geom_hline(yintercept = 0, alpha = 0.3, size = 0.5, color = 'black') +
  theme(legend.position = 'top',
        axis.text.x = element_text(angle = 40, hjust = 1))

if(!file.exists(figure.dir(1, 'gsea.bar.2.5x3.pdf'))){
  ggsave(figure.dir(1, 'gsea.bar.2.5x3.pdf'), height = 2.5, width = 3, units = 'in')
}
```
### C: Venn Diagram of ISG gene sets:
```{r}

tmp.df <- as.data.frame(c(kras.isg.signature, msig.df.h.ifna, msig.df.h.ifng)) %>% 
  select(gene = contains('.'))

liu.isg.signature %>% as.data.frame() %>% select(gene = contains('.')) %>% 
  filter(! gene %in% tmp.df$gene)
  
  
  filter(gene %in% select(liu.isg.signature %>% 
           as.data.frame(), gene = contains('.'))$gene)


temp.vd <- venn.diagram(
  x = list(
    kras.isg.signature %>% unlist(),
    liu.isg.signature %>% unlist(),
    msig.df.h.ifna %>% unlist(),
    msig.df.h.ifng %>% unlist()),
  category.names = c('','','',''),
  filename = NULL,
  output = TRUE,
  col = c(viridis(4)),
  imagetype = 'pdf',
  height = unit(3.5, 'in') ,
  width = unit(3.5, 'in') ,
  resolution = 300,
  compression = "lzw",
  lwd = 1,
  fill = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3), alpha(viridis(4)[3],0.3), alpha(viridis(4)[4],0.3)), 
  fontfamily = "Helvetica",
  scaled = T,
  cat.just = list(c(0.5,0),c(0.5,0),c(0,0),c(0,0)))

grid.draw(temp.vd)

if(!file.exists(figure.dir(1, 'gsea.venn.1.5x1.5.pdf'))){
  library(grDevices)
  pdf(file=figure.dir(1, 'gsea.venn.1.5x1.5.pdf'))
  dev.off()
}
```
### D: ISG volcano Plots:
```{r}
plot.isg.volcano <- function(quant.file) {
  quant.file <- subset(quant.file, gene %in% c(kras.isg.signature,
                                               liu.isg.signature,
                                               msig.df.h.ifna,
                                               msig.df.h.ifng)) %>%
  # for pval == 0, set to the lowest observed pval non 0 pvalue
  mutate(padj = ifelse(padj == 0,
                       8.402919e-301,
                       padj))


  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc)) + 
  theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(1)[1]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3, color = viridis(12, option = 'plasma')[1]) + 
  #scale_x_continuous(limits = c(-6, 6)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_manual(name='', guide = F, values = c(viridis(1),'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 2 ,
                                       as.character(gene), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
} 

isg.aale.de.seq <- plot.isg.volcano(aale.de.seq)
isg.aale.de.seq

if(!file.exists(figure.dir(1, 'isg.volcano.2.5x3.5.pdf'))){
  ggsave(figure.dir(1, 'isg.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

isg.ha1e.de.seq <- plot.isg.volcano(ha1em.de.seq)
isg.ha1e.de.seq

if(!file.exists(figure.dir('supplement', 'ha1em.isg.volcano.2.5x3.5.pdf'))){
  ggsave(figure.dir('supplement', 'ha1em.isg.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

```

## Figure 2:

### A: ISG gene set ATAC histograms
```{r}
liu.isg.hist <- gene.set.hist('total.liu.isg.hist.txt') + 
  ylab('') + 
  xlab('')
kras.isg.hist <- gene.set.hist('total.kras.isg.hist.txt') + 
  ylab('')+ 
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())
msig.ifna.hist <- gene.set.hist('total.msig.ifna.hist.txt') + 
  ylab('') + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())+ 
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())
kras.up.hist <- gene.set.hist('total.kras.up.hist.txt') + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

kras.up.hist <- gene.set.hist('total.kras.up.hist.txt') + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

patchwork::wrap_plots(list(kras.up.hist, msig.ifna.hist, liu.isg.hist, kras.isg.hist), ncol = 2)

if(!file.exists(figure.dir(2, 'stacked.peak.density.3x4.pdf'))){
  ggsave(figure.dir(2, 'stacked.peak.density.3x4.pdf'), height = 3, width = 4, units = 'in')
}
# ggsave('figures/fig.2/stacked.peak.density.3x4.pdf', height = 3, width = 4, units = 'in')
```
### B: KRAS unique ISG ATAC peak table
```{r}
combined.unique.peaks %>% 
  filter(gene %in% kras.isg.signature,
         abs(tss.dist) <= 2500) %>% 
  merge(aale.de.seq, by = 'gene') %>% 
  select(condition, anno, tss.dist, gene, log2fc) %>% 
  tbl_summary(by = anno) %>% 
  bold_labels()
```
### E: ISG TF volcano plot
```{r}
plot.tf.volcano <- function(quant.file) {
  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc,
         color = ifelse(abs(log2fc) > 1 & padj < 0.01, 'not', 'sig'))) +
    theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), alpha = 0.3) + 
  scale_x_continuous(limits = c(-4, 4)) +
  #scale_y_continuous(limits = c(0,-log10(min(aale.de.seq$padj)))) #+
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_manual(name='', guide = F, values = c(viridis(1),'grey')) +
    ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 0.75,
                                       as.character(motif), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
}  

kras.isg.motifs %>% 
  select(motif, log2fc, padj) %>% 
  unique() %>% 
  plot.tf.volcano() -> kras.isg.motifs.volcano
kras.isg.motifs.volcano
# ggsave(plot = kras.isg.motifs.volcano, '../figures/fig.2/updated.kras.isg.volcano.2.3x3.2.pdf', 
#         height = 2.3, width = 3.2, units = 'in')

if(!file.exists(figure.dir(2, 'isg.tf.volcano.2.3x3.2.pdf'))){
  ggsave(figure.dir(2, 'isg.tf.volcano.2.3x3.2.pdf'), height = 3, width = 4, units = 'in')
}

```

## Figure 3:
### A: Single Cell UMAP -- Suerat
```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(kras.aale.sc, 
        reduction = "umap", 
        pt.size = 0.42,  
        cols = c(viridis(12, option = 'plasma')[1],
                 viridis(12, option = 'inferno')[2],
                 viridis(12, option = 'D')[5],
                 viridis(12, option = 'plasma')[7],
                 viridis(12, option = 'inferno')[9],
                 viridis(12, option = 'D')[11],
                 viridis(12, option = 'magma')[6],
                 viridis(12, option = 'magma')[3])) + set_aale_theme()


if(!file.exists(figure.dir(3, 'updated.umap.sc.2.25x3.pdf'))){
  ggsave(figure.dir(3, 'updated.umap.sc.2.25x3.pdf'), height = 2.25, width = 3, units = 'in')
}
```
### B: ISG gene set Single Cell violins:
```{r}
# making the plot objects
make.gene.set.vln <- function(name){
  
  name <- deparse(substitute(name))

  return(VlnPlot(gene.set.kras.aale.sc, 
        features = `name`,pt.size = 0.5, 
        cols = c(viridis(12, option = 'plasma')[1],
                 viridis(12, option = 'inferno')[2],
                 viridis(12, option = 'D')[5],
                 viridis(12, option = 'plasma')[7],
                 viridis(12, option = 'inferno')[9],
                 viridis(12, option = 'D')[11])) + 
  xlab('') + 
    set_aale_theme() +
    theme(plot.margin=margin(-2,0,-1,1,unit = "mm")) +
    theme(legend.position = 'null'))
}


liu.isg.vln <- make.gene.set.vln(`liu.isg.signature`) + 
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank())
kras.isg.vln <- make.gene.set.vln(`kras.isg.signature`)
msig.df.h.ifna.vln <- make.gene.set.vln(`msig.df.h.ifna`)
msig.df.h.ifng.vln <- make.gene.set.vln(`msig.df.h.ifng`)
kras.signal.up.vln <- make.gene.set.vln(`kras.signal.up`) + 
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank())

# pasting them into a single panel
patchwork::wrap_plots(list(kras.signal.up.vln,liu.isg.vln, msig.df.h.ifna.vln,kras.isg.vln))

if(!file.exists(figure.dir(3, 'updated.stacked.vln.2.5x3.5.pdf'))){
  ggsave(figure.dir(3, 'updated.stacked.vln.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}
```
### C: Cluster 4 marker gene distributions:
```{r}
## use seurat functionality to isolate 'marker genes' from each cluster
kras.aale.sc.markers <- 
  FindAllMarkers(kras.aale.sc, 
                 only.pos = F, 
                 min.pct = 0.05, 
                 logfc.threshold = 0.25) %>% 
  filter(p_val_adj < 0.01)
# filter down to the top 5 most significant markers per cluster
kras.aale.sc.markers.heatmap <- 
  kras.aale.sc.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 5, wt = -p_val) 
# filter down to the top 10 most sig in cluster 4
kras.aale.sc.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 10, wt = avg_logFC) %>% 
  filter(cluster %in% c(4)) %>%
  ungroup() %>% 
  select(gene) %>%
  unname() %>% 
  unlist() -> isg.markers

isg.heatmap.norm.counts.plot <-
  kras.aale.sc.norm.counts %>%
  group_by(barcode) %>%
  subset(gene %in% c(isg.markers, 'DDX58', 'IFIH1')) %>%
  spread(gene, norm.counts)

## transform count matrix into a tidy data set to plot cluster v count
as.data.frame(kras.aale.sc@assays$RNA@data) %>% rownames_to_column('gene') %>% 
  filter(gene %in% c(isg.markers, 'IFI27')) %>% 
  gather(barcode, norm.counts, -gene) %>% 
  merge(isg.heatmap.norm.counts.plot %>% select(barcode, cluster), by = 'barcode') %>% 
  distinct() -> dist.sc.plt.df


ggplot(dist.sc.plt.df,
       aes(reorder(gene, -norm.counts), log(norm.counts + 1), color = cluster)) + 
  geom_boxplot(position = position_dodge(width = 1), fill = NA, outlier.colour = NA) + 
  geom_sina(position = position_dodge(width = 1), alpha = 0.4) + xlab('') + 
  scale_color_manual(values = c(viridis(12, option = 'plasma')[1],viridis(12, option = 'inferno')[2],
                       viridis(12, option = 'D')[5],viridis(12, option = 'plasma')[7], 
                       viridis(12, option = 'inferno')[9])) + 
  ylab('log(Seurat normalized counts)')

if(!file.exists(figure.dir(3, 'isg.dist.clusters.2.5x6.7.pdf'))){
  ggsave(figure.dir(3, 'isg.dist.clusters.2.5x6.7.pdf'), height = 2.5, width = 6.7, units = 'in')
}



# # build annotations for pheatmap anno column
# isg.anno.df <- data.frame('cluster' = c(4,2,1,0,3)) %>%
#   mutate(cluster = as.factor(cluster))
# rownames(isg.anno.df) <- isg.anno.df$cluster
# 
# # make color vector to correclty represent colors matching UMAP above
# isg.anno.colors <- list(cluster = c(`0` = viridis(12, option = 'plasma')[1],
#                  `1` = viridis(12, option = 'inferno')[2],
#                  `2` = viridis(12, option = 'D')[5],
#                  `3` = viridis(12, option = 'plasma')[7],
#                  `4` = viridis(12, option = 'inferno')[9]))

# pheatmap
# isg.marker.heatm <- 
#   isg.heatmap.norm.counts.plot %>%
#   gather(gene, count, -barcode, -cluster) %>% 
#   mutate(gene = ifelse(gene == 'DDX58', 'RIG-I', gene),
#          gene = ifelse(gene == 'IFIH1', 'MDA5', gene)) %>% 
#   group_by(cluster, gene) %>% 
#   summarise(mean.count = mean(count)) %>% 
#   ungroup() %>% 
#   mutate(mean.count = log2(mean.count + 1)) %>% 
#   filter(!is.na(mean.count)) %>% 
#   spread(cluster, mean.count) %>% 
#   column_to_rownames('gene') %>% 
#   pheatmap::pheatmap(color = viridis(512), 
#                      border_color = NA, 
#                      treeheight_col = 7, 
#                      treeheight_row = 7,
#                      fontsize = 7,
#                      show_colnames = F,
#                      width = 3.5, height = 2, angle_col = 0,
#                      annotation = isg.anno.df,
#                      annotation_colors = isg.anno.colors) 

# save_pheatmap_pdf(isg.marker.heatm, 'figures/fig.1/updated.isg.marker.heat.pdf', height = 2, width = 3.5)
```

## Figure 4:

### A: TE volcano
```{r}
plot.te.volcano <- function(quant.file) {
  #quant.file <- merge(quant.file, znf.names, by.x = 'Gene', by.y = 'V1')
  ggplot(quant.file, aes(y=-log10(qval),
                     x=log2fc,
         color = clade)) +
    theme(axis.line = element_blank()) + 
  geom_point(size = rel(1), alpha = 0.5) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), alpha = 0.3) + 
  scale_x_continuous(limits = c(-6, 6)) +
  #scale_y_continuous(limits = c(0,-log10(min(aale.de.seq$padj)))) #+
  xlab('log2(Fold Change)') +
  ylab('-log10(qval)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_viridis_d(name = 'clade') + 
  #facet_wrap(~approach, ncol = 1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 2,
                                       as.character(te), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
}

# plot.te.volcano(total.sleuth.out)
# plot.te.volcano(aale.te.de.seq %>% rename(te = 'Gene') %>% merge(te.clades.merge, by = 'te'))
te.aale.sleuth.plt <- 
  plot.te.volcano(total.sleuth.out %>% filter(approach == 'te-only')) + 
  theme(legend.position = 'null')
te.aale.sleuth.plt

if(!file.exists(figure.dir(4, 'te-only.aale.te.volcano.2.5x3.5.pdf'))){
  ggsave(te.aale.sleuth.plt, file = figure.dir(4, 'te-only.aale.te.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

te.ha1em.sleuth.plt <- 
  plot.te.volcano(ha1em.sleuth.out %>% filter(approach == 'te-only')) + 
  theme(legend.position = 'null')
te.ha1em.sleuth.plt

if(!file.exists(figure.dir('supplement', 'te-only.ha1em.te.volcano.2.5x3.5.pdf'))){
  ggsave(te.ha1em.sleuth.plt, file = figure.dir('supplement', 'te-only.ha1em.te.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}
```
### B: TE gene set single cell violin
```{r}
mer.vln <- make.gene.set.vln(`mer`)
line.vln <- make.gene.set.vln(`line`) + theme(axis.text.x = element_blank(),
                                              axis.ticks.x = element_blank())
alu.vln <- make.gene.set.vln(`alu`)+ theme(axis.text.x = element_blank(),
                                              axis.ticks.x = element_blank())
ltr.vln <- make.gene.set.vln(`ltr`)
patchwork::wrap_plots(list(alu.vln, line.vln, mer.vln, ltr.vln)) 

if(!file.exists(figure.dir(4, 'stacked.te.vln.2.5x3.5.pdf'))){
  ggsave(figure.dir(4, 'stacked.te.vln.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}
```
### C: RNA editing in single cells
```{r}
rna.edit.path <- '../data/single.cell.rna.seq/output/subset.rna.editing.out/'

subset.alu.editing.out <- read_csv(paste0(rna.edit.path, 
                                          'sleuth.alu.bed/summary_dir/EditingIndex.csv')) %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  select(sample=Sample, signal=A2GEditingIndex, noise=C2TEditingIndex) %>% 
  filter(grepl('split', sample)) %>% 
  separate(sample, c(NA, 'cluster', NA, NA, NA, 'rep'), '[.]') %>% 
  ggplot(aes(cluster, signal, 
             color = cluster)) +
  geom_boxplot(size = rel(0.5)) + 
  geom_point(size = rel(0.5)) +
  scale_color_manual(values = c(viridis(12, option = 'plasma')[1],
                 viridis(12, option = 'inferno')[2],
                 viridis(12, option = 'D')[5],
                 viridis(12, option = 'plasma')[7],
                 viridis(12, option = 'inferno')[9])) + 
  geom_signif(comparisons =  list(c('4','0'),
                                  c('4', '1'),
                                  c('4', '2'),
                                  c('4', '3')),
              color = 'black',
              test = 't.test', test.args = list( paired=FALSE),
              step_increase = 0.075, 
              size = rel(0.25), inherit.aes = T, textsize = rel(2)) +
  geom_point(aes(cluster, noise), shape = 'diamond', size = rel(1), alpha = 0.5) + 
  xlab('') + ylab('A-to-I Editing') + 
  scale_y_continuous(limits = c(0,7), breaks = c(0,1,3,2,4)) +
  theme(plot.margin=margin(-25,-0.5,-2,2,unit = "mm"),
        legend.box.margin=margin(0,0,0,-6,unit = "mm")) + 
  theme(axis.title.y = element_text(vjust = 1))

if(!file.exists(figure.dir(4, 'subset.sc.editing.3x3.25.pdf'))){
  ggsave(figure.dir(4, 'subset.sc.editing.3x3.25.pdf'), height = 3, width = 3.25, units = 'in')
}
```
### D: Single Cell scatter plots for dsRNA detectors:
```{r}
ggplot(norm.counts.plot %>% filter(cluster %in% c(2,4)), 
       aes(rig.i.mda5.induction, alu.gene.set, color = cluster)) + 
  geom_point(size = 0.42) +
  geom_rug(size = 0.42) +
  # scale_color_manual(values = c(viridis(12, option = 'D')[5], viridis(12, option = 'inferno')[9])) + 
  scale_color_manual(values = c(viridis(12, option = 'plasma')[1],viridis(12, option = 'inferno')[2],
                     viridis(12, option = 'D')[5],viridis(12, option = 'plasma')[7], 
                     viridis(12, option = 'inferno')[9])) + 
  # ylab('KRAS ISG') +
  # xlab('RIG I / MDA5 INDUCTION') + 
  facet_row(~cluster) + 
  geom_smooth(method = 'lm', se = F, size = 0.42) + 
  ggpubr::stat_cor(show.legend = F, inherit.aes = T, size = rel(2), 
                   label.x.npc = 'left', label.y.npc = 1) +
  set_aale_theme() + 
  theme(plot.margin=margin(0,0,1,1,unit = "mm")) + 
  theme(strip.text = element_blank())


if(!file.exists(figure.dir(4, 'rig.i.v.kras.isg.scatter.label.1.25x3.5.pdf'))){
  ggsave(figure.dir(4, 'rig.i.v.kras.isg.scatter.label.1.25x3.5.pdf'), height = 1.25, width = 3.5, units = 'in')
}

```
### E: RIG-I & MDA5 distributions
```{r}
as.data.frame(kras.aale.sc@assays$RNA@data) %>% rownames_to_column('gene') %>% 
  filter(gene %in% c('DDX58', 'IFIH1')) %>% 
  gather(barcode, norm.counts, -gene) %>% 
  merge(isg.heatmap.norm.counts.plot %>% select(barcode, cluster), by = 'barcode') %>% 
  distinct() %>% 
  mutate(gene = ifelse(gene == 'DDX58', 'RIG-I', gene),
  gene = ifelse(gene == 'IFIH1', 'MDA5', gene)) -> dist.sc.plt.df

ggplot(dist.sc.plt.df,
       aes(reorder(gene, -norm.counts), log(norm.counts + 1), color = cluster)) + 
  geom_boxplot(position = position_dodge(width = 1), fill = NA, outlier.colour = NA) + 
  geom_sina(position = position_dodge(width = 1), alpha = 0.4) + xlab('') + 
  scale_color_manual(values = c(viridis(12, option = 'plasma')[1],viridis(12, option = 'inferno')[2],
                       viridis(12, option = 'D')[5],viridis(12, option = 'plasma')[7], 
                       viridis(12, option = 'inferno')[9])) + 
  ylab('log(Seurat normalized counts)')

if(!file.exists(figure.dir(4, 'mda5.dist.clusters.1.25x3.5.pdf'))){
  ggsave(figure.dir(4, 'mda5.dist.clusters.1.25x3.5.pdf'), height = 1.25, width = 3.5, units = 'in')
}
```


## Figure 5:

### A: ZNF volcano
```{r}
plot.znf.volcano <- function(quant.file) {
  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc,
         color = anno )) +
    theme(axis.line = element_blank()) + 
  geom_point(size = rel(1), alpha = 0.5) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), alpha = 0.3) + 
  scale_x_continuous(limits = c(-6, 6)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_viridis_d(name = 'annotation') + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 1.5,
                                       as.character(gene), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
}


znf.aale.de.seq <- plot.znf.volcano(trono.de.krab.znfs) + theme(legend.position = 'null')
znf.aale.de.seq

if(!file.exists(figure.dir(5, 'labelled.aale.znf.volcano.2.5x3.5.no.legend.pdf'))){
  ggsave(znf.aale.de.seq, file = figure.dir(5, 'labelled.aale.znf.volcano.2.5x3.5.no.legend.pdf'), height = 2.5, width = 3.5, units = 'in')
}

znf.ha1e.de.seq <- plot.znf.volcano(ha1em.trono.de.krab.znfs)
znf.ha1e.de.seq

if(!file.exists(figure.dir('supplement', 'labelled.ha1em.znf.volcano.2.5x3.5.no.legend.pdf'))){
  ggsave(znf.ha1e.de.seq, file = figure.dir('supplement', 'labelled.ha1em.znf.volcano.2.5x3.5.no.legend.pdf'), height = 2.5, width = 3.5, units = 'in')
}

```
### B: 
#### ATAC histogram:
```{r}
gene.set.hist('total.unique.dn.znf.peak.hist')

if(!file.exists(figure.dir(5, 'znf.unique.peak.density.1.5x3.25.pdf'))){
  ggsave(figure.dir(5, 'znf.unique.peak.density.1.5x3.25.pdf'), height = 1.5, width = 3.25, units = 'in')
}

```
#### CTRL unique ZNF peaks
```{r}
combined.unique.peaks %>% 
  filter(gene %in% trono.de.krab.znfs$gene,
         abs(tss.dist) <= 2500) %>% 
  merge(aale.de.seq, by = 'gene') %>% 
  select(condition, anno, gene, log2fc) %>% 
  filter(anno %in% c('intron ', 'promoter-TSS '),
         condition == 'CTRL') %>% 
  tbl_summary(by = condition) %>%
  bold_labels()
```


## Supplement:

### lncRNA volcanos:
```{r}
plot.lnc.volcano <- function(quant.file) {
  quant.file <- subset(quant.file, gene %in% c(gen.32.lnc$gene, 'MALAT1')) %>% 
    # for pval == 0, set to the lowest observed pval non 0 pvalue
    mutate(padj = ifelse(padj == 0, 
                         8.402919e-301, 
                         padj))


  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc)) + 
  theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(1)[1]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3, color = viridis(12, option = 'plasma')[1]) + 
  scale_x_continuous(limits = c(-10, 10)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_manual(name='', guide = F, values = c(viridis(1),'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 2 ,
                                       as.character(gene), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
} 

lnc.aale.de.seq <- plot.lnc.volcano(aale.de.seq)
lnc.aale.de.seq
lnc.ha1e.de.seq <- plot.lnc.volcano(ha1em.de.seq)
lnc.ha1e.de.seq


file_test('-f', '../figures/fig.1/updated.lnc.volcano.2.5x3.5.pdf')
file.exists('../figures/fig.1/updated.lnc.volcano.2.5x3.5.pdf')

# ggsave(plot = lnc.aale.de.seq,
#        '../figures/fig.1/updated.lnc.volcano.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')
# ggsave(plot = lnc.ha1e.de.seq,
#        '../figures/fig.1/ha1em.updated.lnc.volcano.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')

```
### lncRNA venn:
```{r}
aale.de.seq %>% subset(gene %in% gen.32.lnc$gene) %>% select(gene) -> lncs.aale.de.seq
ha1em.de.seq %>% subset(gene %in% gen.32.lnc$gene) %>% select(gene) -> lncs.ha1em.de.seq

lncs.aale.de.seq %>% filter(!gene %in% lncs.ha1em.de.seq$gene)
lncs.ha1em.de.seq %>% filter(!gene %in% lncs.aale.de.seq$gene)

# temp.vd <- venn.diagram(
#   x = list(
#     lncs.aale.de.seq$gene %>% unlist(),
#     lncs.ha1em.de.seq$gene %>% unlist()),
#   category.names = c('',''),
#   filename = NULL,
#   output = TRUE,
#   col = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3)),
#   imagetype = 'pdf',
#   height = unit(1.5, 'in') ,
#   width = unit(1.5, 'in') ,
#   resolution = 300,
#   compression = "lzw",
#   lwd = 1,
#   fill = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3)), 
#   fontfamily = "Helvetica",
#   scaled = T)

# library(grDevices)
# 
# pdf(file="../figures/fig.1/lnc.venn.2x2.pdf", width = 2, height = 2)
# grid.draw(temp.vd)
# dev.off()

```
### te venn
```{r}
total.sleuth.out %>% filter(approach == 'te-only') %>% select(te) -> te.aale.de.names
ha1em.sleuth.out %>% filter(approach == 'te-only') %>% select(te) -> te.ha1em.de.names

te.aale.de.names %>% filter(!te %in% te.ha1em.de.names$te) %>% unique()
te.ha1em.de.names %>% filter(!te %in% te.aale.de.names$te) %>% unique()

# temp.vd <- venn.diagram(
#   x = list(
#     te.aale.de.names$te %>% unlist(),
#     te.ha1em.de.names$te %>% unlist()),
#   category.names = c('',''),
#   filename = NULL,
#   output = TRUE,
#   col = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3)),
#   imagetype = 'pdf',
#   height = unit(2, 'in') ,
#   width = unit(2, 'in') ,
#   resolution = 300,
#   compression = "lzw",
#   lwd = 1,
#   fill = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3)), 
#   fontfamily = "Helvetica",
#   scaled = T)
# 
# library(grDevices)
# 
# pdf(file="../figures/fig.1/te.venn.2x2.pdf", width = 2, height = 2)
# grid.draw(temp.vd)
# dev.off()

```
### homer tables
```{r}
getwd()
library(gt)
#kras.unique.peaks.motifs.table <- 
read_tsv(paste0(atac.dir, 'homer.annotations/kras.unique.peaks.motif/knownResults.txt')) %>% 
  select('Motif Name', 'Consensus', 'Log P-value', 'q-value (Benjamini)', '% of Target Sequences with Motif', '% of Background Sequences with Motif') %>% 
  top_n(25, wt = -`Log P-value`) %>% 
  gt() %>% 
  tab_header('Motifs unique to KRAS ATAC library') %>% 
  tab_options(table.font.size = 8) %>% 
  print()

read_tsv(paste0(atac.dir, 'homer.annotations/total.znf.way.dn.peaks.motif/knownResults.txt')) %>% 
  select('Motif Name', 'Consensus', 'Log P-value', 'q-value (Benjamini)', '% of Target Sequences with Motif', '% of Background Sequences with Motif') %>% 
  top_n(25, wt = -`Log P-value`) %>% 
  gt() %>% 
  tab_header('Motifs enriched in downregulated (log2FC <= -1) ZNF loci') %>% 
  tab_options(table.font.size = 8) %>% 
  print()

```
### biotype distribution
```{r}
gencode.te.locus.aale.norm.counts %>% 
  gather(sample, count, -enst:-biotype) %>% 
  filter(biotype != 'Mt_rRNA') %>% 
  group_by(biotype, sample) %>% 
  summarise(count = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(freq = count / sum(count)) %>% 
  mutate(biotype = ifelse(biotype %in% c('lncRNA', 'protein_coding', 'Alu', 'L1', 'ERV1'),
                          biotype, 'other')) %>% 
  #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
  # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
  ggplot(aes(sample, freq, fill = biotype, color = biotype)) + 
    geom_bar(stat = 'identity') +
    # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #geom_sina(position = position_dodge(width = 1)) +
    scale_fill_viridis_d( direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    scale_x_discrete(expand = c(0, 0.5)) +
    geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    xlab('') + ylab('% of total reads') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))

# ggsave('../figures/fig.3/supplement.intra.biotypes.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')

exo.gencode.te.locus.aale.norm.counts %>% 
  gather(sample, count, -enst:-biotype) %>% 
  filter(biotype != 'Mt_rRNA') %>% 
  group_by(biotype, sample) %>% 
  summarise(count = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(freq = count / sum(count)) %>% 
  mutate(biotype = ifelse(biotype %in% c('lncRNA', 'protein_coding', 'Alu', 'L1', 'ERV1'),
                          biotype, 'other')) %>% 
  #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
  # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
  ggplot(aes(sample, freq, fill = biotype, color = biotype)) + 
    geom_bar(stat = 'identity') +
    # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #geom_sina(position = position_dodge(width = 1)) +
    scale_fill_viridis_d( direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    scale_x_discrete(expand = c(0, 0.5)) +
    geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    xlab('') + ylab('% of total reads') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))

# ggsave('../figures/fig.3/supplement.exo.biotypes.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')


aale.exo.locus.count %>% 
  select(-hg, -rmsk, -clade, -range) %>% 
  gather(sample, count, -te, -family) %>% 
  group_by(family, sample) %>% 
  summarise(count = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(freq = count / sum(count)) %>% 
  mutate(family = ifelse(family %in% c('L2', 'Alu', 'L1', 'ERV1', 'ERVK', 'ERVL'),
                          family, 'other')) %>%
  #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
  # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
  ggplot(aes(sample, freq, fill = family, color = family)) + 
    geom_bar(stat = 'identity') +
    # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #geom_sina(position = position_dodge(width = 1)) +
    scale_fill_viridis_d( direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    scale_x_discrete(expand = c(0, 0.5)) +
    geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    xlab('') + ylab('% of total reads') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))

if(!file.exists(figure.dir('supplement', 'exo.te.breakdown.2.5x3.5.pdf'))){
  ggsave(file = figure.dir('supplement', 'exo.te.breakdown.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

gencode.te.locus.aale.norm.counts %>% 
  select(enst, contains('.')) %>%
  merge(te.families.clades, by.x = 'enst', by.y = 'gene') %>% 
  select(-clade) %>% 
  gather(sample, count, -enst, -family) %>% 
  group_by(family, sample) %>% 
  summarise(count = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(freq = count / sum(count)) %>% 
  mutate(family = ifelse(family %in% c('L2', 'Alu', 'L1', 'ERV1', 'ERVK', 'ERVL'),
                          family, 'other')) %>%
  #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
  # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
  ggplot(aes(sample, freq, fill = family, color = family)) + 
    geom_bar(stat = 'identity') +
    # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #geom_sina(position = position_dodge(width = 1)) +
    scale_fill_viridis_d( direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    scale_x_discrete(expand = c(0, 0.5)) +
    geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    xlab('') + ylab('% of total reads') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))

if(!file.exists(figure.dir('supplement', 'intra.te.breakdown.2.5x3.5.pdf'))){
  ggsave(file = figure.dir('supplement', 'intra.te.breakdown.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}
```

### total volcano Plots:
```{r}
plot.isg.volcano <- function(quant.file) {
  quant.file <- quant.file %>% 
    mutate(padj = ifelse(padj == 0,
                       8.402919e-301,
                       padj))


  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc)) + 
  theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(1)[1]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3, color = viridis(12, option = 'plasma')[1]) + 
  #scale_x_continuous(limits = c(-6, 6)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_manual(name='', guide = F, values = c(viridis(1),'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 5 ,
                                       as.character(gene), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
} 

isg.aale.de.seq <- plot.isg.volcano(aale.de.seq)
isg.aale.de.seq

if(!file.exists(figure.dir(1, 'isg.volcano.2.5x3.5.pdf'))){
  ggsave(figure.dir(1, 'isg.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

isg.ha1e.de.seq <- plot.isg.volcano(ha1em.de.seq)
isg.ha1e.de.seq

if(!file.exists(figure.dir('supplement', 'ha1em.isg.volcano.2.5x3.5.pdf'))){
  ggsave(figure.dir('supplement', 'ha1em.isg.volcano.2.5x3.5.pdf'), height = 2.5, width = 3.5, units = 'in')
}

aale.de.seq %>% 
  merge(ha1em.de.seq, by = 'gene') %>% 
  mutate(padj.x = ifelse(padj.x == 0,
                       8.402919e-301,
                       padj.x),
         padj.y = ifelse(padj.y == 0,
                       8.402919e-301,
                       padj.y)) %>% 
  filter(padj.x <= 0.01 && padj.y <= 0.01) %>% 
  ggplot(aes(log2fc.x, log2fc.y)) + 
  geom_point() + 
  xlim(-10,10) + ylim(-10,10)

temp.vd <- venn.diagram(
  x = list(aale.de.seq %>% mutate(padj = ifelse(padj == 0, 8.402919e-301, padj)) %>% 
             filter(padj <= 0.01) %>% select(gene) %>% unlist() %>% unname(),
           ha1em.de.seq %>% mutate(padj = ifelse(padj == 0, 8.402919e-301, padj)) %>% 
             filter(padj <= 0.01) %>% select(gene) %>% unlist() %>% unname()),
  category.names = c('AALE','HA1E'),
  filename = NULL,
  output = TRUE,
  col = c(viridis(2)),
  imagetype = 'pdf',
  height = unit(1.5, 'in') ,
  width = unit(1.5, 'in') ,
  resolution = 300,
  compression = "lzw",
  lwd = 1,
  fill = c(alpha(viridis(4)[1],0.3), alpha(viridis(4)[2],0.3)), 
  fontfamily = "Helvetica",
  scaled = T,
  cat.just = list(c(0.5,0),c(0.5,0))
)

library(grDevices)
grid.draw(temp.vd)

if(!file.exists(figure.dir('supplement', 'de.venn.1.5x1.5.pdf'))){
  library(grDevices)
  pdf(file=figure.dir('supplement', 'de.venn.1.5x1.5.pdf'))
  dev.off()
}

aale.de.seq %>% mutate(padj = ifelse(padj == 0, 8.402919e-301, padj)) %>% 
             filter(padj <= 0.01) %>% select(gene) -> aale.de.count

ha1em.de.seq %>% mutate(padj = ifelse(padj == 0, 8.402919e-301, padj)) %>% 
             filter(padj <= 0.01) %>% select(gene) %>% nrow()
```

### exoTIC comparisons
```{r}
exotic.gencode.norm.counts <-
  read_csv('../data/bulk.rna.seq/exotic/output/exo.gencode.aale.kras.v.ctrl.de-seq.counts.csv')

exotic.gencode.norm.counts %>% 
  select(X1, contains('exotic')) %>% 
  filter_at(vars(contains('.')), any_vars(. >= 5)) %>% 
  gather(sample, count, -X1) %>% 
  mutate(count = log(count + 1)) %>% 
  spread(sample, count) %>% 
  filter(! X1 %in% gen.32.lnc$gene) %>%
  column_to_rownames('X1') #%>% 
  pheatmap(show_rownames = F, color = viridis(24, direction = -1), height = 3, width = 3, border_color = NA, fontsize = 7) -> exotic.heatmap.rc2

save_pheatmap_pdf(exotic.heatmap.rc2, here('exotic.lnc.heatmap.3x3.pdf'), width = 3, height = 3)


exotic.tx.gencode.norm.counts <-
  read_csv('../data/bulk.rna.seq/exotic/output/exo.tx.gencode.aale.kras.v.ctrl.de-seq.counts.csv') %>% 
  rename(X1 = 'tx') %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

```

#### exoTIC biotype dist
```{r}
exotic.tx.gencode.norm.counts %>% 
  gather(sample, count, -enst:-biotype) %>% 
  filter(grepl('exotic', sample)) %>% 
  filter(biotype != 'Mt_rRNA') %>% 
  group_by(biotype, sample) %>% 
  summarise(count = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(freq = count / sum(count)) %>% 
  mutate(biotype = ifelse(biotype %in% c('lncRNA', 'protein_coding'),
                          biotype, 'other')) %>% 
  #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
  # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
  ggplot(aes(sample, freq, fill = biotype, color = biotype)) + 
    geom_bar(stat = 'identity') +
    # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #geom_sina(position = position_dodge(width = 1)) +
    scale_fill_viridis_d(direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    scale_x_discrete(expand = c(0, 0.5)) +
    geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    xlab('') + ylab('% of total counts') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))

# ggsave('../figures/fig.3/supplement.exotic.biotypes.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')
```
#### exoTIC Alu editing
```{r}
rna.edit.path <- '../data/bulk.rna.seq/exotic/output/exotic.alu.edit.out/'

subset.alu.editing.out <- read_csv(paste0(rna.edit.path, 
                                          'summary_dir/EditingIndex.csv')) %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  select(sample=Sample, signal=A2GEditingIndex, noise=C2TEditingIndex) %>% 
  separate(sample, c(NA, 'condition', 'rep', NA), '[.]') %>% 
  ggplot(aes(condition, signal, color = condition)) +
  geom_boxplot(size = rel(0.5)) + 
  geom_point(size = rel(0.5)) +
  geom_signif(comparisons =  list(c('ctrl','kras')),
              color = 'black',
              test = 't.test', test.args = list( paired=FALSE),
              step_increase = 0.075, 
              size = rel(0.25), inherit.aes = T, textsize = rel(2)) +
  geom_point(aes(condition, noise, color = condition), 
             shape = 'diamond', 
             size = rel(1), alpha = 0.5) + 
  xlab('') + ylab('A-to-I Editing') + 
  scale_y_continuous(limits = c(0,0.75)) + 
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F)

subset.alu.editing.out

# ggsave('../figures/fig.3/subset.exotic.editing.2.5x3.5.pdf',
#        height = 2.5, width = 3.5, units = 'in')
```


## OLD FIG 1:

### A: Total volcano Plots:
```{r}
plot.total.volcano <- function(quant.file) {
  quant.file <- quant.file %>% 
  # for pval == 0, set to the lowest observed pval non 0 pvalue
  mutate(padj = ifelse(padj == 0,
                       8.402919e-301,
                       padj))


  ggplot(quant.file, aes(y=-log10(padj),
                     x=log2fc)) + 
  theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(3)[2]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3, color = viridis(3)[2]) + 
  #scale_x_continuous(limits = c(-6, 6)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_manual(name='', guide = F, values = c(viridis(1),'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(-log10(padj)) > 100 | gene == 'KRAS',
                                       as.character(gene), '')),
                         box.padding = unit('0.5', 'lines'),
                         segment.color = 'gray',
                         segment.alpha = 0.2,
                         color = 'black',
                         ylim = c(0.5, Inf), size = rel(2.5))
} 

tot.aale.de.seq <- plot.total.volcano(aale.de.seq)
tot.aale.de.seq

tot.ha1e.de.seq <- plot.total.volcano(ha1em.de.seq)
tot.ha1e.de.seq

aale.de.seq %>% merge(ha1em.de.seq, by = 'gene') %>% 
  ggplot(aes(log2fc.x, log2fc.y, label = gene)) + 
  geom_point(alpha = 0.4) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc.x) >= 1.0 & abs(log2fc.y) > 1.5,
                                     as.character(gene), '')),
                       box.padding = unit('0.5', 'lines'),
                       segment.color = 'gray',
                       segment.alpha = 0.2,
                       color = 'black',
                       ylim = c(-Inf, Inf), size = rel(2.5)) +
  ylim(-6, 10) + 
  xlim(-7.5, 7.5) +
  ylab('HA1E log2fc') + 
  xlab('AALE log2fc') + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) -> aale.v.ha1e.log2fc.scatter

# ggsave(plot = tot.aale.de.seq,
#        paste0(figure.dir.1, '/fig.1/aale.tot.volcano.2.75x3.3.pdf'),
#        height = 2.75, width = 3.3, units = 'in')

if(!file.exists(figure.dir('supplement', 'aale.tot.volcano.2.75x3.3.pdf'))){
  ggsave(tot.aale.de.seq, file = figure.dir('supplement', 'aale.tot.volcano.2.75x3.3.pdf'), height = 2.75, width = 3.3, units = 'in')
}

if(!file.exists(figure.dir('supplement', 'ha1e.tot.volcano.2.75x3.3.pdf'))){
  ggsave(tot.ha1e.de.seq, file = figure.dir('supplement', 'ha1e.tot.volcano.2.75x3.3.pdf'), height = 2.75, width = 3.3, units = 'in')
}

# ggsave(plot = tot.ha1e.de.seq,
#        paste0(figure.dir.1, '/fig.1/ha1em.tot.volcano.2.7x3.5.pdf'),
#        height = 2.75, width = 3.3, units = 'in')

if(!file.exists(figure.dir('supplement', 'ha1e.aale.scatter.2x6.7.pdf'))){
  ggsave(aale.v.ha1e.log2fc.scatter, file = figure.dir('supplement', 'ha1e.aale.scatter.2x6.7.pdf'), height = 2, width = 6.7, units = 'in')
}

```
### B: GSEA of both HA1E and AALE cell types:
```{r}
msig.df <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  select(gene_symbol, gs_name) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

make.fgsea.ranks <- function(de.seq){
  
  # build a ranked gene list in the appropriate format (named list)
  # using shrunken log2 Fold Change values from DESeqII
  de.seq <- 
    de.seq %>%
    mutate(rank = as.double(log2fc),
           Gene = as.character(gene)) %>% 
    select(Gene, rank) %>%
    mutate(rank = scale(rank)) 
  
  de.seq.rnk <- 
    de.seq$rank
  de.seq.rnk <- 
    setNames(de.seq$rank, de.seq$Gene)
  # run the GSEA implementation on the hallmark sets supplemented with custom sets
  print('fgsea')
  de.seq.fgsea.res <- 
    fgsea(pathways = msig.df, 
          stats = de.seq.rnk, 
          nperm = 100000)
  
  print('padj filter and clean names')
  # convert to a tidy format, clean the names for display
  de.seq.fgsea.df <- 
    as_tibble(de.seq.fgsea.res) %>% 
    # filter(padj < 0.05) %>% 
    mutate(pathway = gsub('_', ' ', pathway))
  
  return(de.seq.fgsea.df)
  # clean tmp objects
  rm(de.seq, de.seq.rnk, de.seq.fgsea.res, de.seq.fgsea.df)
}

aale.rnk.df <- make.fgsea.ranks(aale.de.seq %>% mutate(padj = ifelse(padj == 0,8.402919e-301,padj)) %>% filter(padj < 0.01))
ha1e.rnk.df <- make.fgsea.ranks(ha1em.de.seq %>% mutate(padj = ifelse(padj == 0,8.402919e-301,padj)) %>% filter(padj < 0.01))
  

ggplot(aale.rnk.df, aes(NES, reorder(pathway, NES))) + 
  geom_col(fill= viridis(1)) + ylab('') + xlim(-3,3)
           
# ggsave(paste0(figure.dir.1, '/fig.1/aale.h.gsea.2.75x3.3.pdf'),
#        height = 2.75, width = 3.3, units = 'in')

ggplot(ha1e.rnk.df, aes(NES, reorder(pathway, NES))) + 
  geom_col(fill= viridis(3)[2]) + ylab('') + xlim(-3,3)

# ggsave(paste0(figure.dir.1, '/fig.1/ha1em.h.gsea.2.75x3.3.pdf'),
#        height = 2.75, width = 3.3, units = 'in')
```


## Figure 4:

### A: nanosight
```{r}
## dilution factor == 200
kras.sight <- read_csv('../data/other/wetlab/kras.nanosight.data.csv', skip = 2,
                       col_names = c('size', 'r1', 'r2', 'r3', NA, NA, NA, NA)) %>% 
  mutate(condition = 'kras')
ctrl.sight <- read_csv('../data/other/wetlab/ctrl.nanosight.data.csv', skip = 2,
                       col_names = c('size', 'r1', 'r2', 'r3', NA, NA, NA, NA)) %>% 
  mutate(condition = 'ctrl')

sight <- rbind(kras.sight, ctrl.sight)

nanosight.size.dist.plt <- 
  sight %>% 
  select(!contains('X')) %>% 
  gather(read, value, -size, -condition) %>% 
  group_by(size, condition) %>% 
  mutate(mean = (mean(value) * 200)/1e8,
         sd = (sd(value) * 200)/1e8) %>% 
  ggplot(aes(size, y = mean, ymin = mean - sd, ymax = mean + sd, group = condition, fill = condition, color = condition)) + 
  geom_vline(xintercept = seq(0,400,50), linetype = 'dashed', alpha = 0.15) + 
  geom_ribbon(alpha = 0.25, linetype = 'dotted') + 
  geom_line(size = (rel(1)))+
  scale_x_continuous(limits = c(0, 400), breaks = c(0,50,100,150,200,250,300,350,400)) + 
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) + 
  stat_peaks(size = (rel(2))) + 
  ylab('10 million particles per mL') +
  xlab('size (nm)') +
  facet_col(~condition) +
  geom_segment(x = 30, xend = 150, y = 7.75, yend = 7.75)

nanosight.size.dist.plt
# ggsave('../figures/fig.3/nanosight.spline.2.5x3.5.pdf', height = 2.5, width = 3.5, units = 'in')
```
### B: exo v intra DE Volcano
```{r}
plot.exo.volcano <- function(quant.file) {
  quant.file <- quant.file %>% 
  # for pval == 0, set to the lowest observed pval non 0 pvalue
  mutate(padj = ifelse(padj == 0,
                       8.402919e-301,
                       padj))

  quant.file <- quant.file %>% 
    mutate(biotype = ifelse(gene %in% gen.32.lnc$gene, 'lncRNA', 'protein_coding'))
  
  ggplot(quant.file, aes(y=-log10(padj),
                     x=-log2fc, 
         color = biotype)) + 
  theme(axis.line = element_blank()) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(3)[2]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3,color = viridis(3)[2]) + 
  scale_x_continuous(limits = c(-23, 23)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(padj)') +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  # geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  # scale_color_manual(name='', values = c(viridis(1),viridis(3)[2])) +
  facet_col(~biotype, strip.position = 'right') + theme(plot.margin=margin(1,2,2,2,unit = "mm"),
                                                        strip.text = element_text(size = rel(1.5)))
# 
#   ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 16,
#                                        as.character(gene), '')),
#                          box.padding = unit('0.5', 'lines'),
#                          segment.color = 'gray',
#                          segment.alpha = 0.2,
#                          color = 'black',
#                          ylim = c(0.5, Inf), size = rel(2.5))
} 

exo.aale.de.seq.plt <- plot.exo.volcano(exo.vs.intra.kras.de.seq)
exo.aale.de.seq.plt

# ggsave('../figures/fig.5/exo.v.intra.facet.volcano.2.5x3.25.pdf', height = 2.5, width = 3.25, units = 'in')

```
### C: intra v exo log2FC
```{r}
# RidgePlot(kras.aale.sc, features = neuro.diff.markers$gene)
# carla.epi.markers <- c('EPCAM', 'PDPN', 'SFTPB', 'SFTPD', 'FOXJ1', 
#                        'SCGB1A1', 'ETV4', 'SOX9', 'ETV5', 'FOXQ1', 'TM4SF1', 'LEF1',
#                        'COL1A1', 'PECAM1')
# 
# neuro.diff.markers <- read_csv('../data/other/wetlab/ras.markers.csv', col_names = c('gene', 'lineage'))
# 
# yang.markers <- read_tsv('../data/other/wetlab/yang_et_al.ts_score_gencode.v19.tsv') %>% filter(ts_score >= 3)
yang.ts.score.gtex <-
   read_tsv('../data/other/wetlab/yang_et_al.ts_score_gencode.v19.tsv',
            col_names = c('ensg','gene','tissue','tpm','group','ts_score'),
            skip = 1) %>%
   filter(ts_score > 3) %>%
   group_by(gene) %>%
   filter(ts_score == max(ts_score)) %>%
   separate(tissue, sep = ' - ', into = c('tissue', NA)) %>% 
  distinct(gene, tissue)


plot.intra.v.exo.scatter <- function()
  
# exo.aale.de.seq.plt/ 
exo.aale.de.seq %>% 
  merge(aale.de.seq, by = 'gene') %>% 
  mutate(padj.x = ifelse(padj.x == 0,
                     8.402919e-301,
                     padj.x),
         padj.y = ifelse(padj.y == 0,
                     8.402919e-301,
                     padj.y)) %>% 
  # filter(gene %in% yang.markers$Description) %>%
  # merge(yang.ts.score.gtex, by = 'gene') %>% 
  ggplot(aes(log2fc.x, log2fc.y, label = gene)) + 
  geom_smooth(method = 'lm', se = T, size = 0.42, color = viridis(3)[2], alpha = 0.15) +
  geom_point(size = rel(1), alpha = 0.5, color = viridis(3)[2]) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), 
           alpha = 0.3, color = viridis(3)[2]) +
  ggpubr::stat_cor(show.legend = F, inherit.aes = T, size = rel(3),
                   label.x.npc = 'left', label.y.npc = 1) +
  ggrepel::geom_text_repel(aes(label = ifelse(log2fc.x > 0 & log2fc.y > 0 | 
                                                log2fc.x < -2 & log2fc.y < 0 & gene %in% yang.ts.score.gtex$gene | 
                                                log2fc.x < -3 & log2fc.y < 0,
                                     as.character(gene), '')),
                       box.padding = unit('0.5', 'lines'),
                       segment.color = 'gray',
                       segment.alpha = 0.4,
                       color = 'black',
                       ylim = c(-5, Inf), size = rel(2.5)) + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  xlim(-7,7) +
  ylab('intracellular log2 Fold Change') + xlab('exosomal log2 Fold Change') 

exo.aale.de.seq %>% 
  merge(aale.de.seq, by = 'gene', all = T) %>% 
  filter(!is.na(baseMean.x)) %>% 
  mutate(shared = ifelse(is.na(baseMean.y), 'yes', 'no'))


exo.aale.norm.counts %>% 
  merge(exotic.gencode.norm.counts, by = 'X1') %>% 
  select(X1, contains('kras'),  -contains('rneasy')) %>% 
  mutate(exo.mean = rowMeans(.[,c(2,3,4)])) %>% 
  filter(X1 %in% gen.32.lnc$gene) %>% 
  ggplot(aes(exo.mean, exotic.kras)) + 
  geom_point() + xlim(0,5000) +
  ggrepel::geom_text_repel(aes(label = ifelse(exo.mean > 1000 & exotic.kras > 500,
                                     as.character(X1), '')),
                       box.padding = unit('0.5', 'lines'),
                       segment.color = 'gray',
                       segment.alpha = 0.4,
                       color = 'black',
                       ylim = c(-5, Inf), size = rel(2.5))

exotic.gencode.norm.counts %>% 
  filter(X1 %in% gen.32.lnc$gene, X1 %in% exo.vs.intra.kras.de.seq$gene) %>% 
  ggplot(aes(exotic.kras, rneasy.kras, label = X1)) + 
  geom_point() +
  ggrepel::geom_text_repel(
                     box.padding = unit('0.5', 'lines'),
                     segment.color = 'gray',
                     segment.alpha = 0.4,
                     color = 'black',
                     ylim = c(-5, Inf), size = rel(2.5))


# ggsave('../figures/fig.5/intra.exo.scatter.2.5x6.7.pdf', height = 2.5, width = 6.7, units = 'in')


```
### D: TE volcano
```{r}
plot.te.volcano <- function(quant.file) {
  #quant.file <- merge(quant.file, znf.names, by.x = 'Gene', by.y = 'V1')
  ggplot(quant.file, aes(y=-log10(qval),
                     x=log2fc,
         color = clade)) +
    theme(axis.line = element_blank()) + 
  geom_point(size = rel(1), alpha = 0.5) + 
  geom_rug(size = rel(0.5),length = unit(rel(0.1), 'cm'), alpha = 0.3) + 
  #scale_x_continuous(limits = c(-6, 6)) +
  # scale_y_continuous(limits = c(0,-log10(min(aale.de.seq$padj)))) #+
  scale_y_continuous(limits = c(0,10)) +
  xlab('log2(Fold Change)') +
  ylab('-log10(qval)') + 
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = 0, linetype='dotted', alpha = 0.4) +
  geom_hline(yintercept = -log10(0.05), linetype='dotted', alpha = 0.4) +
  geom_vline(xintercept = c(-1,0,1), linetype='dotted', alpha = 0.4) +
  scale_color_viridis_d(name = 'clade') #+ 
  # facet_wrap(~approach, ncol = 1) +
  # ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 2,
  #                                      as.character(te), '')),
  #                        box.padding = unit('0.5', 'lines'),
  #                        segment.color = 'gray',
  #                        segment.alpha = 0.2,
  #                        color = 'black',
  #                        ylim = c(0.5, Inf), size = rel(2.5))
}


te.aale.sleuth.plt <- 
  plot.te.volcano(aale.exo.vs.intra.sleuth.out %>% filter(clade %in% c("SINE", "LINE", "LTR"))) #+ 


aale.exo.vs.intra.sleuth.out.actual %>% 
  filter(qval < 0.01, clade %in% c("SINE", "LINE", "LTR"), ! family %in% c('RTE-BovB', 'RTE-X', '5S-Deu-L2', 'Gypsy', 'CR1', 'LTR')) %>%
  # group_by(te) %>% 
  select(family, exo, intra) %>% 
  gather(condition, count, -family) %>%
  mutate(condition = relevel(as.factor(condition), ref = 'intra')) %>% 
  ggplot(aes(reorder(family, -count), log(count + 1), color = condition, fill = condition)) + 
  geom_boxplot(position = position_dodge(width = 1), fill = NA, outlier.colour = NA) + 
  geom_sina(position = position_dodge(width = 1), alpha = 0.1) + xlab('') +
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2])) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) + 
  stat_compare_means(aes(group = condition), method = 't.test', label = 'p.format', method.args = list(alternative = 'less'))
  # geom_signif()

# te.aale.sleuth.plt
# ggsave('../figures/fig.5/te.locus.count.box.5.15x2.5.pdf',
#        height = 2, width = 5.15, units = 'in')

aale.exo.vs.intra.sleuth.out %>% 
  filter(clade %in% c("SINE", "LINE", "LTR"), ! family %in% c('RTE-BovB', 'RTE-X', '5S-Deu-L2', 'Gypsy', 'CR1', 'LTR')) %>%
  # group_by(te) %>% 
  select(family, contains('kras'), te) %>% 
  filter_at(vars(contains('kras')), any_vars(. > 10)) %>% 
  gather(sample, count, -family, -te) %>%
  mutate(condition = ifelse(grepl('intra', sample), 'intra', 'exo')) %>% 
  mutate(condition = relevel(as.factor(condition), ref = 'intra')) %>%
  group_by(condition, te) %>% 
  mutate(mean.count = mean(count)) %>% 
  distinct(family, sample, mean.count) %>% 
  ggplot(aes(reorder(family, -mean.count), log(mean.count + 1), color = condition, fill = condition)) + 
  geom_boxplot(position = position_dodge(width = 1), fill = NA, outlier.colour = NA) + 
  geom_sina(position = position_dodge(width = 1), alpha = 0.1) + xlab('') +
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2])) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) + 
  stat_compare_means(aes(group = condition), method = 't.test', 
                     label = 'p.format', method.args = list(alternative = 'less'), hide.ns = T, size = rel(2.5))

# ggsave('../figures/fig.5/te.locus.count.box.5.15x2.5.pdf',
#        height = 2, width = 5.15, units = 'in')



aale.exo.vs.intra.sleuth.out %>% 
  filter(clade %in% c("SINE", "LINE", "LTR"), 
         ! family %in% c('RTE-BovB', 'RTE-X', '5S-Deu-L2', 'Gypsy', 'CR1', 'LTR')) %>%
  filter_at(vars(contains('kras')), any_vars(. > 10)) %>% 
  select(contains('kras'), family) %>% 
  gather(sample, count, -family) %>% 
  group_by(sample, family) %>% 
  summarize(mean.count = mean(count)) %>% 
  spread(sample, mean.count) %>% 
  column_to_rownames('family') %>% 
  pheatmap::pheatmap(color = viridis(12), 
                     border_color = 'grey',
                     treeheight_col = 7,
                     treeheight_row = 7,
                     fontsize = 7,
                     width = 5.15, height = 2, angle_col = 0) -> te.locus.heat

# save_pheatmap_pdf(te.locus.heat, '../figures/fig.5/te.marker.heat.pdf', height = 2, width = 5.15)
  
  # theme(legend.position = 'null')

```
### E: exo editing
```{r}
rna.edit.path <- '../data/bulk.rna.seq/aale/output/exo.alu.edit.out/'

subset.alu.editing.out <- read_csv(paste0(rna.edit.path, 
                                          'summary_dir/EditingIndex.csv')) %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>% 
  select(sample=Sample, signal=A2GEditingIndex, noise=C2TEditingIndex) %>% 
  separate(sample, c(NA, 'condition', 'rep', NA), '[.]') %>% 
  ggplot(aes(condition, signal, color = condition)) +
  geom_boxplot(size = rel(0.5)) + 
  geom_point(size = rel(0.5)) +
  geom_signif(comparisons =  list(c('ctrl','kras')),
              color = 'black',
              test = 't.test', test.args = list( paired=FALSE),
              step_increase = 0.075, 
              size = rel(0.25), inherit.aes = T, textsize = rel(2)) +
  geom_point(aes(condition, noise, color = condition), 
             shape = 'diamond', 
             size = rel(1), alpha = 0.5) + 
  xlab('') + ylab('A-to-I Editing') + 
  scale_y_continuous(limits = c(0,0.75)) + 
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F)

subset.alu.editing.out
# 
# ggsave('../figures/fig.3/subset.exo.editing.2.5x1.625.pdf', 
#        height = 2.5, width = 1.625, units = 'in')
```


### I: Cell Titer Glo boxplots:
```{r}
cell.Titer <- read_csv('../data/other/wetlab/cellTiter.csv', skip = 1, 
                       col_names = c('background', 'si', '1','2','3','4','5','6','7','8','9'))
cell.Titer['M1'] <- rowMeans(cell.Titer[, 3:5])
cell.Titer['M2'] <- rowMeans(cell.Titer[, 6:8])
cell.Titer['M3'] <- rowMeans(cell.Titer[, 9:11])

cell.Titer %>% 
  filter(background == 'KRAS_AALE') %>% 
  select(si, M1, M2, M3) %>%
  column_to_rownames('si') %>% 
  t() %>% 
  as.tibble() %>% 
  # rename(siKRAS = 'KRAS',
  #        siCTRL = 'NC.1',
  #        `siRIG-I` = 'RIG-I',
  #        siMDA5 = 'MDA5') %>% 
    rename(KRAS = 'siKRAS',
         NC.1 = 'siCTRL',
         `RIG-I` = 'siRIG-I',
         MDA5 = 'siMDA5') %>% 
  gather(target, Rel.Luminescence) %>%
  mutate(Rel.Luminescence = Rel.Luminescence/1e06) %>% 
  ggplot(aes(target, Rel.Luminescence)) + 
  geom_point(size = rel(0.5), color = viridis(1)) + 
  geom_boxplot(size = rel(0.5), color = viridis(1)) + 
  geom_signif(comparisons =  list(c('siCTRL','siKRAS'),
                                  c('siCTRL', 'siMDA5'),
                                  c('siCTRL', 'siRIG-I')),
                                   test = 't.test', step_increase = 0.075, 
              size = rel(0.25), inherit.aes = T, textsize = rel(2), 
              color = 'black') + 
  scale_x_discrete(limits=c('siCTRL','siKRAS', 'siMDA5', 'siRIG-I')) + 
  xlab('') + scale_y_continuous(limits = c(2,6.5)) + set_aale_theme() + 
  theme(plot.margin =  margin(-2,-.25,-.25,1,unit = "mm")) + 
  ylab('cell viability (ATP)')
                         
# ggsave('../figures/fig.1/updated.ctg.2.5x3.pdf', height = 2.5, width = 3, units = 'in')
```
### C: Alu and Line gene set scatter 
```{r}
ggplot(norm.counts.plot %>% filter(cluster %in% c(2,4)), 
       aes(line.gene.set, alu.gene.set, color = cluster)) + 
  geom_point(size = 0.42) +
  geom_rug() +
  #scale_color_wsj() +
  scale_color_manual(values = c(viridis(12, option = 'D')[5], viridis(12, option = 'inferno')[9])) + 
  # ylab('KRAS ISG') +
  # xlab('RIG I / MDA5 INDUCTION') + 
  facet_row(~cluster, scales = 'free') + 
  geom_smooth(method = 'lm', se = F, size = 0.42) + 
  ggpubr::stat_cor(show.legend = F, inherit.aes = T, size = rel(2), 
                   label.x.npc = 'left', label.y.npc = 1) +
  set_aale_theme() + 
  theme(plot.margin=margin(0,0,1,1,unit = "mm")) + 
  theme(strip.text = element_blank())

  

# ggsave('figures/fig.1/rig.i.v.kras.isg.scatter.label.1.25x3.5.pdf', 
#        height = 1.25, width = 3.5, units = 'in')

```
### D: L1MC4 and AluSx volcano
```{r}
make.gene.vln <- function(name){
  
  name <- deparse(substitute(name))

  return(VlnPlot(gene.set.kras.aale.sc,
        features = c(`name`), pt.size = 0.02, 
        cols = c(viridis(12, option = 'plasma')[1],
                 viridis(12, option = 'inferno')[2],
                 viridis(12, option = 'D')[5],
                 viridis(12, option = 'plasma')[7],
                 viridis(12, option = 'inferno')[9],
                 viridis(12, option = 'D')[11])) + 
  xlab('') + 
  #scale_y_continuous(limits = c(-0.01,1.3)) + 
  theme(legend.position = 'null',
        plot.margin =  margin(-6,0,-3,0,unit = "mm")))

}

l1mc4a.vln <- make.gene.vln(`L1MC4a`) 
alusx.vln <- make.gene.vln(`AluSx`) + theme(axis.title.y = element_blank())
patchwork::wrap_plots(list(l1mc4a.vln, alusx.vln), ncol = 2)
# ggsave('figures/fig.3/alu.line.stacked.vln.1.25x3.5.pdf', height = 1.25, width = 3.5, units = 'in')
```