---
title: "Resubmission of AALE manuscript to Cell Reports"
author: 'Roman E. Reggiardo'
date: '9/22/2021'
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 8, collapse = T)
knitr::opts_chunk$set(message = F)
library(here)
print(here())
```

# setup:
## loading
```{r loading}
suppressPackageStartupMessages({
  library(tidyverse)
  library(here)
  library(ggsci)
  library(ggpubr)
  library(ggsignif)
  library(ggthemes) 
  library(ggrepel)
  library(ggforce)
  library(ggpmisc)
  # library(ggprism)
  library(patchwork)
  library(extrafont)
  library(ggrepel)
  library(pheatmap)
  library(viridis)
  library(patchwork)
  library(fgsea)
  # library(caret)
  library(msigdbr)
  library(matrixStats)
  # library(NMF)
  # library(rsample)
  # library(vip)
  # library(ROCR)
  # library(glmnet)
  # library(mlbench)
  library(tximeta)
  library(SummarizedExperiment)
  library(rjson)
  library(HDF5Array)
  library(DESeq2)
  # library(ggpmisc)
  library(survival)
  library(survminer)
})
suppressMessages(loadfonts())
```
## setting ggplot theme
```{r setting ggplot theme}

base_size = 10

theme_set(theme_foundation(base_size = base_size,
                           base_family = 'Helvetica') + 
            theme(plot.title = element_blank(),
                panel.background = element_rect(colour = NA),
                plot.background = element_rect(colour = NA),
                panel.border = element_rect(colour = NA), 
                axis.line = element_line(), 
                axis.line.x = NULL, 
                axis.line.y = NULL, 
                axis.text = element_text(size = rel(0.95)), 
                axis.text.x = element_text(margin = margin(t = 0.8 * base_size/4)), 
                axis.text.x.top = element_text(margin = margin(b = 0.8 * base_size/4), vjust = 0), 
                axis.text.y = element_text(margin = margin(r = 0.5 * base_size/4), hjust = 1), 
                axis.text.y.right = element_text(margin = margin(l = 0.5 * base_size/4), hjust = 0), 
                axis.ticks = element_line(), 
                axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                axis.ticks.length.y.right = NULL,
                strip.text = element_text(size = rel(0.95), face = 'bold'),
                strip.background = element_blank(),
                legend.key.size= unit(0.08, "in"),
                legend.spacing = unit(0, "in"),
                legend.key = element_rect(colour = NA),
                legend.title = element_text(face="italic"),
                legend.text = element_text(face = 'bold'),
                legend.justification = c("right", "top"),
                legend.box.just = "right",
                legend.margin = margin(6, 6, 6, 6),
                plot.margin=margin(0.04,
                                   0.04,
                                   0.04,
                                   0.04,
                                   unit = "in"),
                # legend.margin = margin(-0.04,
                #                    -0.02,
                #                    0.08,
                #                    0.02,
                #                    unit = "in"),
                legend.box.spacing = unit(-0.02, 'in'),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank()
          ))

theme_set_fun <- function(){
  theme_set(theme_foundation(base_size = base_size,
                             base_family = 'Helvetica') + 
              theme(plot.title = element_blank(),
                  panel.background = element_rect(colour = NA),
                  plot.background = element_rect(colour = NA),
                  panel.border = element_rect(colour = NA), 
                  axis.line = element_line(), 
                  axis.line.x = NULL, 
                  axis.line.y = NULL, 
                  axis.text = element_text(size = rel(0.95)), 
                  axis.text.x = element_text(margin = margin(t = 0.8 * base_size/4)), 
                  axis.text.x.top = element_text(margin = margin(b = 0.8 * base_size/4), vjust = 0), 
                  axis.text.y = element_text(margin = margin(r = 0.5 * base_size/4), hjust = 1), 
                  axis.text.y.right = element_text(margin = margin(l = 0.5 * base_size/4), hjust = 0), 
                  axis.ticks = element_line(), 
                  axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                  axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                  axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                  axis.ticks.length.y.right = NULL,
                  strip.text = element_text(size = 8),
                  strip.background = element_blank(),
                  legend.key.size= unit(0.08, "in"),
                  legend.spacing = unit(0, "in"),
                  legend.key = element_rect(colour = NA),
                  legend.title = element_text(face="italic"),
                  legend.text = element_text(face = 'bold'),
                  legend.justification = c("right", "top"),
                  legend.box.just = "right",
                  legend.margin = margin(6, 6, 6, 6),
                  plot.margin=margin(0.04,
                                     0.04,
                                     0.04,
                                     0.04,
                                     unit = "in"),
                  # legend.margin = margin(-0.04,
                  #                    -0.02,
                  #                    0.08,
                  #                    0.02,
                  #                    unit = "in"),
                  legend.box.spacing = unit(-0.02, 'in'),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank()
            ))
}

```
## figure dirs, and other globally relevant constants
```{r system I/O constants}
figure.dir <-  here('figures')
data.dir <-  here('data')
output.data.dir <- here('output.data')
reference.dir <- here('reference')
job.scripts.dir <- here('R/job.scripts')

chr.order <- c('chr1', 'chr2', 'chr3', 
               'chr4', 'chr5', 'chr6', 
               'chr7', 'chr8', 'chr9',
               'chr10', 'chr11', 'chr12', 
               'chr13', 'chr14', 'chr15', 
               'chr16', 'chr17',
               'chr18', 'chr19', 'chr20', 
               'chr21', 'chr22', 'chrX', 'chrY')
```
## saving panels
```{r saving panel}
panel.figure.width = 85
panel.figure.width.medium = 114
panel.figure.width.double = 174
panel.figure.height = 279.4 
# panel.figure.width.in = 3.4252
# panel.figure.height = 22.5
# panel.figure.height.in = 8.858268

panel.save <- function(figure, name, width = 'single', height = 'single', plot.in = last_plot()){
  
  if (width == 'single'){
    panel.width = panel.figure.width
  } else if (width == 'medium'){
    panel.width = panel.figure.width.medium
  } else{
    panel.width = panel.figure.width.double
  }
  
  if (height == 'single'){
    panel.height = panel.figure.height/3
  } else if (height == 'double'){
    panel.height = (panel.figure.height/3)*1.5
  } else if (height == 'medium'){
    panel.height = (panel.figure.height/3)*2
  } else{
    panel.height = panel.figure.height
  }
  
  if(!dir.exists(paste0(figure.dir, '/','fig.',figure))){
    
    dir.create(paste0(figure.dir, '/','fig.',figure))
    
  }
  
  ggsave(paste0(figure.dir,
                '/',
                'fig.',
                figure,
                '/', 
                name, 
                '.',
                panel.width, 
                'x', 
                panel.height, 
                '.pdf'),
       width = panel.width, 
       height = panel.height, 
       units = 'mm', plot = plot.in)
}

save_pheatmap_pdf <- function(x, filename, width=panel.figure.width, height=3) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=unit(width, 'mm'), height=unit(height, 'mm'), family = 'Helvetica')
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}

## https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/
calc_ht_size = function(ht, unit = "mm") {
    pdf(NULL)
    ht = ComplexHeatmap::draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = grid::convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = grid::convertY(h, unit, valueOnly = TRUE)
    dev.off()

    c(w, h)
}
```
## gene sets for filtering, identification
```{r gene set I/O}

if (! file.exists(file.path(output.data.dir, 'meta_data_df.rds'))) {
  
  meta_data <- lst()
  
  meta_data$te.clades <- 
  read_csv(file.path(output.data.dir, 'reference.data','te.clades.csv'))

  meta_data$te.families.clades <- 
    read_csv(file.path(output.data.dir, 'reference.data','te.family.clade.csv'),
             col_names = c('gene', 'family', 'clade')) %>% 
    distinct()
  
  meta_data$trono.krab.znfs <- 
    read_csv(file.path(output.data.dir, 'krab.znf.trono.list.csv'), 
             col_names = c('gene', 'biotype', 'ensg', 'chr', 'start', 'end', 'strand', 'ZNF'))
  
  meta_data$gen.24.names <- 
    read_csv(file.path(output.data.dir, 'reference.data/gen.24.names.csv'),
             col_names = c('enst','ensg','tx','gene','len','biotype')) %>% 
    select(ensg, gene) %>% 
    distinct()
  
  meta_data$gencode.35.tx.to.gene <- 
    read_csv(file.path(reference.dir, 'gencode.v35.tx.to.gene.csv'), 
                                    col_names = c('tx', 'gene')) %>% 
    separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))
  
  meta_data$gencode.35.tx.to.gene %>% 
    dplyr::select(-enst) %>% 
    group_by(biotype) %>% 
    tally() -> gen.35.biotype.count
  
  meta_data$gencode.35.lnc.gene.names <- 
    read_csv(file.path(reference.dir, 'gencode.v35.lncRNA.gene.names.txt'), 
             col_names = 'gene')
  
  meta_data$ucsc.rmsk.tx.to.gene <-  
    read_csv(file.path(reference.dir, 'gencode.v35.ucsc.rmsk.tx.to.gene.csv'), 
                                    col_names = c('enst', 'gene')) %>%
    filter(grepl('^hg38', enst))
  
  ucsc.rmsk.tx.to.gene %>% 
    mutate(tx = enst) %>% 
    separate(tx, sep = '_', into = c(NA, NA, 'gene', 'position', NA, NA, 'strand', NA)) %>% 
    mutate(position = sub('range=', '', position),
           strand = sub('strand=', '', strand)) %>% 
    merge(te.families.clades, by = 'gene') %>% 
    mutate(tmp.gene = gene,
           ensg = NA) %>% 
    dplyr::select(enst, ensg, 'gene' = tmp.gene, 'len' = clade, 'biotype' = family) -> meta_data$ucsc.rmsk.tx.to.gene.fmt
  
  
  meta_data$tot.gencode.35.tx.to.gene <- 
    read_csv(file.path(reference.dir, 'gencode.v35.tx.to.gene.csv'), 
                                    col_names = c('tx', 'gene')) %>% 
    mutate(enst = tx) %>% 
    separate(tx, sep = '\\|', into = c(NA,'ensg',NA,NA,'tx','gene','len','biotype'))
  
  meta_data$total.tx.to.gene <- 
    rbind(tot.gencode.35.tx.to.gene %>% dplyr::select(enst, ensg, gene, len, biotype),
          ucsc.rmsk.tx.to.gene.fmt %>% dplyr::select(enst = enst, ensg, gene, len, biotype))
  
  
  meta_data$panc_methylation_df <- read_tsv(file.path(output.data.dir, 'GSE119548_avg_beta.txt.gz'))
  
  saveRDS(meta_data, file.path(output.data.dir, 'meta_data_df.rds'))
  
} else {
  meta.data <- readRDS(file.path(output.data.dir, 'meta_data_df.rds'))
}


```

```{r ip tables}
# te.locus.count <- 
#   read_csv(file.path(output.data.dir, 'pancreas.plasma.ev.long.RNA.normalized.deseq.tx.counts.csv')) %>%
#   rename('X1' = 'tx')
# 
# read_csv(file.path(output.data.dir, 
#                    'pancreas.plasma.ev.long.RNA.normalized.deseq.gene.diseaseState_PDAC_vs_healthy.lfcShrink.csv')) %>% 
#   drop_na() %>% 
#   rename('X1' = 'gene') %>% 
#   filter(!grepl('^MT', gene),
#          abs(lfcShrinkResults.log2FoldChange) >= 0.5) %>% 
#   merge(te.families.clades, by = 'gene') %>% 
#   mutate(family = ifelse(is.na(family), clade, family), 
#          clade = factor(clade, levels = c('coding-gencode', 'lncRNA', 'DNA', 'LINE', 'LTR', 'SINE'))) %>% 
#   filter(!grepl('\\(', gene)) %>% 
#   mutate(genome = 'hg38') -> panc_v_ctrl_te.ip.parse
# 
# 
# te.locus.count %>% 
#   filter(!grepl('^ENST', tx)) %>% 
#   filter_at(vars(contains('SRR')), any_vars(. > 5)) %>% 
#   separate(tx, '=', into = c('name', 'position')) %>% 
#   separate(name, '_', into = c(NA,NA,'name',NA)) %>%
#   merge(te.families.clades, by.x = 'name', by.y = 'gene') %>% 
#   separate(position, ':', into = c('chr', 'position')) %>% 
#   separate(position, '_', into = c('position', NA)) %>% 
#   separate(position, '-', into = c('start', 'end')) %>% 
#   dplyr::select(gene='name', chr, start, end, clade, contains('SRR')) %>% 
#   filter(gene %in% panc_v_ctrl_te.ip.parse$gene) %>% 
#   gather(patient, count, -chr, -start, -end, -gene, -clade) %>% 
#   group_by(chr, start, end, gene, clade) %>% 
#   summarize(avg_count = mean(count)) %>% 
#   mutate(start = as.numeric(start),
#            chr = factor(chr, levels = chr.order)) %>% 
#   merge(panc_v_ctrl_te.ip.parse %>% dplyr::select(-clade), by = 'gene') -> panc_v_ctrl_te.ip.table
# 
# write_csv(x = panc_v_ctrl_te.ip.table, 
#           file.path(output.data.dir, 'panc_v_healthy.gut.paper.te.ip.table.RENAME.COLUMNS.csv'))
```

# feature engineering:
## FUN: construct.quant.object
```{r}
load.tximeta.object <- function(reference){
  
  print(file.path(output.data.dir, paste0(reference, '_h5_se')))
  
  txi <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_h5_se')))
  gxi <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_gene_h5_se')))
  
  if (reference == 'process.aware.salmon'){
    
    gxi.split <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_gene_split_h5_se')))
    
    return(list('txi' = txi, 'gxi' = gxi, 'gxi.split' = gxi.split))
    
  } else {
    
    return(list('txi' = txi, 'gxi' = gxi))
  }
  
}

subset.tximeta.object <- function(tximeta.list, grep.string){
  
  if (grepl('\\|', grep.string)){
      
    subset.txi <- SummarizedExperiment::subset(tximeta.list$txi, 
                                         select = !grepl(grep.string, colData(tximeta.list$txi)$names))
  
    subset.gxi <- SummarizedExperiment::subset(tximeta.list$gxi, 
                                         select = !grepl(grep.string, colData(tximeta.list$gxi)$names))
  } else{
        
    subset.txi <- SummarizedExperiment::subset(tximeta.list$txi, 
                                         select = grepl(grep.string, colData(tximeta.list$txi)$names))
  
    subset.gxi <- SummarizedExperiment::subset(tximeta.list$gxi, 
                                         select = grepl(grep.string, colData(tximeta.list$gxi)$names))
  }
  
  return(list('txi' = subset.txi,
              'gxi' = subset.gxi))
}
```
### CALL: construct.quant.object
```{r}
salmon.quant <- load.tximeta.object('salmon')

salmon.quant$aale$intra <- subset.tximeta.object(salmon.quant, 'intra.aale')
salmon.quant$hbec$hbec_wt <- subset.tximeta.object(salmon.quant, 'wt2_kras')
salmon.quant$hbec$hbec_v12 <- subset.tximeta.object(salmon.quant, '_v12')
salmon.quant$hbec$lacz_v_wt <- subset.tximeta.object(salmon.quant, '_v12|mut2|aale|ZNF|Empty')
salmon.quant$hbec$lacz_v_v12 <- subset.tximeta.object(salmon.quant, '_wt|mut2|aale|ZNF|Empty')
salmon.quant$a549 <- subset.tximeta.object(salmon.quant, 'intra.a549')
salmon.quant$all_intra <- subset.tximeta.object(salmon.quant, 'rnaEASY|hold')
salmon.quant$aale$exo <- subset.tximeta.object(salmon.quant, 'rnaEASY.aale')
salmon.quant$aale$kras_compare <- subset.tximeta.object(salmon.quant, 'aale.*.kras')
salmon.quant$aale$ctrl_compare <- subset.tximeta.object(salmon.quant, 'aale.*.ctrl')
salmon.quant$aale$platform_compare <- subset.tximeta.object(salmon.quant, 'aale')

colData(salmon.quant$gxi) %>% 
  as.data.frame() %>% 
  remove_rownames() -> salmon.quant$meta.data.df

rowRanges(salmon.quant$gxi) %>% 
  as.data.frame() %>% 
  dplyr::rename('chr' = seqnames) %>% 
  merge(gencode.35.tx.to.gene %>% select(ensg, gene) %>% distinct(), by.x = 'gene_id', by.y = 'ensg') -> 
  salmon.quant$gxi.bed

rowRanges(salmon.quant$txi) %>% 
  as.data.frame() %>% 
  dplyr::rename('chr' = seqnames) %>% 
  merge(gencode.35.tx.to.gene %>% select(enst, gene) %>% distinct(), by.x = 'tx_name', by.y = 'enst') -> 
  salmon.quant$txi.bed

# ------------------------------------------------------------------------------

ucsc.rmsk.salmon.quant <- load.tximeta.object('ucsc.rmsk.salmon')

ucsc.rmsk.salmon.quant$aale$intra <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'intra.aale')
ucsc.rmsk.salmon.quant$hbec$hbec_wt <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'wt2_kras')
ucsc.rmsk.salmon.quant$hbec$hbec_v12 <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, '_v12')
ucsc.rmsk.salmon.quant$hbec$lacz_v_wt <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, '_v12|mut2|aale|ZNF|Empty')
ucsc.rmsk.salmon.quant$hbec$lacz_v_v12 <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, '_wt|mut2|aale|ZNF|Empty')
ucsc.rmsk.salmon.quant$a549 <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'intra.a549')
ucsc.rmsk.salmon.quant$all_intra <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'rnaEASY|hold')
ucsc.rmsk.salmon.quant$aale$exo <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'rnaEASY.aale')
ucsc.rmsk.salmon.quant$aale$kras_compare <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'aale.*.kras')
ucsc.rmsk.salmon.quant$aale$ctrl_compare <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'aale.*.ctrl')
ucsc.rmsk.salmon.quant$aale$platform_compare <- 
  subset.tximeta.object(ucsc.rmsk.salmon.quant, 'aale')

colData(ucsc.rmsk.salmon.quant$gxi) %>% 
  as.data.frame() %>% 
  remove_rownames() -> ucsc.rmsk.salmon.quant$meta.data.df

```

## FUN: run.de.seq
```{r}
run.de.seq <- function(tximeta.object, base_level = 'ctrl'){
  
  tximeta.name <- deparse(substitute(tximeta.object))
  
  print(tximeta.name)
  
  colData(tximeta.object$gxi) %>% 
    as.data.frame() %>% 
    remove_rownames() %>% 
    mutate(condition = relevel(as.factor(condition), base_level)) -> colData.gxi
  
  print(levels(colData.gxi$condition))
  
  print(colData.gxi)

  count.matrix.gxi <- assay(tximeta.object$gxi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('gene') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('gene')
    
  colData(tximeta.object$txi) %>% 
    as.data.frame() %>% 
    remove_rownames() -> colData.txi

  count.matrix.txi <- assay(tximeta.object$txi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('tmp') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('tmp')
  
  if (grepl('compare', tximeta.name)) {
    
    colData.gxi$platform <- relevel(as.factor(colData.gxi$platform), 'intra')
    
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                    colData = colData.gxi, 
                                    design = as.formula(~platform))

    dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi, 
                                      colData = colData.txi, 
                                      design = as.formula(~platform))
    
  } else {
    
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                    colData = colData.gxi, 
                                    design = as.formula(~condition))

    dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi, 
                                      colData = colData.txi, 
                                      design = as.formula(~condition))
    
  }
  
  dds.txi <- estimateSizeFactors(dds.txi)
  
  dds.txi.norm.counts <- as.data.frame(counts(dds.txi, normalized=T))
  
  dds.txi.vst.counts <- as.data.frame(assay(vst(dds.txi, blind = F)))
  
  dds.gxi <- estimateSizeFactors(dds.gxi)
  
  dds.gxi.norm.counts <- as.data.frame(counts(dds.gxi, normalized=T))
  
  dds.gxi.vst.counts <-  as.data.frame(assay(vst(dds.gxi, blind = F)))
  
  if (grepl('all_intra|a549', tximeta.name)){
    print('no DE calc')
      if (grepl('ucsc.rmsk', tximeta.name)){  
    
        print('rmsk')
    
        dds.gxi.norm.counts <- dds.gxi.norm.counts %>% 
          rownames_to_column('gene') 
        
        return(list('counts' = dds.gxi.norm.counts,
                    'txi.counts' = dds.txi.norm.counts,
                    'gxi.vst.counts' = dds.gxi.vst.counts,
                    'col.data' = colData.txi))
        
        quit()
      }
    
      dds.gxi.norm.counts <- 
        dds.gxi.norm.counts %>% 
          rownames_to_column('gene_id') %>% 
          merge(salmon.quant$gxi.bed, by = 'gene_id') %>% 
          distinct()
      
      dds.txi.norm.counts <- 
        dds.gxi.norm.counts %>% 
          rownames_to_column('tx_name') %>% 
          merge(salmon.quant$txi.bed, by = 'tx_name') %>% 
          distinct() 
    
      return(list('counts' = dds.gxi.norm.counts, 
                  'txi.counts' = dds.txi.norm.counts,
                  'gxi.vst.counts' = dds.gxi.vst.counts,
                  'col.data' = colData.txi))
      quit()

  }
  
  filter <- rowSums(counts(dds.gxi, normalized=T) >= 10) >= 2
  
  dds.gxi <- DESeq(dds.gxi[filter,])
  
  print(resultsNames(dds.gxi))
  
  coef = print(tail(resultsNames(dds.gxi), n=1))

  if (grepl('ucsc.rmsk', tximeta.name)){  
    
    print('rmsk')
    
    dds.gxi.lfc <- lfcShrink(dds.gxi, coef = coef) %>% 
      as.data.frame() %>% 
      # filter(padj <= 0.05) %>% 
      rownames_to_column('gene') %>% 
      dplyr::select(gene, baseMean, log2FoldChange, padj)
    
    dds.gxi.norm.counts <- dds.gxi.norm.counts %>% 
      rownames_to_column('gene') 
    
  } else{
    
    dds.gxi.lfc <- lfcShrink(dds.gxi, coef = coef) %>% 
      as.data.frame() %>% 
      # filter(padj <= 0.05) %>% 
      rownames_to_column('gene_id') %>% 
      merge(salmon.quant$gxi.bed, by = 'gene_id')
    
    plotMA(lfcShrink(dds.gxi, coef = coef))
    
    dds.gxi.norm.counts <- 
      dds.gxi.norm.counts %>% 
        rownames_to_column('gene_id') %>% 
        merge(salmon.quant$gxi.bed, by = 'gene_id') %>% 
        distinct()
    
    dds.txi.norm.counts <- 
      dds.gxi.norm.counts %>% 
        rownames_to_column('tx_name') %>% 
        merge(salmon.quant$txi.bed, by = 'tx_name') %>% 
        distinct() 
  }

return(list('de' = dds.gxi.lfc, 
            'counts' = dds.gxi.norm.counts, 
            'txi.counts' = dds.txi.norm.counts, 
            'txi.vst.counts' = dds.txi.vst.counts,
            'gxi.vst.counts' = dds.gxi.vst.counts,
            'col.data' = colData.txi))
  
}

run.de.seq.contrast <- function(tximeta.object, base_level = 'Empty'){
  
  tximeta.name <- deparse(substitute(tximeta.object))
  
  print(tximeta.name)
  
  colData(tximeta.object$gxi) %>% 
    as.data.frame() %>% 
    remove_rownames() %>% 
    mutate(condition = relevel(as.factor(condition), base_level)) -> colData.gxi
  
  print(levels(colData.gxi$condition))
  
  print(colData.gxi)

  count.matrix.gxi <- assay(tximeta.object$gxi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('gene') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('gene')
    
  colData(tximeta.object$txi) %>% 
    as.data.frame() %>% 
    remove_rownames() -> colData.txi

  count.matrix.txi <- assay(tximeta.object$txi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('tmp') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('tmp')
  
  # colData.gxi$platform <- relevel(as.factor(colData.gxi$platform), 'intra')
  
  dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                  colData = colData.gxi, 
                                  design = as.formula(~condition))

  # dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi, 
  #                                   colData = colData.txi, 
  #                                   design = as.formula(~platform))
  
  dds.gxi <- estimateSizeFactors(dds.gxi)
  
  filter <- rowSums(counts(dds.gxi, normalized=T) >= 10) >= 2
  
  dds.gxi <- DESeq(dds.gxi[filter,])
  
  print(resultsNames(dds.gxi))
  
  # results(dds.gxi, contrast =list("conditionkras.platformrnaEASY"))
  # 
  dds.682.gxi.lfc <- lfcShrink(dds.gxi, coef = "condition_ZNF682_vs_Empty") %>%
    as.data.frame() %>%
    # filter(padj <= 0.05) %>%
    rownames_to_column('gene_id') #%>%
    # merge(salmon.quant$gxi.bed, by = 'gene_id')
  
  dds.257.gxi.lfc <- lfcShrink(dds.gxi, coef = "condition_ZNF257_vs_Empty") %>%
    as.data.frame() %>%
    # filter(padj <= 0.05) %>%
    rownames_to_column('gene_id') #%>%
    # merge(salmon.quant$gxi.bed, by = 'gene_id')

  return(list('de_682' = dds.682.gxi.lfc, 
              'de_257' = dds.257.gxi.lfc,
              'col.data' = colData.txi))
  
  
}

run.de.seq.contrast <- function(tximeta.object, base_level = 'ctrl'){
  
  tximeta.name <- deparse(substitute(tximeta.object))
  
  print(tximeta.name)
  
  colData(tximeta.object$gxi) %>% 
    as.data.frame() %>% 
    remove_rownames() %>% 
    mutate(condition = relevel(as.factor(condition), base_level)) -> colData.gxi
  
  print(levels(colData.gxi$condition))
  
  print(colData.gxi)

  count.matrix.gxi <- assay(tximeta.object$gxi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('gene') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('gene')
    
  colData(tximeta.object$txi) %>% 
    as.data.frame() %>% 
    remove_rownames() -> colData.txi

  count.matrix.txi <- assay(tximeta.object$txi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('tmp') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('tmp')
  
  # colData.gxi$platform <- relevel(as.factor(colData.gxi$platform), 'intra')
  
  # dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
  #                                 colData = colData.gxi, 
  #                                 design = as.formula(~condition + 
  #                                                       platform + 
  #                                                       platform:condition))

  dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi,
                                    colData = colData.txi,
                                    design = as.formula(~platform))
  
  dds.gxi <- estimateSizeFactors(dds.gxi)
  
  filter <- rowSums(counts(dds.gxi, normalized=T) >= 10) >= 2
  
  dds.gxi <- DESeq(dds.gxi[filter,])
  
  print(resultsNames(dds.gxi))
  
  # results(dds.gxi, contrast =list("conditionkras.platformrnaEASY"))
  # 
 results(dds.gxi, contrast =list("conditionkras.platformrnaEASY"))

 ctrl.dds.gxi.lfc <- lfcShrink(dds.gxi, coef = "conditionkras.platformrnaEASY") %>%
   as.data.frame() %>%
   # filter(padj <= 0.05) %>%
   rownames_to_column('gene_id') %>%
   merge(salmon.quant$gxi.bed, by = 'gene_id')
 
 ctrl.dds.gxi.lfc <- lfcShrink(dds.gxi, coef = "conditionctrl.platformrnaEASY") %>%
   as.data.frame() %>%
   # filter(padj <= 0.05) %>%
   rownames_to_column('gene_id') %>%
   merge(salmon.quant$gxi.bed, by = 'gene_id')

  return(list('ctrl' = ctrl.dds.gxi.lfc, 
              'kras' = ctrl.dds.gxi.lfc,
              'col.data' = colData.txi))
  
  
}

```
## CALL: run.de.seq
```{r}
salmon.quant$aale$intra$deseq <- run.de.seq(salmon.quant$aale$intra, base_level = 'ctrl')
salmon.quant$aale$exo$deseq <- run.de.seq(salmon.quant$aale$exo, base_level = 'ctrl')
salmon.quant$aale$kras_compare$deseq <- run.de.seq(salmon.quant$aale$kras_compare, base_level = 'kras')
salmon.quant$aale$ctrl_compare$deseq <- run.de.seq(salmon.quant$aale$ctrl_compare, base_level = 'ctrl')
salmon.quant$aale$platform_compare$deseq <- run.de.seq(salmon.quant$aale$platform_compare, base_level = 'ctrl')
salmon.quant$hbec$hbec_wt$deseq <- run.de.seq(salmon.quant$hbec$hbec_wt, base_level = 'wt2_kras_wt')
salmon.quant$hbec$hbec_v12$deseq <- run.de.seq(salmon.quant$hbec$hbec_v12, base_level = 'wt2_kras_v12')
salmon.quant$hbec$lacz_v_wt$deseq <- run.de.seq(salmon.quant$hbec$lacz_v_wt, base_level = 'wt2_lacz')
salmon.quant$hbec$lacz_v_v12$deseq <- run.de.seq(salmon.quant$hbec$lacz_v_v12, base_level = 'wt2_lacz')
salmon.quant$a549$deseq <- run.de.seq(salmon.quant$a549, base_case = 'Empty')
salmon.quant$all_intra$deseq <- run.de.seq(salmon.quant$all_intra)

a549_de <- run.de.seq.contrast(salmon.quant$a549)

a549_te_de <- run.de.seq.contrast(ucsc.rmsk.salmon.quant$a549)

# tmp <- run.de.seq.contrast(salmon.quant$aale$kras_compare$)

View(tmp %>% filter(padj < 0.01))

tmp %>% filter(gene %in% salmon.quant$aale$intra$deseq$de$gene) %>% View()

salmon.quant$aale$exo$deseq$de %>% 
  merge(tmp, by = 'gene') %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, label = gene)) + 
  geom_point()+
  geom_text_repel(aes(label = ifelse(abs(log2FoldChange.x) > 2, gene, '')))


# ------------------------------------------------------------------------------

ucsc.rmsk.salmon.quant$aale$intra$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$aale$intra, base_level = 'ctrl')
ucsc.rmsk.salmon.quant$aale$exo$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$aale$exo, base_level = 'ctrl')
ucsc.rmsk.salmon.quant$aale$kras_compare$deseq <- 
  run.de.seq(ucsc.rmsk.salmon.quant$aale$kras_compare, base_level = 'kras')
ucsc.rmsk.salmon.quant$aale$ctrl_compare$deseq <- 
  run.de.seq(ucsc.rmsk.salmon.quant$aale$ctrl_compare, base_level = 'ctrl')
ucsc.rmsk.salmon.quant$aale$platform_compare$deseq <- 
  run.de.seq(ucsc.rmsk.salmon.quant$aale$platform_compare, base_level = 'ctrl')
# ucsc.rmsk.salmon.quant$hbec$hbec_wt$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$hbec$hbec_wt, base_level = 'wt2_kras_wt')
# ucsc.rmsk.salmon.quant$hbec$hbec_v12$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$hbec$hbec_v12, base_level = 'wt2_kras_v12')
# ucsc.rmsk.salmon.quant$hbec$lacz_v_wt$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$hbec$lacz_v_wt, base_level = 'wt2_lacz')
ucsc.rmsk.salmon.quant$hbec$lacz_v_v12$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$hbec$lacz_v_v12, base_level = 'wt2_lacz')
ucsc.rmsk.salmon.quant$hbec$a549$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$a549, base_case = 'Empty')
ucsc.rmsk.salmon.quant$all_intra$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$all_intra)

saveRDS(ucsc.rmsk.salmon.quant, file = file.path(output.data.dir, 'ucsc.rmsk.salmon.quant_analysis_set.rds'))



```

## FUN: make.fgsea.ranks
```{r}
make.fgsea.ranks <- function(de.seq){
  
  # build a ranked gene list in the appropriate format (named list)
  # using shrunken log2 Fold Change values from DESeqII
  de.seq <- 
    de.seq %>%
    mutate(rank = as.double(log2FoldChange),
           Gene = as.character(gene)) %>% 
    dplyr::select(Gene, rank) %>%
    mutate(rank = scale(rank)) 
  
  de.seq.rnk <- 
    de.seq$rank
  de.seq.rnk <- 
    setNames(de.seq$rank, de.seq$Gene)

  # run the GSEA implementation on the hallmark sets supplemented with custom sets
  print('fgsea')
  de.seq.fgsea.res <- 
    fgsea(pathways = msig.df, 
          stats = de.seq.rnk, eps = 0.0)
  
  print('padj filter and clean names')
  # convert to a tidy format, clean the names for display
  de.seq.fgsea.df <-
    as_tibble(de.seq.fgsea.res) %>%
    filter(padj < 0.05) %>%
    mutate(pathway = gsub('_', ' ', pathway))

  
  return(de.seq.fgsea.df)
  
}


```
### CALL: make.fgsea.ranks
```{r}
msig.df <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  dplyr::select(gene_symbol, gs_name) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

msig.df$HALLMARK_KRAS_G12D_ZNF_DN <- salmon.quant$aale$intra$deseq$de %>% 
  filter(gene %in% filter(meta.data$trono.krab.znfs, ZNF == 'KRAB-ZNFs')$gene,
         log2FoldChange <= -1) %>% 
  dplyr::select(gene) %>% unlist() %>% unname()

salmon.quant$aale$intra$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$aale$intra$deseq$de)
salmon.quant$aale$exo$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$aale$exo$deseq$de)
salmon.quant$aale$kras_compare$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$aale$kras_compare$deseq$de)
salmon.quant$aale$ctrl_compare$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$aale$ctrl_compare$deseq$de)
salmon.quant$aale$platform_compare$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$aale$platform_compare$deseq$de)
salmon.quant$hbec$hbec_wt$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$hbec$hbec_wt$deseq$de)
salmon.quant$hbec$hbec_v12$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$hbec$hbec_v12$deseq$de)
salmon.quant$hbec$lacz_v_wt$deseq$H.fgsea <-make.fgsea.ranks(salmon.quant$hbec$lacz_v_wt$deseq$de)
salmon.quant$hbec$lacz_v_v12$deseq$H.fgsea <- make.fgsea.ranks(salmon.quant$hbec$lacz_v_v12$deseq$de)

# saveRDS(salmon.quant, file = file.path(output.data.dir, 'salmon.quant_analysis_set.rds'))

msig.df <- 
  msigdbr(species = 'Homo sapiens') %>% 
  dplyr::select(gene_symbol, gs_name) %>% 
  filter(!grepl('HALLMARK', gs_name)) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

salmon.quant$aale$intra$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$aale$intra$deseq$de)
salmon.quant$aale$exo$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$aale$exo$deseq$de)
salmon.quant$aale$kras_compare$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$aale$kras_compare$deseq$de)
salmon.quant$aale$ctrl_compare$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$aale$ctrl_compare$deseq$de)
salmon.quant$aale$platform_compare$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$aale$platform_compare$deseq$de)
salmon.quant$hbec$hbec_wt$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$hbec$hbec_wt$deseq$de)
salmon.quant$hbec$hbec_v12$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$hbec$hbec_v12$deseq$de)
salmon.quant$hbec$lacz_v_wt$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$hbec$lacz_v_wt$deseq$de)
salmon.quant$hbec$lacz_v_v12$deseq$A.fgsea <- make.fgsea.ranks(salmon.quant$hbec$lacz_v_v12$deseq$de)

saveRDS(salmon.quant, file = file.path(output.data.dir, 'salmon.quant_analysis_set.rds'))


salmon.quant$aale$exo$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 0.5) %>% nrow()

salmon.quant$aale$exo$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -0.5, gene %in% meta.data$gencode.35.lnc.gene.names$gene) %>% nrow()

```

```{r}

salmon.quant <- readRDS(file = file.path(output.data.dir, 'salmon.quant_analysis_set.rds'))
ucsc.rmsk.salmon.quant <- readRDS(file = file.path(output.data.dir, 'ucsc.rmsk.salmon.quant_analysis_set.rds'))

lapply(c(1,2), function(i) { 
  
  imap(salmon.quant[names(salmon.quant) %in% c('aale', 'hbec')][[i]], function(x, x_name) {
    
    x$deseq$de %>%
      filter(abs(log2FoldChange) > 1, padj < 0.05, gene %in% msig.df$HALLMARK_INTERFERON_ALPHA_RESPONSE) %>% 
      mutate(id = x_name)
    
  })
}) %>% purrr::flatten() %>% 
  bind_rows() %>% 
  filter(!grepl('compare|hbec_|v_wt', id)) %>% 
  group_by(gene) %>% 
  summarize(context = list(id)) %>% 
  ggplot(aes(context)) + 
  geom_bar() + 
  ggupset::scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)))

msig.df[grepl('INTERFERON|INFLAMMATORY', names(msig.df))] %>% 
  lapply(function(x) as.data.frame(x)) %>% 
  bind_rows(.id = 'pathway') %>% 
  rbind(salmon.quant$aale$intra$deseq$de %>% 
          filter(log2FoldChange > 1, padj < 0.05) %>% 
          mutate(pathway = 'kras.intra') %>% select(pathway, 'x' = gene)) %>% 
  group_by(x) %>% 
  summarize(pathways = list(pathway)) %>% 
  ggplot(aes(pathways)) + 
  geom_bar() + 
  ggupset::scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)), vjust = -0.5)

salmon.quant$aale$intra$deseq$de %>% 
          filter(padj < 0.05) %>%
  mutate(pathway = ifelse(gene %in% msig.df$HALLMARK_INTERFERON_ALPHA_RESPONSE, 'IFNA', 'other')) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = pathway)) + 
  geom_point(alpha = 0.3)

salmon.quant$aale$intra$deseq$de %>% filter(padj< 0.05) %>% 
  merge(salmon.quant$aale$exo$deseq$de %>% filter(padj < 0.05), by = 'gene') %>% 
  mutate(pathway = ifelse(gene %in% msig.df$HALLMARK_INTERFERON_ALPHA_RESPONSE, 'IFNA', 'other')) %>%
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, color = pathway)) + 
  geom_point(alpha = 0.3)

library(rGREAT)

salmon.quant$gxi.bed %>% 
  filter(gene %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.01, log2FoldChange > 1)$gene) %>% 
  select(chr, start, end, gene, strand) -> input_bed

consensus_atac_boolean_anno

job <- rGREAT::submitGreatJob(input_bed, species = 'hg38')

tb <- rGREAT::getEnrichmentTables(job)

rGREAT::plotRegionGeneAssociationGraphs(job)


availableOntologies(job)



```

```{r}
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome)

ifn_g <- msig.df$HALLMARK_INTERFERON_ALPHA_RESPONSE
ifn_a <- msig.df$HALLMARK_INTERFERON_GAMMA_RESPONSE
jak_stat <- msig.df$HALLMARK_IL6_JAK_STAT3_SIGNALING

options(meme_db = file.path(output.data.dir, 'reference.data', 
                            'HOCOMOCOv11_core_HUMAN_mono_meme_format.meme'))



salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 1.5, gene %in% c(ifn_g, ifn_a, jak_stat)) %>% 
  # filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj < 0.05, log2FoldChange > 0.5)$gene) %>%
  select(gene_id) %>% unlist() %>% unname() -> de_list

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj > 0.05, gene %in% c(ifn_g, ifn_a, jak_stat)) %>% 
  # filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj > 0.05, log2FoldChange > 0)$gene) %>%
  select(gene_id) %>% unlist() %>% unname() -> bg_de_list

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 0, gene %in% c(ifn_g, ifn_a), !gene_id %in% c(de_list)) %>% 
  filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj < 0.05, log2FoldChange < 0)$gene) %>%
  select(gene_id) %>% unlist() %>% unname() -> de_list_w_hbec


# salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
#   filter(padj < 0.05, log2FoldChange > 0) %>% #, gene %in% c(ifn_g, ifn_a)) %>% 
#   filter(!gene %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 0)$gene) %>% 
#   select(gene_id) %>% unlist() %>% unname() -> de_list_w_hbec

gr <- rowRanges(salmon.quant$gxi)

ifn_de_gr <- gr[names(gr) %in% de_list]
ng_de_gr <- gr[names(gr) %in% bg_de_list]

promoter_seqs <- 
  GenomicFeatures::getPromoterSeq(subject = BSgenome.Hsapiens.UCSC.hg38, 
                                  query = ifn_de_gr, upstream = 500, downstream = 500)

promoter_seqs_hbec <- 
  GenomicFeatures::getPromoterSeq(subject = BSgenome.Hsapiens.UCSC.hg38, 
                                   query = ng_de_gr, upstream = 500, downstream = 500)


# dreme_results <- memes::runMeme(input = promoter_seqs, control = NA)
# 
tomTom_res <- memes::runAme(control = promoter_seqs_hbec,
                            input = promoter_seqs)

tomTom_res %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ame_tf
# %>% 
  # select(TF) %>% unlist() %>% unname() -> ame_tf

salmon.quant$aale$intra$deseq$de %>% 
  merge(ame_tf, by.x = 'gene', by.y = 'TF') %>% 
  # filter(gene %in% ame_tf) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.35, 0.79)) 

panel.save(figure = 1, name = 'enriched_tf_volcano', width = 'single', height = 'single')


isre_motif <- MotifDb::query(MotifDb::MotifDb, "ISRE") %>% 
  universalmotif::convert_motifs()

isre_motif <- isre_motif[[1]]

irf9_motif <- MotifDb::query(MotifDb::MotifDb, "IRF9_HUMAN.H11MO.0.C") %>% 
  universalmotif::convert_motifs()

irf7_motif <- MotifDb::query(MotifDb::MotifDb, "IRF7_HUMAN.H11MO.0.C") %>% 
  universalmotif::convert_motifs()

irf1_motif <- MotifDb::query(MotifDb::MotifDb, "IRF1_HUMAN.H11MO.0.A") %>% 
  universalmotif::convert_motifs()

fos_motif <- MotifDb::query(MotifDb::MotifDb, "FOS") %>% 
  universalmotif::convert_motifs()

foxp2_motif <- MotifDb::query(MotifDb::MotifDb, 'FOXP2_HUMAN.H11MO.0.C') %>% 
  universalmotif::convert_motifs()


irf1_fimo_results <- memes::runFimo(promoter_seqs, irf1_motif, thresh = 1e-4) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id')
  # select(gene) %>% distinct()

irf7_fimo_results <- memes::runFimo(promoter_seqs, irf7_motif, thresh = 1e-4) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id')

irf9_fimo_results <- memes::runFimo(promoter_seqs, irf9_motif, thresh = 1e-4) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id') 

fos_fimo_results <- memes::runFimo(promoter_seqs, fos_motif, thresh = 1e-6) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id') 
  # select(gene) %>% distinct()

foxp2_fimo_results <- memes::runFimo(promoter_seqs, foxp2_motif, thresh = 1e-3) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id') 
  # select(gene) %>% distinct()

irf9_fimo_results <- memes::runFimo(promoter_seqs, irf9_motif, thresh = 1e-3) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id') 
  # select(gene) %>% distinct()

```

```{r}
salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 0.5, gene %in% meta.data$gencode.35.lnc.gene.names$gene) %>% View()
  select(gene_id) %>% unlist() %>% unname() -> de_list

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -0.5, gene %in% meta.data$gencode.35.lnc.gene.names$gene) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list_w_hbec

gr <- rowRanges(salmon.quant$gxi)

ifn_de_gr <- gr[names(gr) %in% de_list]
ifn_de_gr_hbec <- gr[names(gr) %in% de_list_w_hbec]

promoter_seqs <- 
  GenomicFeatures::getPromoterSeq(subject = BSgenome.Hsapiens.UCSC.hg38, 
                                  query = ifn_de_gr, upstream = 1000, downstream = 500)

promoter_seqs_hbec <- 
  GenomicFeatures::getPromoterSeq(subject = BSgenome.Hsapiens.UCSC.hg38, 
                                   query = ifn_de_gr_hbec, upstream = 1000, downstream = 500)


# dreme_results <- memes::runDreme(promoter_seqs, control = "shuffle")
# 
tomTom_res <- memes::runAme(control = 'shuffle',
                            input = promoter_seqs)

```

```{r}

ctrl_norm_factor <- 43576060
kras_norm_factor <- 60276212

kras_bam_path <- file.path(output.data.dir, 'narrow_output.data/bwa/mergedLibrary/kras_R1.mLb.clN.sorted.bam')

ctrl_bam_path <- file.path(output.data.dir, 'narrow_output.data/bwa/mergedLibrary/ctrl_R1.mLb.clN.sorted.bam')


salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 1.5, gene %in% c(ifn_g, ifn_a, jak_stat)) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list

# salmon.quant$aale$intra$deseq$de %>% 
#   filter(padj > 0.5) %>%  #, log2FoldChange < -1.5) %>%  #, log2FoldChange > 1.5, gene %in% c(ifn_g, ifn_a)) %>% 
#   select(gene_id) %>% unlist() %>% unname() -> ns_list

gr <- rowRanges(salmon.quant$gxi)

ifn_de_gr <- gr[names(gr) %in% de_list]

# ns_gr <- gr[names(gr) %in% ns_list]

de_promoter_regions <- promoters(ifn_de_gr, upstream = 1000, downstream = 500)
# ns_promoter_regions <- promoters(ns_gr, upstream = 1000, downstream = 200)



bamsignals::bamCoverage(kras_bam_path, de_promoter_regions, paired.end = 'extend') -> kras_de_prom_profile
bamsignals::bamCoverage(ctrl_bam_path, de_promoter_regions, paired.end = 'extend') -> ctrl_de_prom_profile

# bamsignals::bamCoverage(kras_bam_path, ns_promoter_regions, paired.end = 'extend') -> kras_ns_prom_profile
# bamsignals::bamCoverage(ctrl_bam_path, ns_promoter_regions, paired.end = 'extend') -> ctrl_ns_prom_profile

bamsignals::alignSignals(kras_de_prom_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(position = position - 1000,
         cov = (cov/kras_norm_factor) * 1e6,
         condition = 'kras',
         context = 'de') -> kras_de_prom_df

# bamsignals::alignSignals(kras_ns_prom_profile) %>% 
#   as.data.frame() %>% 
#   mutate(position = row_number()) %>% 
#   gather(prom, cov, -position) %>%
#   mutate(position = position - 1000,
#          cov = (cov/kras_norm_factor) * 1e6,
#          condition = 'kras',
#          context = 'ns') -> kras_ns_prom_df

bamsignals::alignSignals(ctrl_de_prom_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(position = position - 1000,
         cov = (cov/ctrl_norm_factor) * 1e6,
         condition = 'ctrl',
         context = 'de') -> ctrl_de_prom_df

# rbind(kras_de_prom_df, kras_ns_prom_df) %>% 
#   group_by(position, context) %>% 
#   summarise(ci = list(mean_cl_normal(cov))) %>% 
#   unnest(ci) %>% 
#   ggplot(aes(x = position, y = y, ymin = ymin, ymax = ymax, color = context)) + 
#   geom_ribbon() + 
#   geom_point()
  

rbind(kras_de_prom_df, ctrl_de_prom_df) %>%
  ggplot(aes(position, cov, group = condition, color = condition)) +
    stat_summary(geom="ribbon", fun.data=mean_cl_boot, width=0.1, conf.int=0.95, 
                 fill='grey', alpha = 0.1, size = rel(0.05), show.legend = F)+
    # stat_summary(geom="line", fun=mean, linetype="dashed", alpha = 0.15, size = rel(0.05)) +
    stat_summary(geom="point", fun=mean, alpha = 0.25, size = rel(1)) +
    scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
    xlab('distance to TSS') +
    ylab('mean CPM (95% CI)') + 
    xlim(-1000, 500) + 
    theme(legend.position = c(0.3, 0.99))

panel.save(figure = 2, name = 'isg_atac_enrichment', width = 'single', height = 'single')


de_promoter_regions_for_meme <- promoters(ifn_de_gr, upstream = 300, downstream = 500) %>% 
  memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38)

salmon.quant$aale$intra$deseq$de %>%
  filter(padj > 0.5, gene %in% c(ifn_g, ifn_a)) %>%
  select(gene_id) %>% unlist() %>% unname() -> ns_list
ns_gr <- gr[names(gr) %in% ns_list]
ns_promoter_regions <- promoters(ns_gr, upstream = 300, downstream = 500) %>% 
  memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38)


ifn_ame_out <- memes::runAme(de_promoter_regions_for_meme, control = ns_promoter_regions)

ifn_ame_out %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ifn_ame_out
# %>% 
  # select(TF) %>% unlist() %>% unname() -> ame_tf

salmon.quant$aale$intra$deseq$de %>% 
  merge(ifn_ame_out, by.x = 'gene', by.y = 'TF') %>% 
  # filter(gene %in% ame_tf) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.35, 0.79)) 


# avgBinnedSig <- rowMeans(bamsignals::alignSignals(kras_prom_profile))
# 
# avgBinnedSig %>% 
#   as.data.frame() %>% 
#   rownames_to_column('bin') %>% 
#   ggplot(aes(bin, .)) + 
#   geom_col()
# 
# bamsignals::alignSignals(bamsignals::bamCoverage(kras_bam_path, promoter_regions, paired.end = 'extend')) %>% dim()
```

```{r}


smRNA_de <- read_csv(file = file.path(output.data.dir, 'AALE_DESeq2.csv')) %>% 
  dplyr::rename(gene = 1,
                log2fc = 2)

smRNA_de %>% 
  # filter(gene %in% salmon.quant$aale$exo$deseq$de$gene)
  filter(padj < 0.05) %>% 
  # mutate()
  ggplot(aes(log2fc, -log10(padj))) + 
  geom_point() + 
  xlim(-4,4) + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2fc) > 1.5, gene, '')))

panel.save(figure = 'fig.supplement', name = 'smRNA_exo_de', width = 'single', height = 'single')


salmon.quant$aale$exo$deseq$de %>% 
  # filter(gene %in% salmon.quant$aale$exo$deseq$de$gene)
  filter(padj < 0.05) %>% 
  # mutate()
  ggplot(aes(log2FoldChange, -log10(padj))) + 
  geom_point() + 
  # xlim(-4,4) + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 1.5, gene, '')))


smRNA_de %>% 
  select(gene, log2fc, padj) %>% 
  filter(padj < 0.05) %>% 
  merge(salmon.quant$aale$intra$deseq$de %>% filter(padj < 0.05), by = 'gene') %>% 
  ggplot(aes(log2FoldChange, log2fc)) + 
  xlab('intra -- RNA') + ylab('exo -- OTTRseq') +
  geom_point() + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = gene), show.legend = F)

panel.save(figure = 'fig.supplement', name = 'smRNA_exo_v_intra_de', width = 'single', height = 'single')


iso_de <- read_csv(file = file.path(output.data.dir, 'unique_to_exo_DE_isoforms_P05.csv'))[, 2:15]

iso_exo <- read_csv(file = file.path(output.data.dir, 'annotated_unique_exo_isoforms.csv'))
iso_intra <- read_csv(file = file.path(output.data.dir, 'annotated_unique_intra_isoforms.csv'))


iso_exo %>% 
  filter(gene_biotype %in% c('lncRNA', 'protein_coding')) %>% 
  group_by(gene_biotype) %>% 
  summarize(count = length(gene_biotype)) %>% 
  ggplot(aes(gene_biotype, count, label = count)) + 
  geom_col() + 
  geom_hline(yintercept = seq(0,300,50), color = 'white') +
  coord_flip() +
  xlab('')

panel.save(figure = 'fig.supplement', name = 'iso_exo_count', width = 'single', height = 'single')


iso_de %>% 
  filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj))) + 
  geom_point() +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 7.5, gene, '')))
panel.save(figure = 'fig.supplement', name = 'iso_exo_de', width = 'single', height = 'single')


annotated_unique_exo_isoforms.csv



```

## Ex RNA
```{r}
salmon.quant$aale$exo$deseq$de %>% 
  dplyr::rename('EXO_aale' = log2FoldChange) %>% 
  filter(padj < 0.05) %>% 
  merge(salmon.quant$aale$intra$deseq$de %>% 
          dplyr::rename('INTRA_aale' = log2FoldChange) %>% 
          filter(padj < 0.05), by = 'gene') %>% 
  mutate(unique = ifelse(gene %in% uniq_kras_peak_genes$`Gene Name`, 'KRAS', 
                         ifelse(gene %in% uniq_ctrl_peak_genes$`Gene Name`, 'CTRL', NA))) %>% 
  ggplot(aes(EXO_aale, INTRA_aale, color = unique)) + 
  geom_point() + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2]), name = 'unique ATAC peak') +
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = ifelse(abs(EXO_aale) > 1 & abs(INTRA_aale) > 1, gene, '')), show.legend = F) +
  theme(legend.position = c(0.99, 0.25)) +
  geom_line(stat = 'smooth', method = 'lm', linetype = 'dashed', alpha = 0.8, color = 'purple') +
  annotate(geom = 'text', x = 2.35, y = -2, fontface = 'italic',
           label = paste0('y', ' ', ' = -0.99 + 0.66x; \nadj. R2 = 0.49'), 
           size = rel(3.75), alpha = 0.8)

salmon.quant$aale$exo$deseq$de %>% 
  dplyr::rename('EXO_aale' = log2FoldChange) %>% 
  filter(padj < 0.05) %>% 
  merge(salmon.quant$aale$intra$deseq$de %>% 
          dplyr::rename('INTRA_aale' = log2FoldChange) %>% 
          filter(padj < 0.05), by = 'gene') %>% 
  lm(EXO_aale ~ INTRA_aale, data = .) %>% 
  summary()

salmon.quant$aale$kras_compare$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.lnc.gene.names$gene,
         padj < 0.01) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 1, gene, '')), show.legend = F) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99,0.99))

salmon.quant$aale$ctrl_compare$deseq$de %>% 
  filter(gene %in% meta.data$gencode.35.lnc.gene.names$gene,
         padj < 0.01) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 1, gene, '')), show.legend = F) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99,0.99))


salmon.quant$aale$ctrl_compare$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.lnc.gene.names$gene,
         padj < 0.01) %>%
  mutate(ctrl = log2FoldChange) %>% 
  merge(salmon.quant$aale$kras_compare$deseq$de %>% 
    filter(!gene %in% meta.data$gencode.35.lnc.gene.names$gene,
           padj < 0.01) %>%
    mutate(kras = log2FoldChange), by = 'gene') %>% 
  ggplot(aes(kras, ctrl, label = gene)) + 
  geom_point(alpha = 0.6) + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(kras) > 3 & abs(ctrl) > 3, gene, '')), show.legend = F) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99,0.99))



salmon.quant$aale$exo$deseq$de %>% 
  dplyr::rename('EXO_aale' = log2FoldChange) %>% 
  filter(padj < 0.05) %>% 
  merge(salmon.quant$aale$intra$deseq$de %>% 
          dplyr::rename('INTRA_aale' = baseMean) %>% 
          filter(padj < 0.05), by = 'gene') %>% 
  mutate(unique = ifelse(gene %in% uniq_kras_peak_genes$`Gene Name`, 'KRAS', 
                         ifelse(gene %in% uniq_ctrl_peak_genes$`Gene Name`, 'CTRL', NA))) %>% 
  ggplot(aes(EXO_aale, INTRA_aale, color = unique)) + 
  geom_point() + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2]), name = 'unique ATAC peak') +
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = ifelse(abs(EXO_aale) > 1 & abs(INTRA_aale) > 1, gene, '')), show.legend = F) +
  theme(legend.position = c(0.99, 0.99)) +
  geom_line(stat = 'smooth', method = 'lm', linetype = 'dashed', alpha = 0.8, color = 'purple')


panel.save(figure = 3, name = 'intra_v_exo_de', width = 'single', height = 'single')


salmon.quant$aale$exo$deseq$de %>% 
  mutate(biotype = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', 'coding')) %>%
  # filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = biotype)) + 
  geom_point(alpha = 0.6) + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 1, gene, '')), show.legend = F) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99,0.99))

panel.save(figure = 3, name = 'exo_de', width = 'single', height = 'single')



ucsc.rmsk.salmon.quant$aale$exo$deseq$de %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # xlim(-2, 2) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95))

panel.save(figure = 3, name = 'exo_te_de', width = 'single', height = 'single')

ucsc.rmsk.salmon.quant$aale$kras_compare$deseq$de %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  filter(clade %in% c('SINE', 'LINE', 'LTR', 'DNA')) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # xlim(-2, 2) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95)) + 
  facet_col(~clade)

ucsc.rmsk.salmon.quant$aale$ctrl_compare$deseq$de %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  filter(clade %in% c('SINE', 'LINE', 'LTR', 'DNA')) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # xlim(-2, 2) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95)) + 
  facet_col(~clade)

ucsc.rmsk.salmon.quant$aale$ctrl_compare$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.01) %>%
  mutate(ctrl = log2FoldChange) %>% 
  merge(ucsc.rmsk.salmon.quant$aale$kras_compare$deseq$de %>% 
    mutate(kras = log2FoldChange), by = 'gene') %>% 
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  filter(clade %in% c('SINE', 'LINE', 'LTR', 'DNA')) %>% 
  ggplot(aes(kras, ctrl, label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(kras) > 3 & abs(ctrl) > 3, gene, '')), show.legend = F) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99,0.99)) + 
  facet_col(~clade)



salmon.quant$aale$intra$deseq$H.fgsea %>% 
  dplyr::rename(INTRA_aale = NES) %>% 
  merge(salmon.quant$aale$exo$deseq$H.fgsea %>% 
          dplyr::rename(EXO_aale = NES), by = 'pathway') %>% View()
  ggplot(aes(EXO_aale, INTRA_aale, label = pathway)) + 
  geom_point() +
  ggrepel::geom_text_repel()
  
  
salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, abs(log2FoldChange) > 0.5) %>% 
  mutate(context = ifelse(log2FoldChange > 0, 'INTRA_up', 'INTRA_dn')) %>% 
  select(gene, context) %>% 
  rbind(
    salmon.quant$aale$exo$deseq$de %>% 
    filter(padj < 0.05, abs(log2FoldChange) > 0.5) %>% 
    mutate(context = ifelse(log2FoldChange > 0, 'EXO_up', 'EXO_dn')) %>% 
    select(gene, context)
    ) %>% 
  group_by(gene) %>% 
  summarize(contexts = list(context)) %>% 
  ggplot(aes(contexts)) + 
  geom_bar() + 
  ggupset::scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)), vjust = -0.5) +
  xlab('') + ylab('overlapping genes') -> exo_v_intra_upset
  
panel.save(plot.in = exo_v_intra_upset, figure = 3, name = 'intra_v_exo_upset', width = 'single', height = 'single')




kras.sight <- read_csv(file.path(output.data.dir, 'kras.nanosight.data.csv'), skip = 2,
                       col_names = c('size', 'r1', 'r2', 'r3', NA, NA, NA, NA)) %>% 
  mutate(condition = 'kras')
ctrl.sight <- read_csv(file.path(output.data.dir, 'ctrl.nanosight.data.csv'), skip = 2,
                       col_names = c('size', 'r1', 'r2', 'r3', NA, NA, NA, NA)) %>% 
  mutate(condition = 'ctrl')

nano_sight <- rbind(kras.sight, ctrl.sight)

nano_sight %>% 
  select(size, r1, r2, r3, condition) %>% 
  gather(rep, value, -condition, -size) %>% 
  filter(size <= 400) %>% 
  mutate(value = value/1e6) %>% 
  group_by(size, condition) %>% 
  summarize(y = mean(value),
            ymin = mean(value) - sd(value),
            ymax = mean(value) + sd(value)) %>% 
  ggplot(aes(size, y = y, ymin = ymin, ymax = ymax)) +
  geom_ribbon(fill = 'grey', alpha = 0.4) +
  geom_line(color = 'darkgrey') +
  stat_peaks(geom = 'text', span = 35, vjust = -0.5, y.label.fmt = '%#f') +
  facet_row(~condition) + 
  ylab('1e6 particles per mL') + xlab('nm')
      
panel.save(figure = 3, name = 'exo_nanosight_dist', width = 'single', height = 'single')


ucsc.rmsk.salmon.quant$aale$exo$deseq$counts %>% 
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene') %>%
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR')) %>% 
  gather(sample, count, -gene, -clade, -family) %>% 
  separate(sample, sep = '[.]', into = c(NA, NA, NA, 'condition')) %>%
  ggplot(aes(condition, log1p(count))) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_jitter(alpha = 0.2) + 
  facet_wrap(~clade) +
  stat_compare_means(method = 'wilcox', label = 'p.format')
  
panel.save(figure = 3, name = 'exo_nanosight_dist', width = 'single', height = 'single')



ucsc.rmsk.salmon.quant$aale$platform_compare$deseq$counts %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene') %>%
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR')) %>% 
  gather(sample, count, -gene, -clade, -family) %>% 
  separate(sample, sep = '[.]', into = c('platform', NA, NA, 'condition')) %>%
  mutate(platform = ifelse(platform == 'rnaEASY', 'EXO', 'INTRA'),
         condition = paste0(platform, '_', condition)) %>% 
  filter(platform == 'EXO') %>% 
  ggplot(aes(condition, log1p(count))) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_jitter(alpha = 0.1) + 
  facet_wrap(~clade) +
  stat_compare_means(method = 'wilcox', label = 'p.format', hide.ns = T) +#, label.y.npc = 0.15
  #                    comparisons = list(c('INTRA_ctrl', 'INTRA_kras'),
  #                                       c('EXO_ctrl', 'EXO_kras'),
  #                                       c('INTRA_kras', 'EXO_kras'),
  #                                       c('INTRA_ctrl', 'EXO_ctrl'))) + 
  xlab('') + ylab('log(norm.count + 1)')

panel.save(figure = 4, name = 'te_exo_abundance_compare', width = 'single', height = 'single')


a549_te_de$de_682 %>% 
  mutate(oe = 'ZNF682') %>% 
  filter(!gene_id %in% meta.data$gencode.35.tx.to.gene$gene,
         !grepl(')n', gene_id)) %>% 
  rbind(a549_te_de$de_257 %>% 
    mutate(oe = 'ZNF257') %>% 
    filter(!gene_id %in% meta.data$gencode.35.tx.to.gene$gene,
         !grepl(')n', gene_id))) %>% 
  # merge(salmon.quant$aale$intra$deseq$de, by = 'gene') %>% 
  filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene_id)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel() +
  # ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange.x < -0.1 & log2FoldChange.y > 0.1, gene, '')), max.overlaps = 5) +
  facet_col(~oe) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  geom_hline(yintercept = 0, alpha = 0.3, linetype = 'dashed') #+ 
  # ylab('AALE log2FC') + xlab('A549 log2FC')


ucsc.rmsk.salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # xlim(-2, 2) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95))


ucsc.rmsk.salmon.quant$aale$intra$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  # xlim(-2, 2) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95))


ucsc.rmsk.salmon.quant$aale$intra$deseq$de %>% 
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  xlim(-6, 6) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) >= 1.15, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99,0.95))

ucsc.rmsk.salmon.quant$aale$intra$deseq$de %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  mutate(EXO_aale = log2FoldChange) %>% 
  merge(ucsc.rmsk.salmon.quant$aale$exo$deseq$de %>%
    filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
           padj < 0.05) %>%
    mutate(INTRA_hbec = log2FoldChange), by = 'gene') %>% 
  ggplot(aes(EXO_aale, INTRA_hbec, label = gene)) + 
  geom_point() + 
  ggrepel::geom_text_repel() + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4)


ucsc.rmsk.salmon.quant$aale$intra$deseq$de %>%
  filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
         padj < 0.05) %>%
  mutate(context = 'aale') %>% 
  rbind(ucsc.rmsk.salmon.quant$hbec$lacz_v_v12$deseq$de %>%
    filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
           padj < 0.05) %>% 
      mutate(context = 'hbec')) %>% 
  merge(meta.data$te.families.clades, by = 'gene') %>% 
  mutate(clade = factor(clade, levels = c('DNA', 'LINE', "LTR", 'Satellite', 'SINE'))) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = clade)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange > 0, gene, '')), show.legend = F, max.overlaps = 40) +
  scale_color_viridis_d(direction = -1) +
  facet_col(~context, scales = 'free_y') +
  xlim(-6,4) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(.39,.39))

panel.save(figure = 4, name = 'intra_te_de', width = 'single', height = 'single')





ucsc.rmsk.salmon.quant$aale$exo$deseq$counts %>% 
  # filter(gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(is.na(clade), 'gencode', clade),
         clade = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', clade)) %>% 
  gather(sample, count, -family, -clade, -gene) %>% 
  mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', 'kras')) %>% 
  group_by(sample, condition, clade) %>% 
  summarize(clade_counts = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(frac = clade_counts/sum(clade_counts)) %>% 
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR', 'Satellite', 'gencode', 'lncRNA')) %>%
  ggplot(aes(sample, frac, fill = clade)) + 
  geom_col()
  # ggplot(aes(condition, frac)) + 
  # geom_boxplot(fill = NA, outlier.colour = NA) + 
  # geom_point() +
  # facet_wrap(~clade, scales = 'free_y')

ucsc.rmsk.salmon.quant$aale$exo$deseq$counts %>% 
  # filter(gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(is.na(clade), 'coding', clade),
         clade = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', clade)) %>% 
  gather(sample, count, -family, -clade, -gene) %>% 
  filter(!grepl('^MT|^RP', gene)) %>% 
  mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', 'kras'),
         sample = str_replace(sample, 'rnaEASY.aale.', '')) %>% 
  group_by(sample, condition, clade) %>% 
  summarize(clade_counts = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(frac = clade_counts/sum(clade_counts)) %>% 
  mutate(clade = factor(clade, levels = c('SINE', "LINE",'DNA', 'LTR', 'Satellite', 'lncRNA', 'coding'))) %>% 
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR', 'Satellite', 'coding', 'lncRNA')) %>%
  ggplot(aes(sample, frac, fill = clade)) + 
  geom_col() + 
  geom_hline(yintercept = seq(0,1,0.25), color = 'white', alpha = 0.6) +
  scale_fill_viridis_d(direction = -1) + 
  facet_col(~condition, scales = 'free_x') + 
  xlab('') + 
  ylab('fraction of norm. counts') + 
  theme(legend.position = 'top', legend.justification = 'center')

panel.save(figure = 3, name = 'biotype_dist', width = 'single', height = 'single')


ucsc.rmsk.salmon.quant$aale$intra$deseq$counts %>% 
  # filter(gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(is.na(clade), 'coding', clade),
         clade = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', clade)) %>% 
  gather(sample, count, -family, -clade, -gene) %>% 
  filter(!grepl('^MT|^RP', gene)) %>% 
  mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', 'kras'),
         sample = str_replace(sample, 'rnaEASY.aale.', '')) %>% 
  group_by(sample, condition, clade) %>% 
  summarize(clade_counts = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(frac = clade_counts/sum(clade_counts)) %>% 
  mutate(clade = factor(clade, levels = c('SINE', "LINE",'DNA', 'LTR', 'Satellite', 'lncRNA', 'coding'))) %>% 
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR', 'Satellite', 'lncRNA')) %>%
  ggplot(aes(sample, frac, fill = clade)) + 
  geom_col() + 
  # geom_hline(yintercept = seq(0,1,0.25), color = 'white', alpha = 0.6) +
  scale_fill_manual(values = c(viridis(7, direction = -1)[[1]],viridis(7, direction = -1)[[2]],viridis(7, direction = -1)[[3]],
                               viridis(7, direction = -1)[[4]],viridis(7, direction = -1)[[5]],viridis(7, direction = -1)[[6]])) +
  facet_col(~condition, scales = 'free_x') + 
  xlab('') + 
  ylab('fraction of norm. counts') + 
  theme(legend.position = 'top', legend.justification = 'center')

panel.save(figure = 'fig.supplement', name = 'intra_biotype_dist', width = 'single', height = 'single')


ucsc.rmsk.salmon.quant$aale$ctrl_compare$deseq$counts %>% 
  # filter(gene %in% meta.data$gencode.35.tx.to.gene$gene) %>% 
  merge(meta.data$te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(is.na(clade), 'gencode', clade),
         clade = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', clade)) %>% 
  gather(sample, count, -family, -clade, -gene) %>% nrow()
  filter(!grepl('^MT|^RP', gene)) %>% 
  mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', 'kras')) %>% 
  group_by(sample, condition, clade) %>% 
  summarize(clade_counts = sum(count)) %>% 
  group_by(sample) %>% 
  mutate(frac = clade_counts/sum(clade_counts)) %>% 
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR', 'Satellite', 'gencode', 'lncRNA')) %>%
  ggplot(aes(sample, frac, fill = clade)) + 
  geom_col()

  
```

```{r}
Biostrings::letterFrequency(x = seqs, letters = "GC", as.prob = TRUE)

gr[, 'GC'] <- Biostrings::letterFrequency(x = seqs, letters = "GC", as.prob = TRUE)


rowRanges(salmon.quant$gxi)[, 'gene_id' %in% 
                              filter(salmon.quant$aale$intra$deseq$de, padj < 0.01, log2FoldChange > 1)$gene]


subsetByOverlaps(rowRanges(salmon.quant$gxi), 
                 makeGRangesFromDataFrame(salmon.quant$gxi.bed %>% 
                                            filter(filter(salmon.quant$aale$intra$deseq$de, padj < 0.01,
                                                          log2FoldChange > 1)$gene)))


Biostrings::letterFrequency(x = b)

makeGRangesFromDataFrame(salmon.quant$gxi.bed %>% 
  filter(gene %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.01, log2FoldChange > 1)$gene))

```

```{r}

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 1.5, gene %in% c(ifn_g, ifn_a, jak_stat)) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list

hm_col_data <- 
  salmon.quant$all_intra$deseq$counts[salmon.quant$all_intra$deseq$counts$gene %in%
                                        de_list, 
                                      !grepl('intra', 
                                             colnames(salmon.quant$all_intra$deseq$counts))]
hm_count_data <- 
  salmon.quant$all_intra$deseq$counts[salmon.quant$all_intra$deseq$counts$gene %in%
                                        de_list,
                                      grepl('aale|wt2_kras_v12|wt2_lacz',
                                          colnames(salmon.quant$all_intra$deseq$counts))]

sample_col_data <- salmon.quant$all_intra$deseq$col.data %>% 
  filter(names %in% colnames(hm_count_data)) %>% 
  mutate(cell_line = names) %>% 
  separate(cell_line, sep = '[.]', into = c(NA, 'cell_line', NA, NA)) %>% 
  column_to_rownames('names')

rownames(hm_count_data) <- hm_col_data$gene

hm_count_data <- scale(t(log1p(hm_count_data)), center = T)

side_ha <- ComplexHeatmap::rowAnnotation(df = sample_col_data %>% select(cell_line), 
                                         col = list(cell_line = c('aale' = viridis(2)[1],
                                                                  'hbec' = viridis(2)[2])), show_annotation_name = F)

salmon.quant$all_intra$deseq$counts %>% 
  filter(gene %in% de_list) %>% 
  select(gene) %>% 
  mutate(FOS = ifelse(gene %in% fos_fimo_results$gene, T, F),
         ISRE = ifelse(gene %in% isre_fimo_results$gene, T, F)) %>% 
  column_to_rownames('gene') -> motif_anno_df

bottom_ha <- ComplexHeatmap::columnAnnotation(df = motif_anno_df,
                                              col = list(FOS = c('TRUE' = 'black', 'FALSE' = 'white'),
                                                         ISRE = c('TRUE' = 'black', 'FALSE' = 'white')))

set.seed(129)

ComplexHeatmap::Heatmap(hm_count_data, col = viridis(16, option = 'plasma'),
                        name = 'z-score',
                        width = panel.figure.width.double, 
                        height = panel.figure.height,
                        row_km = 2, row_title = c('WT', 'G12'),
                        use_raster = T, show_row_names = F,
                        left_annotation = side_ha,
                        bottom_annotation = bottom_ha)
  
```

```{r}
msig.df[grepl('INTERFERON', names(msig.df))] %>% 
                             unlist() %>% unname() -> ifn_gene_list

msig.df[grepl('OXIDATIVE|XENOBIOTIC', names(msig.df))] %>% 
                             unlist() %>% unname() -> metab_gene_list


salmon.quant$aale$intra$deseq$H.fgsea %>% 
  group_by(pathway) %>% 
  mutate(pathway = str_replace_all(pathway, 'HALLMARK ', ''),
         pathway = str_replace_all(pathway, ' ', '\n'),
         coverage = round(length(unlist(leadingEdge))/size, 3),
         NES = round(NES, 3)) %>% 
  arrange(-NES, padj) %>% 
  filter(!grepl('ZNF', pathway)) %>% 
  ggplot(aes(NES, reorder(pathway, NES), fill = padj)) + 
  geom_col(color = 'white') + 
  geom_text(aes(label = coverage, x = NES - 0.6 * sign(NES)), color = 'white') + 
  scale_fill_gsea(reverse = T) + 
  ylab('') + 
  theme(legend.position = c(0.95, 0.16), axis.text.y = element_text(size = base_size))

panel.save(figure = 1, name = 'gsea_labelled_bar', width = 'single', height = 'double')


exp_tbl %>% 
  select(pathway, padj, NES, coverage) %>% 
  mutate(padj = format(padj, scientific= T, digits = 2)) %>% 
  gridExtra::grid.table()
  


salmon.quant$aale$intra$deseq$de %>% 
  mutate(padj = ifelse(padj == 0, 
                       min(filter(salmon.quant$aale$intra$deseq$de, padj != 0)$padj), padj)) %>% 
  mutate(gene_set = ifelse(gene %in% ifn_gene_list,'IFN', 'other'),
         gene_set = ifelse(gene %in% msig.df$HALLMARK_KRAS_SIGNALING_UP | gene == 'KRAS', 'KRAS', gene_set),
         gene_set = ifelse(gene %in% meta.data$trono.krab.znfs$gene, 'ZNF', gene_set)) %>% 
  mutate(gene_set = factor(gene_set, levels = c('KRAS','IFN','ZNF', 'other'))) %>% 
  filter(padj < 0.05, gene_set != 'other') %>% 
  mutate(kras = ifelse(gene == 'KRAS', T, F)) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene, color = kras)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 5 & gene_set != 'other' | 
                                                padj < 0.0001 & gene_set != 'other' | gene == "KRAS", gene, ''))) +
  facet_col(~gene_set, scales = 'free_y') +
  scale_color_manual(values = c('black', 'red'), guide = 'none') +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed')

panel.save(figure = 1, name = 'facet_set_volcano', width = 'single', height = 'double')

salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
  mutate(padj = ifelse(padj == 0, 
                       min(filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj != 0)$padj), padj)) %>% 
  mutate(gene_set = ifelse(gene %in% ifn_gene_list,'IFN', 'other'),
         gene_set = ifelse(gene %in% msig.df$HALLMARK_KRAS_SIGNALING_UP, 'KRAS', gene_set),
         gene_set = ifelse(gene %in% meta.data$trono.krab.znfs$gene, 'ZNF', gene_set)) %>% 
  mutate(gene_set = factor(gene_set, levels = c('KRAS','IFN','ZNF', 'other'))) %>% 
  filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 5 & gene_set != 'other' | 
                                                padj < 0.0001 & gene_set != 'other', gene, ''))) +
  facet_col(~gene_set, scales = 'free_y') +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + xlim(-10,10)

panel.save(figure = 'fig.supplement', name = 'hbec_facet_set_volcano', width = 'single', height = 'double')


salmon.quant$aale$intra$deseq$de %>% 
  mutate(padj = ifelse(padj == 0, 
                       min(filter(salmon.quant$aale$intra$deseq$de, padj != 0)$padj), padj)) %>% 
  mutate(biotype = ifelse(gene %in% meta.data$gencode.35.lnc.gene.names$gene, 'lncRNA', 'protein-coding')) %>% 
  # filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  # scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 5 |
                                                -log10(padj) > 250, gene, ''))) +
  facet_row(~biotype, scales = 'free')

panel.save(figure = 'fig.supplement', name = 'facet_biotype_volcano', width = 'double', height = 'single')


salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
   filter(padj < 0.05, log2FoldChange > 1) %>% 
   select(gene) %>% unlist() %>% unname() -> hbec_up

salmon.quant$aale$intra$deseq$de %>% 
  mutate(padj = ifelse(padj == 0, 
                       min(filter(salmon.quant$aale$intra$deseq$de, padj != 0)$padj), padj)) %>% 
  filter(grepl('^HOX', gene)) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel() +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed')


# salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
salmon.quant$aale$intra$deseq$de %>% 
  mutate(cell_line = 'AALE') %>% 
  rbind(salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
    mutate(cell_line = 'HBEC')) %>% 
  filter(padj < 0.05) %>% 
  filter(gene %in% c(ifn_a, ifn_g)) %>% 
  mutate(IRF9 = ifelse(gene %in% irf9_fimo_results$gene, T, F)) %>%
  ggplot(aes(IRF9, log2FoldChange)) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_jitter(alpha = 0.4, width = 0.15) +
  facet_col(~cell_line, scales = 'free_y') +
  # geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 1, gene, ''))) +
  stat_compare_means(label.x.npc = 0.55, label.y.npc = 0.03) + 
  xlab('IRF9 motif')
  
panel.save(figure = 1, name = 'facet_set_irf9_expression', width = 'single', height = 'single')


salmon.quant$aale$intra$deseq$de %>% 
  dplyr::rename('AALE' = log2FoldChange) %>% 
  filter(padj < 0.05,
         gene %in% c(ifn_a, ifn_g)) %>% 
  merge(salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
          dplyr::rename('HBEC' = log2FoldChange) %>% 
          filter(padj < 0.05) %>% 
          mutate(HBEC = HBEC), by = 'gene') %>% 
  # mutate(ISRE = ifelse(gene %in% irf1_fimo_results$gene, T, F)) %>% 
  ggplot(aes(AALE, HBEC)) +
  geom_point() + 
  xlab('AALE log2FC') + ylab('HBEC log2FC') +
  # scale_color_manual(values = c(viridis(3)[1], viridis(3)[2]), name = 'ISRE motif') +
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = ifelse(AALE > 1 & HBEC > 1, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99, 0.22))
  
  salmon.quant$aale$intra$deseq$de %>% 
  dplyr::rename('AALE' = log2FoldChange) %>% 
  filter(padj < 0.05,
         gene %in% c(ifn_a, ifn_g)) %>% 
  merge(salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
          dplyr::rename('HBEC' = log2FoldChange) %>% 
          filter(padj < 0.05) %>% 
          mutate(HBEC = HBEC), by = 'gene') %>% 
    lm(AALE ~ HBEC, data = .) %>% summary()


panel.save(figure = 1, name = 'de_scatter_ifn', width = 'single', height = 'single')




salmon.quant$aale$intra$deseq$de %>% 
  dplyr::rename('AALE' = log2FoldChange) %>% 
  filter(padj < 0.05,
         gene %in% meta.data$gencode.35.lnc.gene.names$gene) %>%
  merge(salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
          dplyr::rename('HBEC' = log2FoldChange) %>% 
          filter(padj < 0.05) %>% 
          mutate(HBEC = HBEC), by = 'gene') %>% 
  # mutate(ISRE = ifelse(gene %in% irf9_fimo_results$gene, T, F)) %>% 
  ggplot(aes(AALE, HBEC)) + 
  geom_point() + 
  scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = ifelse(AALE > 1 | HBEC > 1, gene, '')), show.legend = F) + 
  theme(legend.position = c(0.99, 0.92))



```
### rerwip PCA
```{r}
set.seed(234)
### Gencode PCA

salmon.quant$all_intra$deseq$col.data

salmon.quant$aale$exo$deseq$gxi.vst.counts %>% apply(., 1, sd) -> pca_sd
salmon.quant$aale$exo$deseq$gxi.vst.counts[!pca_sd == 0, ] -> pca_filt
salmon.quant$aale$exo$deseq$col.data -> pca_col_data

pca_filt %>%
  t() %>% 
  as.data.frame() %>% 
  prcomp(scale=T) -> pca_out_raw

summary(pca_out_raw)
pcs.of.interest <- apply(pca_out_raw$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[1]
 
# convert output to dataframe
pca.out <- as.data.frame(pca_out_raw$x) %>% 
  rownames_to_column('names') %>% 
  merge(pca_col_data, by ='names')
# summarize PC contributions
pca.out.summary <- 
  as.data.frame(summary(pca_out_raw)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)
# summarize PC contributions
pca.out.contributions <- 
  as.data.frame(pca_out_raw$rotation) %>% 
  rownames_to_column('tx') %>% 
  gather(pc, contrib, -tx) %>% 
  #merge(gencode.reference,
  #      by.x = 'tx',
  #      by.y = 'gene.name') %>% 
  group_by(pc) %>% 
  mutate(contrib = abs(contrib)) %>% 
  filter(contrib > quantile(contrib, 0.95))
# plot PC1 and PC2
pca.out %>% 
  rownames_to_column('sample') %>% 
  #separate(sample, sep = '[.]', into = c(NA, 'condition', NA)) %>% 
  ggplot(aes(PC1, PC2,
             fill = condition,
             color = condition)) + 
    geom_point(size = rel(2)) + 
    scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
    xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                            digits = 3), sep = ' ')) +
    scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
    geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
    geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
    theme(legend.position = c(0.30, 0.25),
            legend.key.width = unit(0.125,'cm'),
            legend.key.height = unit(0.4, 'cm'))

panel.save(figure = 'fig.supplement', name = 'aale_exo_pca', width = 'single', height = 'single')


# print top contibutors per PC
pca.out.contributions %>%
  group_by(pc) %>%
  filter(pc %in% c("PC1", "PC2")) %>% 
  select(pc, contrib, tx) %>% 
  mutate(contrib = abs(contrib)) %>% 
  top_n(5, wt = contrib) %>% 
  arrange(pc, -contrib)
```

```{r}
toil.survival <- read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.survival.tsv.gz')) 
luad.phenotypes <- 
  read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.GDC_phenotype.tsv.gz')) %>% 
  select(sample = submitter_id.samples, 
       sample_type = sample_type.samples,
       stage = tumor_stage.diagnoses,
       age = age_at_index.demographic,
       sex = gender.demographic,
       race = race.demographic)
luad.mutect <- read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.mutect2_snv.tsv.gz')) %>% 
  filter(gene %in% c('KRAS')) %>% 
  select(sample = Sample_ID, gene, aa_change = Amino_Acid_Change) %>% 
  pivot_wider(names_from = gene, values_from = aa_change) %>% 
  mutate(across(.cols = where(is.list), paste)) 

luad.phenotypes.merge <- luad.phenotypes %>% 
  merge(luad.mutect, by = 'sample', all.x = T) %>% 
  filter(sample_type %in% c('Primary Tumor', 'Solid Tissue Normal')) %>% 
  merge(toil.survival, by = 'sample') %>% 
  mutate(sample = str_sub(sample, end = str_length(sample) - 1)) %>% 
  replace_na(list(KRAS = 'WT')) %>% 
  mutate(KRAS = ifelse(sample_type == 'Solid Tissue Normal', 'WT_normal', KRAS))

norm.counts.final <- read_tsv(file.path(output.data.dir, 'tcga.data', 'luad.data.frame.tsv'))
```

```{r}

norm.counts.final %>% 
  filter(sample != 'TCGA-44-3917-01') %>% 
  filter(KRAS %in% c('p.G12D', 'WT_normal')) -> working_toil_counts

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 1.5, gene %in% c(ifn_a, ifn_g, jak_stat)) -> aale_ifn

working_toil_counts %>% 
  filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  # filter(gene %in% uniq_peak_genes$`Gene Name`) %>% 
  # filter(gene %in% aale_ifn$gene) %>%
  group_by(gene) %>%
  mutate(count = scale(log1p(count), center = T)) %>%
  select(sample, gene, count) %>%
  pivot_wider(names_from = gene, values_from = count) %>%
  distinct() %>% 
  column_to_rownames('sample') %>% 
  t() -> toil_hm_counts

working_toil_counts %>% 
  select(sample, 'G12D' = KRAS) %>%
  distinct() %>% 
  column_to_rownames('sample') %>% 
  ComplexHeatmap::rowAnnotation(df = . , 
                                   col = list(G12D = c('WT_normal' = 'white', 
                                                       'p.G12D' = viridis(3)[2])),
                                   show_legend = F,
                                annotation_name_side = 'bottom',
                                simple_anno_size = unit(2.5, "mm")) -> toil_sample_anno



working_toil_counts %>% 
  filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  # filter(gene %in% uniq_peak_genes$`Gene Name`) %>% 
  # filter(gene %in% aale_ifn$gene) %>%
  select(gene) %>% 
  distinct() %>% 
  mutate(IRF9 = ifelse(gene %in% irf9_fimo_results$gene, T, F),
         IRF1 = ifelse(gene %in% irf1_fimo_results$gene, T, F),
         IRF7 = ifelse(gene %in% irf7_fimo_results$gene, T, F)) %>% 
  column_to_rownames('gene') %>% 
  ComplexHeatmap::columnAnnotation(df =.,
                                col = list(IRF9 = c('TRUE' = viridis(2)[1], 'FALSE' = 'white'),
                                           IRF1 = c('TRUE' = viridis(2)[1], 'FALSE' = 'white'),
                                           IRF7 = c('TRUE' = viridis(2)[1], 'FALSE' = 'white')),
                                show_legend = F,
                                annotation_name_side = 'right',
                                simple_anno_size = unit(2.5, "mm")) -> gene_motif_anno




ComplexHeatmap::Heatmap(t(toil_hm_counts), 
                        name = 'z-score', 
                        show_row_names = F,
                        left_annotation = toil_sample_anno,
                        bottom_annotation = gene_motif_anno,
                        # heatmap_width = unit(panel.figure.width, 'mm'),
                        # heatmap_height = unit(panel.figure.height/2, 'mm'),
                        # column_dend_height = unit(3, 'mm'),
                        # row_dend_width = unit(4.5, 'mm'),
                        height = unit((panel.figure.height/2)/1.325, 'mm'),
                        width = unit((panel.figure.width*4)/1.645, 'mm')) #-> toil_hm_plt

size = calc_ht_size(toil_hm_plt)
size

pdf(file = file.path(figure.dir, 'fig.1', 'toil_hm.pdf'),
    width=unit(size[1], 'mm'), height=unit(size[2], 'mm'), family = 'Helvetica')

toil_hm_plt

dev.off()

# save_pheatmap_pdf(x = toil_hm, filename = 'test_hm.pdf', height = panel.figure.height/2, width = panel.figure.width)


working_toil_counts %>% 
  filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  ggplot(aes(KRAS, count)) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_point() + 
  facet_wrap(~gene) + 
  stat_compare_means(comparisons = list(c('p.G12D', 'WT_normal')), method = 'wilcox')
```

```{r}
myc.avg.counts.tidy.stratified <-
  norm.counts.final %>%
  filter(sample_type == 'Primary Tumor') %>%
  # filter(gene %in% tmp.df$gene) %>%
  filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'toil_ifn') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum)

surv.obj <- Surv(time = filter(luad.phenotypes.merge %>% select(sample, OS.time, OS) %>% distinct() , 
                               sample %in% myc.avg.counts.tidy.stratified$sample)$OS.time,
                 event = filter(luad.phenotypes.merge %>% select(sample, OS.time, OS) %>% distinct(), 
                                sample %in% myc.avg.counts.tidy.stratified$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ myc.avg.counts.tidy.stratified$toil_ifn), 
                    data = myc.avg.counts.tidy.stratified)
ggsurvplot(surv.fit,
           pval = T, 
           legend.title = "",
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)])
surv.plt

```


## znf
```{r}
salmon.quant$aale$intra$deseq$de %>% 
  dplyr::rename('AALE' = log2FoldChange) %>% 
  filter(padj < 0.05,
         gene %in% meta.data$trono.krab.znfs$gene) %>% 
  merge(salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
          dplyr::rename('HBEC' = log2FoldChange) %>% 
          filter(padj < 0.05) %>% 
          mutate(HBEC = HBEC), by = 'gene') %>% 
  merge(meta.data$trono.krab.znfs, by = 'gene') %>% View()
  ggplot(aes(AALE, HBEC, color = ZNF)) + 
  geom_point() + 
  scale_color_viridis_d() +
  # scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
  geom_text_repel(aes(label = ifelse(AALE < -1 & HBEC < -0.35, gene, '')), show.legend = F) + 
  theme(legend.position = c(1, 1))

panel.save(figure = 5, name = 'de_scatter_znf', width = 'single', height = 'single')

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -1, gene %in% meta.data$trono.krab.znfs$gene) %>%   #, gene %in% uniq_ctrl_peak_genes$`Gene Name`) %>% 
  # filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj < 0.05, log2FoldChange > 0)$gene) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -0.5, grepl('^HOX|^KRT', gene) | gene %in% meta.data$trono.krab.znfs$gene) %>% 
  # filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj < 0.05, log2FoldChange > 0)$gene) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list

salmon.quant$aale$intra$deseq$de %>% 
  # filter(gene == 'ZNF563') %>% 
  filter(padj < 0.05 , log2FoldChange > 2) %>% 
  # filter(gene %in% meta.data$trono.krab.znfs$gene,
  #        !gene_id %in% de_list) %>% 
  # filter(gene %in% filter(salmon.quant$hbec$lacz_v_v12$deseq$de, padj < 0.05, log2FoldChange > 0)$gene) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list_w_hbec


# salmon.quant$hbec$lacz_v_v12$deseq$de %>% 
#   filter(padj < 0.05, log2FoldChange > 0) %>% #, gene %in% c(ifn_g, ifn_a)) %>% 
#   filter(!gene %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 0)$gene) %>% 
#   select(gene_id) %>% unlist() %>% unname() -> de_list_w_hbec

gr <- rowRanges(salmon.quant$gxi)

znf_de_gr <- gr[names(gr) %in% de_list]
znf_de_gr_hbec <- gr[names(gr) %in% de_list_w_hbec]

znf_promoter_seqs <- 
  GenomicRanges::promoters(znf_de_gr, upstream = 1000, downstream = 200) %>% 
  memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38)

znf_promoter_seqs_hbec <- 
  GenomicRanges::promoters(znf_de_gr_hbec, upstream = 1000, downstream = 200) %>% 
  memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38)

znf_uniq_peaks_seqs <- 
  memes::get_sequence(regions = ctrl_znf_uniq_peak_ranges, genome = BSgenome.Hsapiens.UCSC.hg38)

znf_ame_out <- memes::runAme(znf_promoter_seqs, 
                             control = znf_promoter_seqs_hbec)

znf_ame_out %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> znf_ame_out

# znf_ame_out %>% 
#   filter(adj.pvalue < 0.05) %>% 
#   separate(motif_id, sep = '[.]', into = c('TF', NA, NA, NA)) -> znf_ame_out

salmon.quant$aale$intra$deseq$de %>% 
  merge(znf_ame_out, by.x = 'gene', by.y = 'TF') %>% 
  # filter(gene %in% ame_tf) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 0.3, gene, '')), color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99, 0.99)) 

panel.save(figure = 6, name = 'tf_enrich_volcano', width = 'single', height = 'single')

consensus_atac_boolean_anno %>% 
  filter(kras_R1.mLb.clN.bool == F) %>%
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, 
                                 log2FoldChange < -0.5)$gene) %>%
  filter(`Gene Name` %in% meta.data$trono.krab.znfs$gene) %>%
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  # filter(annotation == "promoter-TSS ") %>%
  # filter(chr != 'chr19') %>% 
  mutate(strand = '+') %>%
  select(chr, start, end, strand, name = `Gene Name`) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> ctrl_znf_uniq_peak_ranges


alx1_peak <- memes::get_sequence(regions = 'chr12:85280200-85280907', genome = BSgenome.Hsapiens.UCSC.hg38)


etv1_motif <- MotifDb::query(MotifDb::MotifDb, c('NFYC_HUMAN')) %>% 
  universalmotif::convert_motifs()


etv1_fimo_results <- memes::runFimo(znf_uniq_peaks_seqs, etv1_motif) 
etv1_fimo_results
# %>%
#   as.data.frame() %>%
#   merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id')
 
tomTom_res <- memes::runAme(control = 'shuffle', input = alx1_peak)


ctrl_znf_uniq_peak_ranges %>% 
  resize(661, 'center') %>% 
  memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38) %>% 
  memes::runDreme('shuffle') -> tmp

tmp$sites_hits

meme_res <- memes::runDreme(input = znf_uniq_peaks_seqs)
meme_res %>% 
  memes::runTomTom()


create_motif('TACAAAAW', altname = 'test', type = 'PCM') %>% memes::runTomTom()

znf_tomTom_res <- memes::runAme(control = 'shuffle',
                            input = promoter_seqs, 
                            database = 
                              file.path(output.data.dir, 
                                        'reference.data', 
                                        'Hughes_ZNF_motifs.meme'))

memes::runFimo(promoter_seqs, motifs = fos_motif) %>% as.data.frame() %>% View()

tomTom_res %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ame_tf

salmon.quant$aale$intra$deseq$de %>% 
  merge(ame_tf, by.x = 'gene', by.y = 'TF') %>% 
  # filter(gene %in% ame_tf) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.35, 0.99)) 


znf563_motif <- MotifDb::query(MotifDb::MotifDb, 'NFYC_HUMAN.H11MO.0.A') %>% 
  universalmotif::convert_motifs()

znf563_fimo_results <- memes::runFimo(promoter_seqs, znf563_motif, thresh = 1e-3) %>% 
  as.data.frame() %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'seqnames', by.y = 'gene_id')





consensus_atac_boolean_anno %>% 
  # filter(kras_R1.mLb.clN.bool == F) %>% 
  filter(grepl('IRF9', `Gene Name`)) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  mutate(strand = '+') %>% 
  select(chr, start, end, strand, name = `Gene Name`) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> irf9_peak

names(irf9_peak) <- 'IRF9'

getSeq(BSgenome.Hsapiens.UCSC.hg38, irf9_peak) -> irf9_peak_seq

memes::runAme(input = irf9_peak_seq, control = other_uniq_seqs) -> irf9_motifs
memes::runDreme(input = ifn_uniq_seqs, control = 'shuffle')


meme_in <- memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38, regions = ctrl_znf_uniq_peak_ranges)

ctrl_meme_in <- memes::get_sequence(genome = BSgenome.Hsapiens.UCSC.hg38, regions = other_uniq_peak_ranges)
names(meme_in) <- ctrl_znf_uniq_peak$name

memes::runAme(input = meme_in, control = 'shuffle') -> ame_out


ame_out %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ame_tf

salmon.quant$aale$intra$deseq$de %>% 
  merge(ame_tf, by.x = 'gene', by.y = 'TF') %>% 
  # filter(gene %in% ame_tf) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') #+ 
  theme(legend.position = c(0.35, 0.99)) 

names(ifn_uniq_peak_ranges) <- uniq_peak_genes$`Gene Name`

consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == F) %>% 
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, log2FoldChange > 0)$gene) %>%
  filter(`Gene Name` %in% meta.data$trono.krab.znfs$gene) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  mutate(strand = '+') %>% 
  select(chr, start, end, strand, name = `Gene Name`) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> other_uniq_peak_ranges

names(other_uniq_peak_ranges) <- consensus_atac_boolean_anno %>% 
                                  filter(ctrl_R1.mLb.clN.bool == F) %>% 
                                  filter(`Gene Name` %in%
                                           filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 1.5)$gene) %>%
                                  filter(!`Gene Name` %in% c(ifn_g, ifn_a), ! is.na(`Gene Name`)) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  select(`Gene Name`) %>% unlist() %>% unname()

getSeq(BSgenome.Hsapiens.UCSC.hg38, ifn_uniq_peak_ranges) -> ifn_uniq_seqs
getSeq(BSgenome.Hsapiens.UCSC.hg38, other_uniq_peak_ranges) -> other_uniq_seqs

memes::runAme(input = ifn_uniq_seqs, control = other_uniq_seqs) -> ifn_uniq_motifs

gr <- rowRanges(salmon.quant$gxi)

salmon.quant$aale$intra$deseq$de %>% 
  filter(gene %in% meta.data$trono.krab.znfs$gene,
         padj < 0.05, log2FoldChange < -4.5) %>% 
  select(gene_id) %>% unlist() %>% unname() -> znf_de_list

# merge(ctrl_znf_uniq_peak, salmon.quant$aale$intra$deseq$de, ctrl_znf_uniq_peak, by.y = 'gene', by.x = 'name') %>% 
#   filter(log2FoldChange < 0, padj < 0.05) -> ctrl_znf_uniq_peak

znf_de_gr <- gr[names(gr) %in% znf_de_list]

de_promoter_regions <- promoters(znf_de_gr, upstream = 300, downstream = 300)

bamsignals::bamCoverage(kras_bam_path, de_promoter_regions, paired.end = 'extend') -> kras_de_prom_profile
bamsignals::bamCoverage(ctrl_bam_path, de_promoter_regions, paired.end = 'extend') -> ctrl_de_prom_profile

# memes::runFimo(sequences = ctrl_znf_uniq_peak, motifs = fos_motif)


bamsignals::alignSignals(kras_de_prom_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(position = position - 300,
         cov = (cov/kras_norm_factor) * 1e6,
         condition = 'kras',
         context = 'de') -> kras_de_prom_df

bamsignals::alignSignals(ctrl_de_prom_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(position = position - 300,
         cov = (cov/ctrl_norm_factor) * 1e6,
         condition = 'ctrl',
         context = 'de') -> ctrl_de_prom_df
  
rbind(kras_de_prom_df, ctrl_de_prom_df) %>%
  ggplot(aes(position, cov, group = condition, color = condition)) +
    stat_summary(geom="ribbon", fun.data=mean_cl_normal, width=0.1, conf.int=0.95, 
                 fill='grey', alpha = 0.1, size = rel(0.05), show.legend = F)+
    # stat_summary(geom="line", fun=mean, linetype="dashed", alpha = 0.15, size = rel(0.05)) +
    stat_summary(geom="point", fun=mean, alpha = 0.25, size = rel(1)) +
    scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
    xlab('distance to TSS') +
    ylab('mean CPM (95% CI)') + 
    xlim(-300, 300) + 
    theme(legend.position = c(0.99, 0.99))

panel.save(figure = 6, name = 'atac_coverage_znf', width = 'single', height = 'single')



 norm.counts.final %>% 
  filter(sample != 'TCGA-44-3917-01') %>% 
  filter(KRAS %in% c('p.G12D', 'WT_normal')) -> working_toil_counts

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -1, gene %in% meta.data$trono.krab.znfs$gene) -> aale_znf

working_toil_counts %>% 
  # filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  # filter(gene %in% uniq_peak_genes$`Gene Name`) %>% 
  filter(gene %in% aale_znf$gene) %>%
  # filter(gene %in% c('ETV1', 'ALX1', 'ZNF90', 'ZNF736', 'ZNF682', 'MEF2B'), KRAS == 'p.G12D') %>% 
  group_by(gene) %>%
  mutate(count = scale(log1p(count), center = T)) %>%
  select(sample, gene, count) %>%
  pivot_wider(names_from = gene, values_from = count) %>%
  distinct() %>% 
  column_to_rownames('sample') %>% 
  t() -> toil_hm_counts

working_toil_counts %>% 
  select(sample, 'G12D' = KRAS) %>%
  distinct() %>% 
  column_to_rownames('sample') %>% 
  ComplexHeatmap::rowAnnotation(df = . , 
                                 col = list(G12D = c('WT_normal' = 'white', 
                                                     'p.G12D' = viridis(3)[2])),
                                 show_legend = F,
                                  annotation_name_side = 'bottom',
                                  simple_anno_size = unit(2.5, "mm")) -> toil_sample_anno

# working_toil_counts %>% 
#   select(sample, 'G12' = KRAS) %>%
#   distinct() %>% 
#   column_to_rownames('sample') %>% 
#   ComplexHeatmap::rowAnnotation(df = . , 
#                                    col = list(G12D = c('WT_normal' = 'white', 
#                                                        'p.G12' = viridis(3)[2])),
#                                    show_legend = F,
#                                 annotation_name_side = 'bottom',
#                                 simple_anno_size = unit(2.5, "mm")) -> toil_sample_anno



working_toil_counts %>% 
  # filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  filter(gene %in% aale_znf$gene) %>%
  # filter(gene %in% aale_ifn$gene) %>%
  select(gene) %>% 
  distinct() %>% 
  mutate(unique = ifelse(gene %in% ctrl_znf_uniq_peak$name, T, F)) %>% 
  column_to_rownames('gene') %>% 
  ComplexHeatmap::columnAnnotation(df =.,
                                col = list(unique = c('TRUE' = viridis(2)[1], 'FALSE' = 'white')),
                                show_legend = F,
                                annotation_name_side = 'right',
                                simple_anno_size = unit(2.5, "mm")) -> gene_motif_anno




ComplexHeatmap::Heatmap(t(toil_hm_counts), 
                        name = 'z-score', 
                        show_row_names = F,
                        # left_annotation = toil_sample_anno,
                        # bottom_annotation = gene_motif_anno,
                        # heatmap_width = unit(panel.figure.width, 'mm'),
                        # heatmap_height = unit(panel.figure.height/2, 'mm'),
                        # column_dend_height = unit(3, 'mm'),
                        # row_dend_width = unit(4.5, 'mm'),
                        height = unit((panel.figure.height/3)/2.18, 'mm'),
                        width = unit((panel.figure.width*3)/1.81, 'mm')) -> toil_hm_plt

size = calc_ht_size(toil_hm_plt)
size

pdf(file = file.path(figure.dir, 'fig.5', 'toil_znf_hm.pdf'),
    width=unit(size[1], 'mm'), height=unit(size[2], 'mm'), family = 'Helvetica')

toil_hm_plt

dev.off()

working_toil_counts %>% 
  filter(KRAS %in% c('p.G12D', 'WT_normal')) %>% 
  mutate(gene_set = ifelse(gene %in% aale_znf$gene, 'dn_ZNF', 
                           ifelse(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 
                                              'IFITM1', 'OAS1', 'MOV10', 'OAS3'), 'up_IFN', 'other'))) %>% 
  filter(gene_set != 'other') -> toil_dist_counts



toil_dist_counts %>% 
  group_by(sample, KRAS, gene_set) %>% 
  summarize(exp_mean = mean(count)) %>% 
  pivot_wider(names_from = gene_set, values_from = exp_mean) %>% 
  ggplot(aes(dn_ZNF, up_IFN, color = KRAS)) + 
  geom_point()
  ggplot(aes(KRAS, znf_mean)) + 
  geom_jitter(width = 0.15) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  stat_compare_means(method = 'wilcox.test', comparisons = list(c('p.G12D', 'WT_normal'))) +
  ylab('mean(dn ZNF counts)') + xlab('')

panel.save(figure = 7, name = 'avg_znf_wilcox', width = 'single', height = 'single')


toil_dist_counts %>% 
  ggplot(aes(KRAS, count)) + 
  geom_jitter(alpha =0.4) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  facet_row(~gene) + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        legend.position = c(0.99,0.99)) + 
  stat_compare_means(method = 'wilcox.test', comparisons = list(c('p.G12D', 'WT_normal')), hide.ns = F, label = 'p.signif') + 
  xlab('')




myc.avg.counts.tidy.stratified <-
  norm.counts.final %>%
  filter(sample_type == 'Primary Tumor') %>%
  # filter(gene %in% aale_znf$gene) %>%
  filter(gene %in% ctrl_znf_uniq_peak$name) %>%
  # filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'toil_znf') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum)

surv.obj <- Surv(time = filter(luad.phenotypes.merge %>% select(sample, OS.time, OS) %>% distinct() , 
                               sample %in% myc.avg.counts.tidy.stratified$sample)$OS.time,
                 event = filter(luad.phenotypes.merge %>% select(sample, OS.time, OS) %>% distinct(), 
                                sample %in% myc.avg.counts.tidy.stratified$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ myc.avg.counts.tidy.stratified$toil_znf), 
                    data = myc.avg.counts.tidy.stratified)
ggsurvplot(surv.fit,
           pval = T, 
           legend.title = "",
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)]) -> surv.plt

surv.plt

panel.save(plot.in = surv.plt$plot, figure = 5, name = 'dn_znf_km', width = 'single', height = 'single')


# save_pheatmap_pdf(x = toil_hm, filename = 'test_hm.pdf', height = panel.figure.height/2, width = panel.figure.width)


working_toil_counts %>% 
  filter(gene %in% c('ISG20', 'ISG15', 'MX2', 'CFB', 'IFITM1', 'OAS1', 'MOV10', 'OAS3')) %>%
  ggplot(aes(KRAS, count)) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_point() + 
  facet_wrap(~gene) + 
  stat_compare_means(comparisons = list(c('p.G12D', 'WT_normal')), method = 'wilcox')


```

###A549
```{r}

salmon.quant$all_intra$deseq$counts %>% 
  select(gene, contains('aale'), contains('a549')) %>% 
  filter(gene %in% aale_ifn$gene) %>% 
  gather(sample, count, -gene) %>% 
  separate(sample, sep = '[.]', into = c(NA, 'cell_line', NA, 'condition')) %>% 
  group_by(condition, cell_line, gene) %>% 
  summarize(count = mean(count)) %>% 
  ggplot(aes(condition, log1p(count), label = gene)) + 
  geom_boxplot() +
  geom_point() +
  geom_text_repel(aes(label = ifelse(condition %in% c('Empty', 'ctrl') & log1p(count > 0.1), gene, '')), max.overlaps = 12) +
  geom_line(aes(group = gene), alpha = 0.3) +
  facet_row(~cell_line, scales = 'free_x')

## -----------------------------------------------------------------------------
a549_de$de_682 %>% 
  mutate(oe = 'ZNF682') %>% 
  rbind(a549_de$de_257 %>% 
    mutate(oe = 'ZNF257')) %>% 
  merge(salmon.quant$aale$intra$deseq$de, by = 'gene') %>% 
  filter(gene %in% c(ifn_a, ifn_g),
         padj.x < 0.05) %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, label = gene)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange.x < -0.1 & log2FoldChange.y > 0.1, gene, '')), max.overlaps = 5) +
  facet_col(~oe) +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  geom_hline(yintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  ylab('AALE log2FC') + xlab('A549 log2FC')


panel.save(figure = 5, name = 'facet_oe_a549_volcano', width = 'single', height = 'medium')

a549_te_de$de_682 %>% 
  mutate(oe = 'ZNF682') %>% 
  rbind(a549_te_de$de_257 %>% 
    mutate(oe = 'ZNF257')) %>% 
  filter(!gene_id %in% meta.data$gencode.35.tx.to.gene$gene,
         !grepl(')n', gene_id),
         padj < 0.05) %>% 
  merge(meta.data$te.families.clades, by.x = 'gene_id', by.y = 'gene') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene_id, color = clade)) + 
  geom_point(alpha = 0.6) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.5) +
  scale_color_viridis_d(direction = -1) +
  ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange < -0.5, gene_id, '')), 
                           show.legend = F) +
  facet_col(~oe) + 
  theme(legend.position = c(0.3, 0.99))

panel.save(figure = 5, name = 'facet_te_a549_volcano', width = 'single', height = 'single')


salmon.quant$aale$intra$deseq$de %>% 
  mutate(padj = ifelse(padj == 0, 
                       min(filter(salmon.quant$aale$intra$deseq$de, padj != 0)$padj), padj)) %>% 
  mutate(gene_set = ifelse(gene %in% ifn_gene_list,'IFN', 'other'),
         gene_set = ifelse(gene %in% msig.df$HALLMARK_KRAS_SIGNALING_UP, 'KRAS', gene_set),
         gene_set = ifelse(gene %in% meta.data$trono.krab.znfs$gene, 'ZNF', gene_set)) %>% 
  mutate(gene_set = factor(gene_set, levels = c('KRAS','IFN','ZNF', 'other'))) %>% 
  filter(padj < 0.05) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), label = gene)) + 
  geom_point(alpha = 0.6) + 
  ggrepel::geom_text_repel(aes(label = ifelse(abs(log2FoldChange) > 5 & gene_set != 'other' | 
                                                padj < 0.0001 & gene_set != 'other', gene, ''))) +
  facet_col(~gene_set, scales = 'free_y') +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed')

## -----------------------------------------------------------------------------
  


ucsc.rmsk.salmon.quant$all_intra$deseq$counts %>% 
  select(gene, contains('aale'), contains('a549')) %>% 
  merge(meta.data$te.families.clades, by = 'gene') %>%
  filter(clade %in% c('DNA', "LINE",'SINE', 'LTR')) %>% 
  gather(sample, count, -gene, -clade, -family) %>% 
  separate(sample, sep = '[.]', into = c(NA, 'cell_line', NA, 'condition')) %>% 
  group_by(condition, cell_line, clade, gene) %>% 
  summarize(count = mean(count)) %>% 
  ggplot(aes(condition, count)) + 
  geom_boxplot() + 
  facet_row(~cell_line+clade, scales = 'free_x')

ggplot(ucsc.rmsk.salmon.quant$a549$deseq$de %>% 
         merge(salmon.quant$aale$deseq$de, by = 'gene') %>% 
         filter(gene %in% gencode.35.lnc.gene.names$gene), 
       aes(log2FoldChange.x,log2FoldChange.y, label = gene)) + 
  geom_point() + 
  geom_hline(yintercept = 0, linetype = 'dotted') + 
  geom_vline(xintercept = 0, linetype = 'dotted') + 
  ggrepel::geom_text_repel()


ucsc.rmsk.salmon.quant$a549$deseq$counts %>% 
  filter(grepl('^Alu', gene)) %>% 
  gather(sample, counts)
  column_to_rownames('gene') %>% 
  pheatmap()
```


## ATAC
```{r}
kras_atac_anno_peaks <- 
  read_tsv(file.path(output.data.dir, 
                     'narrow_output.data/bwa/mergedLibrary/macs/narrowPeak/kras_R1.mLb.clN_peaks.annotatePeaks.txt'))

ctrl_atac_anno_peaks <- 
  read_tsv(file.path(output.data.dir, 
                     'narrow_output.data/bwa/mergedLibrary/macs/narrowPeak/ctrl_R1.mLb.clN_peaks.annotatePeaks.txt'))

consensus_atac_anno_peak_counts <- 
  read_tsv(file.path(output.data.dir, 
                     'narrow_output.data/bwa/mergedLibrary/macs/narrowPeak/consensus/consensus_peaks.mLb.clN.featureCounts.txt'),
           skip = 1)

consensus_atac_boolean_anno <- 
  read_tsv(file.path(output.data.dir, 
                     'narrow_output.data/bwa/mergedLibrary/macs/narrowPeak/consensus/consensus_peaks.mLb.clN.boolean.annotatePeaks.txt'
                     ))


irf9_fimo_results %>% 
  mutate(start.x = start.y + start.x,
         end.x = start.y + end.x) %>% 
  select(chr, start = start.x, end = end.x, strand = strand.x, gene, motif_id) %>% 
  unite(col = 'name', sep = '_', c(gene, motif_id)) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = T) -> irf9_bed

consensus_atac_anno_peak_counts %>% 
  select(chr = Chr, start = Start, end = End, strand = Strand, name = Geneid) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = T) -> consensus_bed

consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == F) %>% 
  select(chr, start, end, name = interval_id) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = T) -> kras_atac_bed

consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == F) %>% 
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 1.5)$gene) %>%
  # filter(`Gene Name` %in% c(ifn_g, ifn_a)) %>%
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>%
  # group_by(annotation) %>%
  group_by(`Gene Name`) %>%
  tally() -> uniq_kras_peak_genes#%>% filter(grepl('ZNF', `Gene Name`)) %>% 
  # arrange(-n)

consensus_atac_boolean_anno %>% 
  # filter(ctrl_R1.mLb.clN.bool == T & kras_R1.mLb.clN.bool == T) %>% 
  # filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 1.5)$gene) %>%
  # filter(`Gene Name` %in% c(ifn_g, ifn_a)) %>%
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  # filter(annotation == "promoter-TSS ") %>%
  select(chr, start, end, gene = `Gene Name`) %>% 
  mutate(strand = '+') -> bg_peak_input_bed

consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == F) %>% 
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 1.5)$gene) %>%
  # filter(`Gene Name` %in% c(ifn_g, ifn_a)) %>%
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  # filter(annotation == "promoter-TSS ") %>%
  select(chr, start, end, gene = `Gene Name`) %>% 
  mutate(strand = '+') -> uniq_kras_peak_input_bed

job <- rGREAT::submitGreatJob(gr = uniq_kras_peak_input_bed, species = 'hg38', bg = bg_peak_input_bed)

tb <- rGREAT::getEnrichmentTables(job, download_by = 'tsv')

base::do.call(rbind, tb) %>% 
  group_by(Ontology) %>% 
  top_n(30, -HyperFdrQ) %>%
  filter(HyperFdrQ < 0.05,
         NumFgGenesHit > 1) %>% 
  ggplot(aes(NumFgGenesHit, -log10(HyperFdrQ), label = Desc)) + 
  geom_point() + 
  # scale_color_viridis_d() +
  ggrepel::geom_text_repel(show.legend = F, color = 'black') +
  theme(legend.position = c(0.99, 0.99)) + 
  xlab('observed genes') +
  facet_row(~Ontology, scales = 'free')

panel.save(figure = 2, 
           name = 'atac_GREAT_enrichment_facet', width = 'double', height = 'single')


consensus_atac_boolean_anno %>% 
  filter(kras_R1.mLb.clN.bool == F) %>% 
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange < -1.5)$gene) %>%
  # filter(`Gene Name` %in% meta.data$trono.krab.znfs$gene) %>%
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  # filter(annotation == "promoter-TSS ") %>%
  select(chr, start, end, gene = `Gene Name`) %>% 
  mutate(strand = '+') -> uniq_ctrl_peak_input_bed

ctrl_job <- rGREAT::submitGreatJob(uniq_ctrl_peak_input_bed, species = 'hg38', bg = bg_peak_input_bed)

ctrl_tb <- rGREAT::getEnrichmentTables(ctrl_job, download_by = 'tsv')
  
base::do.call(rbind, ctrl_tb) %>% 
  group_by(Ontology) %>% 
  top_n(30, -HyperFdrQ) %>% 
  filter(HyperFdrQ < 0.05,
         NumFgGenesHit > 1) %>% 
  ggplot(aes(NumFgGenesHit, -log10(HyperFdrQ), label = Desc)) + 
  geom_point() + 
  # scale_color_viridis_d() +
  ggrepel::geom_text_repel(show.legend = F, color = 'black', max.overlaps = 2) +
  theme(legend.position = c(0.99, 0.99)) + 
  xlab('observed genes') +
  facet_row(~Ontology, scales = 'free')
  
panel.save(figure = 6, 
           name = 'atac_GREAT_enrichment_facet', width = 'single', height = 'single')


consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == F) %>% 
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 1.5)$gene) %>%
  filter(`Gene Name` %in% c(ifn_g, ifn_a)) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  mutate(strand = '+') %>% 
  select(chr, start, end, strand, name = `Gene Name`) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> ifn_uniq_peak_ranges

names(ifn_uniq_peak_ranges) <- uniq_peak_genes$`Gene Name`

consensus_atac_boolean_anno %>% 
  filter(ctrl_R1.mLb.clN.bool == T) %>%
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 0.5)$gene) %>%
  filter(`Gene Name` %in% c(ifn_g, ifn_a), ! is.na(`Gene Name`), ! `Gene Name` %in% names(ifn_uniq_peak_ranges)) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  mutate(strand = '+') %>% 
  select(chr, start, end, strand, name = `Gene Name`) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> other_uniq_peak_ranges

names(other_uniq_peak_ranges) <- consensus_atac_boolean_anno %>% 
                                    filter(ctrl_R1.mLb.clN.bool == T) %>%
  filter(`Gene Name` %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, log2FoldChange > 0.5)$gene) %>%
  filter(`Gene Name` %in% c(ifn_g, ifn_a), ! is.na(`Gene Name`), ! `Gene Name` %in% names(ifn_uniq_peak_ranges)) %>% 
  separate(Annotation, sep = '\\(', into = c('annotation', NA)) %>% 
  filter(annotation == "promoter-TSS ") %>% 
  select(`Gene Name`) %>% unlist() %>% unname()

getSeq(BSgenome.Hsapiens.UCSC.hg38, ifn_uniq_peak_ranges) -> ifn_uniq_seqs
getSeq(BSgenome.Hsapiens.UCSC.hg38, other_uniq_peak_ranges) -> other_uniq_seqs

memes::runAme(input = ifn_uniq_seqs, control = other_uniq_seqs) -> ifn_uniq_motifs

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, abs(log2FoldChange) > 1) %>%
  select(gene_id) %>% unlist() %>% unname() -> de_list

de_list_ranges <- gr[names(gr) %in% de_list]

names(de_list_ranges) <- paste0(seqnames(de_list_ranges), '_', names(de_list_ranges))

de_list_promoter_seqs <-
  GenomicFeatures::getPromoterSeq(subject = BSgenome.Hsapiens.UCSC.hg38, 
                                  query = de_list_ranges, upstream = 1000, downstream = 500)

lapply(ifn_uniq_motifs$motif_id, function(motif) {
  
  print(motif)
  
  this_motif <- MotifDb::query(MotifDb::MotifDb, motif) %>% 
                  universalmotif::convert_motifs()
  
  memes::runFimo(de_list_promoter_seqs, motifs = this_motif) -> de_list_promoter_regions
  
  bamsignals::bamCount(kras_bam_path,
                          de_list_promoter_regions) -> kras_de_prom_profile

  bamsignals::bamCount(ctrl_bam_path,
                          de_list_promoter_regions) -> ctrl_de_prom_profile

  lst('kras' = kras_de_prom_profile,
      'ctrl' = ctrl_de_prom_profile)
  

}) -> de_list_fimo_out

de_list_fimo_out[[4]]$kras




meta.data$ucsc.rmsk.tx.to.gene.fmt %>% 
  select(enst, gene, biotype) %>% 
  separate(enst, sep = '_', into = c(NA, NA, 'gene', 'position', NA, NA, 'strand', NA)) %>% 
  mutate(position = sub('range=', '', position),
         strand = sub('strand=', '', strand)) -> rmsk_bed

rmsk_bed %>% 
  filter(grepl(':', position)) %>% 
  separate(position, sep = ':', into = c('chr', 'range')) %>% 
  separate(range, sep = '-', into = c('start', 'end')) -> rmsk_parsed_bed



imap(list('aale_intra' = ucsc.rmsk.salmon.quant$aale$intra$deseq$de
          # 'h358_intra' = ucsc.rmsk.salmon.quant$hbec$lacz_v_v12$deseq$de,
          # 'aale_exo' = ucsc.rmsk.salmon.quant$aale$exo$deseq$de), function(x, x_name) {
         ), function(x, x_name) {

           x %>% 
             filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
                    padj < 0.05, log2FoldChange > 0.65, !grepl('\\)n', gene)) %>% 
             mutate(context = x_name)
         }) %>% bind_rows() -> upreg_te

imap(list('aale_intra' = ucsc.rmsk.salmon.quant$aale$intra$deseq$de,
          'h358_intra' = ucsc.rmsk.salmon.quant$hbec$lacz_v_v12$deseq$de,
          'aale_exo' = ucsc.rmsk.salmon.quant$aale$exo$deseq$de), function(x, x_name) {
           x %>% 
             filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
                    padj < 0.05, log2FoldChange < -1, !grepl('\\)n', gene)) %>% 
             mutate(context = x_name)
         }) %>% bind_rows() -> dnreg_te

salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange > 1.5) %>% 
  select(gene_id) %>% unlist() %>% unname() -> de_list

gr <- rowRanges(salmon.quant$gxi)

ifn_de_gr <- gr[names(gr) %in% de_list]

de_promoter_regions <- promoters(ifn_de_gr, upstream = 1000, downstream = 200)

getSeq(BSgenome.Hsapiens.UCSC.hg38, de_promoter_regions) -> upreg_gencode_seqs


salmon.quant$aale$intra$deseq$de %>% 
  filter(padj < 0.05, log2FoldChange < -1.5,
         !gene %in% meta.data$trono.krab.znfs$gene) %>%
  select(gene_id) %>% unlist() %>% unname() -> dn_de_list

gr <- rowRanges(salmon.quant$gxi)

ifn_de_gr <- gr[names(gr) %in% de_list]

de_promoter_regions <- promoters(ifn_de_gr, upstream = 1000, downstream = 200)

dn_gr <- gr[names(gr) %in% dn_de_list]

dn_promoter_regions <- promoters(dn_gr, upstream = 1000, downstream = 200)

getSeq(BSgenome.Hsapiens.UCSC.hg38, de_promoter_regions) -> upreg_gencode_seqs

getSeq(BSgenome.Hsapiens.UCSC.hg38, dn_promoter_regions) -> dnreg_gencode_seqs


# lapply(1:length(te_ranges_list), function(te) findOverlaps(de_promoter_regions, te_ranges_list[[te]], minoverlap = 0.75))
# 
# te_ranges_list[[2]][822]
# de_promoter_regions[9]

upreg_te %>% 
  select(gene) %>% 
  distinct() %>% 
  unlist() %>% 
  unname() %>% 
  lapply(function(te) {

    rmsk_parsed_bed %>% 
      filter(gene %in% upreg_te$gene,
             gene == te) %>% 
      GenomicRanges::makeGRangesFromDataFrame() -> te_ranges
    
    names(te_ranges) <- rep_len(te, length(te_ranges))
    
    te_ranges
    
  }) -> te_ranges_list

dnreg_te %>% 
  select(gene) %>% 
  distinct() %>% 
  unlist() %>% 
  unname() %>% 
  lapply(function(te) {

    rmsk_parsed_bed %>% 
      filter(gene %in% dnreg_te$gene,
             gene == te) %>% 
      GenomicRanges::makeGRangesFromDataFrame() -> te_ranges
    
    names(te_ranges) <- rep_len(te, length(te_ranges))
    
    te_ranges
    
  }) -> dn_te_ranges_list

width(dn_te_ranges_list[[9]])

median(width(dn_te_ranges_list[[7]]))

dn_te_ranges_list[[9]][width(dn_te_ranges_list[[9]]) > 10, ] -> te_ranges_tmp

resize(te_ranges_tmp, width = median(max(dn_te_ranges_list[[9]]))) -> te_ranges_tmp

bamsignals::bamCoverage(kras_bam_path, 
                        te_ranges_tmp, paired.end = 'extend') -> kras_de_te_profile
bamsignals::bamCoverage(ctrl_bam_path, 
                        te_ranges_tmp, paired.end = 'extend') -> ctrl_de_te_profile

bamsignals::alignSignals(kras_de_te_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(cov = (cov/kras_norm_factor) * 1e6,
         condition = 'kras',
         context = 'de') -> kras_te_prom_df

bamsignals::alignSignals(ctrl_de_te_profile) %>% 
  as.data.frame() %>% 
  mutate(position = row_number()) %>% 
  gather(prom, cov, -position) %>%
  mutate(cov = (cov/ctrl_norm_factor) * 1e6,
         condition = 'ctrl',
         context = 'de') -> ctrl_te_prom_df
  
rbind(kras_te_prom_df, ctrl_te_prom_df) %>%
  ggplot(aes(position, cov, group = condition, color = condition)) +
    stat_summary(geom="ribbon", fun.data=mean_cl_normal, width=0.1, conf.int=0.95, 
                 fill='grey', alpha = 0.1, size = rel(0.05), show.legend = F)+
    # stat_summary(geom="line", fun=mean, linetype="dashed", alpha = 0.15, size = rel(0.05)) +
    stat_summary(geom="point", fun=mean, alpha = 0.25, size = rel(1)) +
    scale_color_manual(values = c(viridis(3)[1], viridis(3)[2])) +
    ylab('mean CPM (95% CI)') + 
    theme(legend.position = c(0.3, 0.99))

do.call("c", dn_te_ranges_list) %>% 
  getSeq(BSgenome.Hsapiens.UCSC.hg38, .) %>% length()

te_ranges_list %>% 
  lapply(function(te_ranges) {
    
    
    if(length(te_ranges) > 0 & length(te_ranges) < 4000) {
      
      names(te_ranges) %>% unique() -> name
      
      print(name)
      print(length(te_ranges))
      
      getSeq(BSgenome.Hsapiens.UCSC.hg38, te_ranges) -> te_seqs

      memes::runAme(te_seqs, control = 'shuffle', 
                    database = file.path(output.data.dir, 'reference.data', 
                            'Hughes_ZNF_motifs.meme')) -> te_motif_enrich

      te_motif_enrich %>%
        mutate(tmp = motif_id,
               TE = name) %>%
        separate(tmp, sep = '_', into = c('TF', NA)) -> ame_tf

      ame_tf
    
    }
  }) -> ame_tf_list


ame_tf_list %>% 
  bind_rows() %>% 
  filter(adj.pvalue < 0.05) %>% 
  separate(motif_id, sep = '[.]', into = c('TF', NA, NA, NA)) %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'TF', by.y = 'gene') %>%
  merge(meta.data$te.families.clades, by.x = 'TE', by.y = 'gene') %>% 
  merge(upreg_te, by.x = 'TE', by.y = 'gene') -> tmp

tmp %>% 
  filter(padj.x < 0.05) %>% 
  # ggplot(aes(reorder(TE, -log2FoldChange.y), reorder(TF, -log2FoldChange.x), fill = log(evalue))) + 
  # geom_tile()
  ggplot(aes(reorder(TE, -log2FoldChange.y), fp_percent, label = TF)) + 
  geom_boxplot() +
  geom_point() #+
  ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange.x < -1.5, TF, '')))
  
  
tmp %>% 
  filter(padj.x < 0.05) %>% 
  # mutate(TF = ifelse(TF %in% c('ZNF682', 'ZNF257', 'ZNF492', 'ZNF90'), 
  #                    TF, 'other')) %>% 
  ggplot(aes(reorder(TE, -log2FoldChange.y), log2FoldChange.x)) + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_point() +
  # geom_line(aes(group = TF)) +
  ggrepel::geom_text_repel(aes(label = ifelse(log2FoldChange.x < -4, TF, ''))) + 
  # facet_row(~clade, scales = 'free_x') + 
  ylab('log2FoldChange') + 
  xlab('') + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1))

panel.save(figure = 4, 
           name = 'znf_TF_enrichment', width = 'single', height = 'single')


tmp %>% 
  select(TF, log2FoldChange.x, padj.x, clade, context) %>% 
  filter(clade %in% c('LTR', 'LINE'), grepl('aale', context)) %>% 
  distinct() %>% 
  group_by(TF) %>% 
  summarise(log2FoldChange = median(log2FoldChange.x), 
            padj = median(padj.x), 
            clade = list(clade)) %>% 
  mutate(gene_set = ifelse(TF %in% c(ifn_a, ifn_g), 'IFN', 'other')) %>% 
  mutate(gene_set = ifelse(TF %in% meta.data$trono.krab.znfs$gene, 'ZNF', gene_set)) %>%
  mutate(gene_set = ifelse(TF %in% msig.df$HALLMARK_KRAS_SIGNALING_UP, 'KRAS', gene_set)) %>%
  mutate(gene_set = factor(gene_set, levels = c('KRAS', 'IFN', 'ZNF', 'other'))) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = gene_set)) + 
  geom_point(alpha = 0.4) + 
  scale_color_manual(values = c(viridis(4)[1], viridis(4)[2], viridis(4)[3], 'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(gene_set != 'other' | TF == 'FOS' | abs(log2FoldChange) > 3, TF, '')), 
                               show.legend = F,
                               max.overlaps = 20) + 
  theme(legend.position = c(0.99,0.75))


panel.save(figure = 4, name = 'ltr_line_TF_enrichment', width = 'single', height = 'single')

te_ranges_list %>% 
  lapply(function(te_ranges) {
    
    
    if(length(te_ranges) > 0 & length(te_ranges) < 4000) {
      
      names(te_ranges) %>% unique() -> name
      
      print(name)
      print(length(te_ranges))
      
      getSeq(BSgenome.Hsapiens.UCSC.hg38, te_ranges) -> te_seqs

      memes::runAme(te_seqs, control = dn_te_seqs, 
                    database = file.path(output.data.dir, 'reference.data', 
                            'Hughes_ZNF_motifs.meme')) -> te_motif_enrich

      te_motif_enrich %>%
        mutate(tmp = motif_id,
               TE = name) %>%
        separate(tmp, sep = '_', into = c('TF', NA)) -> ame_tf
      # %>%
      #   mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
      #          TF = str_replace(TF, pattern = 'ZNFF', replacement = 'ZNF'),
      #          TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ame_tf

      ame_tf
    
    }
  }) -> ame_tf_list

ame_tf_list %>% 
  bind_rows() %>% 
  filter(adj.pvalue < 0.05) %>% 
  merge(salmon.quant$aale$intra$deseq$de, by.x = 'TF', by.y = 'gene') %>%
  merge(meta.data$te.families.clades, by.x = 'TE', by.y = 'gene') %>% 
  merge(upreg_te, by.x = 'TE', by.y = 'gene') -> tmp

tmp %>% 
  select(TF, log2FoldChange.x, padj.x, clade, context) %>% 
  filter(clade %in% c('LTR', 'LINE', "DNA"), grepl('aale', context)) %>% 
  distinct() %>% 
  group_by(TF) %>% 
  summarise(log2FoldChange = median(log2FoldChange.x), 
            padj = median(padj.x), 
            clade = list(clade)) %>% 
  mutate(gene_set = ifelse(TF %in% c(ifn_a, ifn_g), 'IFN', 'other')) %>% 
  mutate(gene_set = ifelse(TF %in% meta.data$trono.krab.znfs$gene, 'ZNF', gene_set)) %>%
  mutate(gene_set = ifelse(TF %in% msig.df$HALLMARK_KRAS_SIGNALING_UP, 'KRAS', gene_set)) %>%
  mutate(gene_set = factor(gene_set, levels = c('KRAS', 'IFN', 'ZNF', 'other'))) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = gene_set)) + 
  geom_point(alpha = 0.4) + 
  scale_color_manual(values = c(viridis(4)[1], viridis(4)[2], viridis(4)[3], 'grey')) +
  ggrepel::geom_text_repel(aes(label = ifelse(gene_set != 'other' | TF == 'FOS' | abs(log2FoldChange) > 3, TF, '')), 
                               show.legend = F,
                               max.overlaps = 20) + 
  theme(legend.position = c(0.99,0.75))

tmp %>% 
  filter(grepl('ZNF', TF),
         grepl('aale', context),
         padj.x < 0.05) %>% 
  ggplot(aes(reorder(TF, log2FoldChange.x), log2FoldChange.y, color = context, label = TE)) + 
  geom_boxplot(fill = NA, outlier.colour = NA, position = position_dodge(width = 1)) + 
  geom_point(position = position_jitterdodge(jitter.width = 0.25, dodge.width = 1)) +
  ggrepel::geom_text_repel() +
  theme(axis.text.x = element_text(angle = 40, hjust = 1))


de_list_promoter_seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, de_promoter_regions)

MotifDb::query(MotifDb::MotifDb, "ZNF257") %>% 
  universalmotif::convert_motifs()

universalmotif::read_meme(file.path(output.data.dir, 'reference.data', 
                      'Hughes_ZNF_motifs.meme')) ->znf_meme

memes::plot_sequence_heatmap(te_seqs)

znf_meme[60]

znf_meme[192]




memes::runFimo(sequences = 
                 de_list_promoter_seqs[names(de_list_promoter_seqs) == 'ENSG00000089127.14'],
               motifs = znf_meme[60]) -> ifn_znf_motif_enrich

znf_score <- read_csv(file = file.path(output.data.dir, 'all.znf.te.score.csv'))

znf_score %>% 
  dplyr::rename('gene' = '...1') -> znf_score


znf_score %>% 
  gather(TE, score, -gene) %>% 
  group_by(gene) %>% 
  filter(score > 0) %>% 
  mutate(rank = dense_rank(-score)) %>% 
  filter(!grepl('Mamrep', TE)) %>% 
  separate(TE, sep = '\\/', into = c('clade', 'family', 'TE')) %>% 
  filter(gene %in% filter(salmon.quant$aale$intra$deseq$de, padj < 0.05, 
                          log2FoldChange < -0.75)$gene) -> znf_score


znf_score %>% 
  merge(ucsc.rmsk.salmon.quant$aale$exo$deseq$de %>% 
          filter(padj < 0.05, log2FoldChange > 0.5), by.x = 'TE', by.y = 'gene') %>% 
  merge(salmon.quant$aale$intra$deseq$de, by = 'gene') %>% 
  filter(rank < 50) %>% 
  ggplot(aes(TE, rank, color = log2FoldChange.y)) + 
  geom_point() + 
  scale_color_viridis_c(name = 'znf log2FC') +
  ggrepel::geom_text_repel(aes(label = ifelse(rank < 20, gene, ''))) +
  xlab('') + 
  theme(legend.position = c(0.99,0.99),
        axis.text.x = element_text(angle = 40, hjust = 1))

panel.save(figure = 4, name = 'trono_znf_target_ranks', width = 'single', height = 'single')



znf_score %>%
  merge(upreg_te, by.x = "TE", by.y = 'gene') %>% 
  ggplot(aes(rank, score)) +
  # ggplot(aes(reorder(TE, rank), rank, label=gene)) + 
  # geom_boxplot(outlier.colour = NA, fill = NA) + 
  geom_point() #+ 
  # theme(axis.text.x = element_text(hjust = 1, angle = 40))

znf_score[, colnames(znf_score) %in% c('gene', upreg_te$gene)]
  
znf_score %>% 
  group_by(gene) %>% 
  filter(TE %in% filter(upreg_te, grepl('aale', context))$gene) %>% 
  group_by(gene) %>% 
  top_n(1, wt = rank) %>% View()
  merge(salmon.quant$aale$intra$deseq$de, by = 'gene') %>% 
  filter(padj < 0.05) %>%  
  ggplot(aes(rank, log2FoldChange, color = clade, label = paste0(gene, ':', TE))) + 
  geom_point() + 
  ggrepel::geom_text_repel()

  


ucsc.rmsk.salmon.quant$aale$exo$deseq$de %>% 
           filter(!gene %in% meta.data$gencode.35.tx.to.gene$gene,
                  padj < 0.05, log2FoldChange > 1) -> upreg_te


rmsk_parsed_bed %>% 
  filter(gene %in% upreg_te$gene,
         grepl('LTR27C', gene)) %>% 
  GenomicRanges::makeGRangesFromDataFrame() -> te_ranges

names(te_ranges) <- rmsk_parsed_bed %>% filter(gene %in% upreg_te$gene, grepl('LTR27C', gene)) %>% select(gene) %>% unlist() %>% unname()
  

getSeq(BSgenome.Hsapiens.UCSC.hg38, te_ranges) -> te_seqs

# memes::runDreme(te_seqs, control = 'shuffle') -> te_motif_denovo

memes::runAme(te_seqs, control = 'shuffle') -> te_motif_enrich

te_motif_enrich %>% 
  mutate(tmp = motif_id) %>% 
  separate(tmp, sep = '_', into = c('TF', NA)) %>% 
  mutate(TF = str_replace(TF, pattern = '^ZN', replacement = 'ZNF'),
         TF = str_replace(TF, pattern = '^NFAC', replacement = 'NFATC')) -> ame_tf
# %>% 
  # select(TF) %>% unlist() %>% unname() -> ame_tf

salmon.quant$aale$intra$deseq$de %>% 
  merge(ame_tf, by.x = 'gene', by.y = 'TF') %>% 
  # filter(grepl('ZNF', gene)) %>%
  ggplot(aes(log2FoldChange, -log10(padj), color = adj.pvalue, label = gene)) + 
  geom_point() + 
  scale_color_viridis_c(direction = -1, name = 'ame adj.pval') +
  geom_text_repel(color = 'black') + 
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') + 
  theme(legend.position = c(0.99, 0.99)) 

  
```

